<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Remade version</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../../../">
<script src="../../../../site_libs/quarto-html/quarto.js"></script>
<script src="../../../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<script src="../../../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../../../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../../../../styles.css">
</head>

<body class="nav-fixed fullcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../../../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../../posts/DanielKick/index.html"> 
<span class="menu-text">Posts</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="https://github.com/danielkick/Resume/blob/main/Resume.pdf"> 
<span class="menu-text">Resume</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://github.com/danielkick/Resume/blob/main/Curriculum-Vitae.pdf"> 
<span class="menu-text">CV</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/danielkick"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/in/daniel-kick-5a449b9a/"> <i class="bi bi-linkedin" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Remade version</h1>
                      </div>
  </div>
    
  
  <div class="quarto-title-meta">

      
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<div id="c9a53a2d" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1"></a><span class="im">import</span> json, re, os</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="im">from</span> datetime <span class="im">import</span> datetime</span>
<span id="cb1-3"><a href="#cb1-3"></a></span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="kw">def</span> read_json(json_path):</span>
<span id="cb1-5"><a href="#cb1-5"></a>    <span class="cf">with</span> <span class="bu">open</span>(json_path, <span class="st">'r'</span>) <span class="im">as</span> f:</span>
<span id="cb1-6"><a href="#cb1-6"></a>        dat <span class="op">=</span> json.load(f)</span>
<span id="cb1-7"><a href="#cb1-7"></a>    <span class="cf">return</span> dat</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="e4facc25" class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1"></a><span class="co"># os.listdir('/Users/danielkick/Downloads/Schulz Lab Slack export Jul 19 2016 - Apr 9 2024/')</span></span>
<span id="cb2-2"><a href="#cb2-2"></a>os.listdir(<span class="st">'./Schulz Lab Slack export Jul 19 2016 - Apr 9 2024/'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="3">
<pre><code>['daveiscool',
 'canvases.json',
 'integration_logs.json',
 'data_analysis',
 '.DS_Store',
 'interesting-reads',
 'general',
 'mpg_reads',
 'schulzlab_science',
 'channels.json',
 'users.json',
 'better_code',
 'model_systems',
 'professional-development',
 'ephys',
 'tacos',
 'random',
 'stgmeeting2020']</code></pre>
</div>
</div>
<div id="5b13386b" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1"></a><span class="co"># loc = '/Users/danielkick/Downloads/Schulz Lab Slack export Jul 19 2016 - Apr 9 2024/better_code/'</span></span>
<span id="cb4-2"><a href="#cb4-2"></a><span class="co"># loc_assets = '/Users/danielkick/Downloads/better_code_assets/'</span></span>
<span id="cb4-3"><a href="#cb4-3"></a></span>
<span id="cb4-4"><a href="#cb4-4"></a>loc <span class="op">=</span> <span class="st">'./Schulz Lab Slack export Jul 19 2016 - Apr 9 2024/better_code/'</span></span>
<span id="cb4-5"><a href="#cb4-5"></a>loc_assets <span class="op">=</span> <span class="st">'./better_code_assets/'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="3380e8ad" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1"></a>posts <span class="op">=</span> os.listdir(loc)</span>
<span id="cb5-2"><a href="#cb5-2"></a>posts.sort()</span>
<span id="cb5-3"><a href="#cb5-3"></a>posts[<span class="dv">0</span>:<span class="dv">3</span>]<span class="op">+</span>posts[<span class="op">-</span><span class="dv">4</span>:<span class="op">-</span><span class="dv">1</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="5">
<pre><code>['2020-04-01.json',
 '2020-04-10.json',
 '2020-04-20.json',
 '2023-04-06.json',
 '2023-05-23.json',
 '2023-07-12.json']</code></pre>
</div>
</div>
<div id="6c6a7ddb" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1"></a>posts <span class="op">=</span> [read_json(loc<span class="op">+</span>e) <span class="cf">for</span> e <span class="kw">in</span> posts]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="8f59d87c" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1"></a>post <span class="op">=</span> posts[<span class="dv">1</span>]</span>
<span id="cb8-2"><a href="#cb8-2"></a>post <span class="op">=</span> [e <span class="cf">for</span> e <span class="kw">in</span> post <span class="cf">if</span> <span class="st">'subtype'</span> <span class="kw">not</span> <span class="kw">in</span> e.keys()]</span>
<span id="cb8-3"><a href="#cb8-3"></a><span class="bu">len</span>(post)</span>
<span id="cb8-4"><a href="#cb8-4"></a></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="8">
<pre><code>1</code></pre>
</div>
</div>
<div id="55a49829" class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1"></a><span class="kw">def</span> format_entry(entry <span class="op">=</span> post[<span class="dv">0</span>]):</span>
<span id="cb10-2"><a href="#cb10-2"></a>    out <span class="op">=</span> []</span>
<span id="cb10-3"><a href="#cb10-3"></a></span>
<span id="cb10-4"><a href="#cb10-4"></a>    ts <span class="op">=</span> datetime.fromtimestamp(<span class="bu">int</span>(entry[<span class="st">'ts'</span>].split(<span class="st">'.'</span>)[<span class="dv">0</span>])).date()</span>
<span id="cb10-5"><a href="#cb10-5"></a>    ts <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>ts<span class="sc">.</span>year<span class="sc">}</span><span class="ss">-</span><span class="sc">{</span>ts<span class="sc">.</span>month<span class="sc">}</span><span class="ss">-</span><span class="sc">{</span>ts<span class="sc">.</span>day<span class="sc">}</span><span class="ss">'</span></span>
<span id="cb10-6"><a href="#cb10-6"></a>    ts <span class="op">=</span> <span class="ss">f'&lt;span style="color: #33b2ff"&gt;</span><span class="sc">{</span>ts<span class="sc">}</span><span class="ss">&lt;/span&gt;'</span></span>
<span id="cb10-7"><a href="#cb10-7"></a>    </span>
<span id="cb10-8"><a href="#cb10-8"></a>    out <span class="op">+=</span> [ts]</span>
<span id="cb10-9"><a href="#cb10-9"></a></span>
<span id="cb10-10"><a href="#cb10-10"></a>    <span class="cf">if</span> <span class="st">'user_profile'</span> <span class="kw">in</span> entry.keys():</span>
<span id="cb10-11"><a href="#cb10-11"></a>        name <span class="op">=</span> entry[<span class="st">'user_profile'</span>][<span class="st">'display_name'</span>]</span>
<span id="cb10-12"><a href="#cb10-12"></a>        name <span class="op">=</span> <span class="ss">f'&lt;span style="color: #33b2ff"&gt;</span><span class="sc">{</span>name<span class="sc">}</span><span class="ss">&lt;/span&gt;'</span></span>
<span id="cb10-13"><a href="#cb10-13"></a>        out <span class="op">+=</span> [name]</span>
<span id="cb10-14"><a href="#cb10-14"></a></span>
<span id="cb10-15"><a href="#cb10-15"></a>    text <span class="op">=</span> entry[<span class="st">'text'</span>]</span>
<span id="cb10-16"><a href="#cb10-16"></a>    out <span class="op">+=</span> [text]</span>
<span id="cb10-17"><a href="#cb10-17"></a></span>
<span id="cb10-18"><a href="#cb10-18"></a>    <span class="cf">if</span> <span class="st">'files'</span> <span class="kw">in</span> entry.keys():</span>
<span id="cb10-19"><a href="#cb10-19"></a>        <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(entry[<span class="st">'files'</span>])):</span>
<span id="cb10-20"><a href="#cb10-20"></a>            out <span class="op">+=</span> [<span class="ss">f'![</span><span class="sc">{</span>entry[<span class="st">"files"</span>][i][<span class="st">"title"</span>]<span class="sc">}</span><span class="ss">](</span><span class="sc">{</span>entry[<span class="st">"files"</span>][i][<span class="st">"name"</span>]<span class="sc">}</span><span class="ss">)'</span>]</span>
<span id="cb10-21"><a href="#cb10-21"></a></span>
<span id="cb10-22"><a href="#cb10-22"></a>    out <span class="op">=</span> <span class="st">'</span><span class="ch">\n</span><span class="st">'</span>.join(out)</span>
<span id="cb10-23"><a href="#cb10-23"></a>    <span class="cf">return</span> out</span>
<span id="cb10-24"><a href="#cb10-24"></a></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="af66539d" class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1"></a></span>
<span id="cb11-2"><a href="#cb11-2"></a><span class="kw">def</span> format_post(post):</span>
<span id="cb11-3"><a href="#cb11-3"></a>    post <span class="op">=</span> [e <span class="cf">for</span> e <span class="kw">in</span> post <span class="cf">if</span> <span class="st">'subtype'</span> <span class="kw">not</span> <span class="kw">in</span> e.keys()]</span>
<span id="cb11-4"><a href="#cb11-4"></a>    <span class="co"># print(len(post))</span></span>
<span id="cb11-5"><a href="#cb11-5"></a>    post_out <span class="op">=</span> [<span class="st">'</span><span class="ch">\n</span><span class="st">---</span><span class="ch">\n</span><span class="st">'</span>]<span class="op">+</span>[format_entry(entry <span class="op">=</span> e) <span class="cf">for</span> e <span class="kw">in</span> post]</span>
<span id="cb11-6"><a href="#cb11-6"></a>    <span class="cf">return</span> post_out</span>
<span id="cb11-7"><a href="#cb11-7"></a></span>
<span id="cb11-8"><a href="#cb11-8"></a>formatted_posts <span class="op">=</span> [format_post(post <span class="op">=</span> e) <span class="cf">for</span> e <span class="kw">in</span> posts]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="ee6d41ee" class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1"></a>formatted_posts <span class="op">=</span> <span class="bu">sum</span>(formatted_posts, [])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="fc803f7e" class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1"></a><span class="co"># idx_w_name_image = [e for e in range(len(formatted_posts)) if re.search('image.png', formatted_posts[e])]</span></span>
<span id="cb13-2"><a href="#cb13-2"></a><span class="co"># idx_w_name_image = [e for e in range(len(formatted_posts)) if re.finditer('image.png', formatted_posts[e])]</span></span>
<span id="cb13-3"><a href="#cb13-3"></a></span>
<span id="cb13-4"><a href="#cb13-4"></a><span class="co"># len(idx_w_name_image)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="3e7a3b3d" class="cell" data-execution_count="17">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1"></a><span class="co"># len([e for e in os.listdir(loc_assets) if re.match('^image.*', e)])</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="12bc0499" class="cell" data-execution_count="18">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1"></a><span class="co"># [e for e in re.finditer('a', 'a a a')]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="9a3a7fa8" class="cell" data-execution_count="19">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1"></a><span class="co"># formatted_posts[7]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="f2d683f3" class="cell" data-execution_count="20">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1"></a><span class="co"># detect = re.search('image.png', formatted_posts[7])</span></span>
<span id="cb17-2"><a href="#cb17-2"></a><span class="co"># detect</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="b834c656" class="cell" data-execution_count="27">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="27">
<pre><code>['\n---\n',
 '&lt;span style="color: #33b2ff"&gt;2020-4-10&lt;/span&gt;\nIf you\'re iteratively making plots, resampling, or doing another task that would have your reach for a for loop or `lapply()` use `furrr::future_map()` instead. `furrr` gives parallel processing ready versions of tidyverse\'s `purrr` functions (e.g. `map()`,  `walk()`). It\'s easy to install the dependencies and takes a lot of the headache out of parallel processing.\n![cpu_use_1.PNG](cpu_use_1.PNG)\n![cpu_use_0.PNG](cpu_use_0.PNG)']</code></pre>
</div>
</div>
<div id="9dd88768" class="cell" data-scrolled="false" data-execution_count="2">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1"></a></span>
<span id="cb20-2"><a href="#cb20-2"></a>frontmatter <span class="op">=</span> <span class="st">"""---</span></span>
<span id="cb20-3"><a href="#cb20-3"></a><span class="st">title: "better_code"</span></span>
<span id="cb20-4"><a href="#cb20-4"></a><span class="st">format: html</span></span>
<span id="cb20-5"><a href="#cb20-5"></a><span class="st">editor: source</span></span>
<span id="cb20-6"><a href="#cb20-6"></a><span class="st">---"""</span></span>
<span id="cb20-7"><a href="#cb20-7"></a></span>
<span id="cb20-8"><a href="#cb20-8"></a>write_text <span class="op">=</span> frontmatter<span class="op">+</span><span class="st">'</span><span class="ch">\n</span><span class="st">'</span><span class="op">+</span><span class="st">'</span><span class="ch">\n\n</span><span class="st">'</span>.join(formatted_posts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<div class="ansi-escaped-output">
<pre><span class="ansi-red-fg">---------------------------------------------------------------------------</span>
<span class="ansi-red-fg">NameError</span>                                 Traceback (most recent call last)
Cell <span class="ansi-green-fg">In[2], line 7</span>
<span class="ansi-green-fg ansi-bold">      1</span> frontmatter <span style="color:rgb(98,98,98)">=</span> <span style="color:rgb(175,0,0)">"""</span><span style="color:rgb(175,0,0)">---</span>
<span class="ansi-green-fg ansi-bold">      2</span> <span style="color:rgb(175,0,0)">title: </span><span style="color:rgb(175,0,0)">"</span><span style="color:rgb(175,0,0)">better_code</span><span style="color:rgb(175,0,0)">"</span>
<span class="ansi-green-fg ansi-bold">      3</span> <span style="color:rgb(175,0,0)">format: html</span>
<span class="ansi-green-fg ansi-bold">      4</span> <span style="color:rgb(175,0,0)">editor: source</span>
<span class="ansi-green-fg ansi-bold">      5</span> <span style="color:rgb(175,0,0)">---</span><span style="color:rgb(175,0,0)">"""</span>
<span class="ansi-green-fg">----&gt; 7</span> write_text <span style="color:rgb(98,98,98)">=</span> frontmatter<span style="color:rgb(98,98,98)">+</span><span style="color:rgb(175,0,0)">'</span><span style="font-weight:bold;color:rgb(175,95,0)">\n</span><span style="color:rgb(175,0,0)">'</span><span style="color:rgb(98,98,98)">+</span><span style="color:rgb(175,0,0)">'</span><span style="font-weight:bold;color:rgb(175,95,0)">\n</span><span style="font-weight:bold;color:rgb(175,95,0)">\n</span><span style="color:rgb(175,0,0)">'</span><span style="color:rgb(98,98,98)">.</span>join(<span class="ansi-yellow-bg">formatted_posts</span>)

<span class="ansi-red-fg">NameError</span>: name 'formatted_posts' is not defined</pre>
</div>
</div>
</div>
<div id="81fe3b7a" class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1"></a><span class="co"># fix code blocks</span></span>
<span id="cb21-2"><a href="#cb21-2"></a>write_text <span class="op">=</span> write_text.replace(<span class="st">'```'</span>, <span class="st">'</span><span class="ch">\n</span><span class="st">```</span><span class="ch">\n</span><span class="st">'</span>)</span>
<span id="cb21-3"><a href="#cb21-3"></a>write_text <span class="op">=</span> write_text.replace(<span class="st">'&amp;gt;'</span>, <span class="st">'&gt;'</span>)</span>
<span id="cb21-4"><a href="#cb21-4"></a>write_text <span class="op">=</span> write_text.replace(<span class="st">'&amp;lt;'</span>, <span class="st">'&lt;'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="86168115" class="cell" data-execution_count="24">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1"></a>write_text <span class="op">=</span> write_text.split(<span class="st">'```'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="8342e93c" class="cell" data-execution_count="25">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1"></a>write_text <span class="op">=</span> <span class="st">''</span>.join([write_text[i]<span class="op">+</span><span class="st">'``` r'</span> <span class="cf">if</span> (i <span class="op">%</span> <span class="dv">2</span>) <span class="op">==</span> <span class="dv">0</span> <span class="cf">else</span> write_text[i]<span class="op">+</span><span class="st">'```'</span> <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(write_text))])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="0db4e82d" class="cell" data-execution_count="26">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="26">
<pre><code>'./better_code_assets/'</code></pre>
</div>
</div>
<div id="c8e7bd19" class="cell" data-execution_count="27">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1"></a><span class="cf">with</span> <span class="bu">open</span>(loc_assets<span class="op">+</span><span class="st">'temp.qmd'</span>, <span class="st">'w'</span>) <span class="im">as</span> f:</span>
<span id="cb26-2"><a href="#cb26-2"></a>    f.write(write_text)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="97b73129" class="cell" data-execution_count="28">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1"></a><span class="co"># replace image file names with 'image (x).png' for all but the first entry</span></span>
<span id="cb27-2"><a href="#cb27-2"></a></span>
<span id="cb27-3"><a href="#cb27-3"></a>tmp <span class="op">=</span> write_text.split(<span class="st">'[image.png](image.png)'</span>)</span>
<span id="cb27-4"><a href="#cb27-4"></a></span>
<span id="cb27-5"><a href="#cb27-5"></a>tmp <span class="op">=</span> [</span>
<span id="cb27-6"><a href="#cb27-6"></a>    [tmp[i], <span class="ss">f'[image.png](image.png)'</span>] <span class="cf">if</span> i <span class="op">==</span> <span class="dv">0</span> <span class="cf">else</span> </span>
<span id="cb27-7"><a href="#cb27-7"></a>    [tmp[i], <span class="ss">f'[image (</span><span class="sc">{</span>i<span class="sc">}</span><span class="ss">).png](image%20(</span><span class="sc">{</span>i<span class="sc">}</span><span class="ss">).png)'</span>] </span>
<span id="cb27-8"><a href="#cb27-8"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(tmp))]</span>
<span id="cb27-9"><a href="#cb27-9"></a></span>
<span id="cb27-10"><a href="#cb27-10"></a>tmp <span class="op">=</span> <span class="bu">sum</span>(tmp, [])[<span class="dv">0</span>:<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb27-11"><a href="#cb27-11"></a></span>
<span id="cb27-12"><a href="#cb27-12"></a>tmp <span class="op">=</span> <span class="st">''</span>.join(tmp)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="9c702b08" class="cell" data-execution_count="29">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1"></a><span class="cf">with</span> <span class="bu">open</span>(loc_assets<span class="op">+</span><span class="st">'temp.qmd'</span>, <span class="st">'w'</span>) <span class="im">as</span> f:</span>
<span id="cb28-2"><a href="#cb28-2"></a>    f.write(tmp)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="fb861dfc" class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1"></a></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="bae4478e" class="cell" data-scrolled="false" data-execution_count="24">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1"></a></span>
<span id="cb30-2"><a href="#cb30-2"></a><span class="co"># %20(43)</span></span>
<span id="cb30-3"><a href="#cb30-3"></a></span>
<span id="cb30-4"><a href="#cb30-4"></a></span>
<span id="cb30-5"><a href="#cb30-5"></a><span class="co"># if no name use previous name?</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="993976ba" class="cell" data-scrolled="false" data-execution_count="25">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1"></a><span class="co"># posts[1]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="942ed59a" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1"></a><span class="im">import</span> shutil</span>
<span id="cb32-2"><a href="#cb32-2"></a><span class="kw">def</span> mkdir(path):</span>
<span id="cb32-3"><a href="#cb32-3"></a>    <span class="cf">if</span> <span class="kw">not</span> os.path.exists(path):</span>
<span id="cb32-4"><a href="#cb32-4"></a>        os.makedirs(path)</span>
<span id="cb32-5"><a href="#cb32-5"></a></span>
<span id="cb32-6"><a href="#cb32-6"></a>mkdir(path <span class="op">=</span> <span class="st">'./output/'</span>)    </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="89fcc5e9" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1"></a>loc <span class="op">=</span> <span class="st">'./Schulz Lab Slack export Jul 19 2016 - Apr 9 2024/better_code/'</span></span>
<span id="cb33-2"><a href="#cb33-2"></a>loc_assets <span class="op">=</span> <span class="st">'./better_code_assets/'</span></span>
<span id="cb33-3"><a href="#cb33-3"></a></span>
<span id="cb33-4"><a href="#cb33-4"></a><span class="co"># loc = './Schulz Lab Slack export Jul 19 2016 - Apr 9 2024/data_analysis/'</span></span>
<span id="cb33-5"><a href="#cb33-5"></a><span class="co"># loc_assets = './data_analysis_assets/'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="3332a2a2" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1"></a><span class="co"># e = '2023-07-12.json' </span></span>
<span id="cb34-2"><a href="#cb34-2"></a>entries <span class="op">=</span> [read_json(loc<span class="op">+</span>e) <span class="cf">for</span> e <span class="kw">in</span> os.listdir(loc)]</span>
<span id="cb34-3"><a href="#cb34-3"></a>entries <span class="op">=</span> [[ee <span class="cf">for</span> ee <span class="kw">in</span> e <span class="cf">if</span> <span class="st">'subtype'</span> <span class="kw">not</span> <span class="kw">in</span> ee.keys()]</span>
<span id="cb34-4"><a href="#cb34-4"></a>            <span class="cf">for</span> e <span class="kw">in</span> entries]</span>
<span id="cb34-5"><a href="#cb34-5"></a>entries <span class="op">=</span> [ee[<span class="dv">0</span>] <span class="cf">for</span> ee <span class="kw">in</span> entries <span class="cf">if</span> ee <span class="op">!=</span> []]</span>
<span id="cb34-6"><a href="#cb34-6"></a></span>
<span id="cb34-7"><a href="#cb34-7"></a></span>
<span id="cb34-8"><a href="#cb34-8"></a></span>
<span id="cb34-9"><a href="#cb34-9"></a><span class="kw">def</span> parse_entry(entry):</span>
<span id="cb34-10"><a href="#cb34-10"></a>    ts <span class="op">=</span> datetime.fromtimestamp(<span class="bu">int</span>(entry[<span class="st">'ts'</span>].split(<span class="st">'.'</span>)[<span class="dv">0</span>])).date()</span>
<span id="cb34-11"><a href="#cb34-11"></a>    ts <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>ts<span class="sc">.</span>year<span class="sc">}</span><span class="ss">-</span><span class="sc">{</span>ts<span class="sc">.</span>month<span class="sc">}</span><span class="ss">-</span><span class="sc">{</span>ts<span class="sc">.</span>day<span class="sc">}</span><span class="ss">'</span></span>
<span id="cb34-12"><a href="#cb34-12"></a></span>
<span id="cb34-13"><a href="#cb34-13"></a>    <span class="co"># name = entry['user_profile']['display_name']</span></span>
<span id="cb34-14"><a href="#cb34-14"></a>    name <span class="op">=</span> entry[<span class="st">'user'</span>]</span>
<span id="cb34-15"><a href="#cb34-15"></a></span>
<span id="cb34-16"><a href="#cb34-16"></a>    text <span class="op">=</span> entry[<span class="st">'text'</span>]</span>
<span id="cb34-17"><a href="#cb34-17"></a></span>
<span id="cb34-18"><a href="#cb34-18"></a>    files <span class="op">=</span> []</span>
<span id="cb34-19"><a href="#cb34-19"></a>    <span class="cf">if</span> <span class="st">'files'</span> <span class="kw">in</span> entry.keys():</span>
<span id="cb34-20"><a href="#cb34-20"></a>        files <span class="op">=</span> [<span class="ss">f'![</span><span class="sc">{</span>entry[<span class="st">"files"</span>][i][<span class="st">"title"</span>]<span class="sc">}</span><span class="ss">](</span><span class="sc">{</span>entry[<span class="st">"files"</span>][i][<span class="st">"name"</span>]<span class="sc">}</span><span class="ss">)'</span> </span>
<span id="cb34-21"><a href="#cb34-21"></a>                <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(entry[<span class="st">'files'</span>]))]</span>
<span id="cb34-22"><a href="#cb34-22"></a></span>
<span id="cb34-23"><a href="#cb34-23"></a>    <span class="cf">return</span> (ts, name, text, files)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="6bfc9d88" class="cell" data-execution_count="139">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1"></a>parse_entry(entry<span class="op">=</span>entries[<span class="dv">1</span>])</span>
<span id="cb35-2"><a href="#cb35-2"></a><span class="co"># Is this me? U1T7R897Y </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="139">
<pre><code>('2022-5-26',
 'U1T7R897Y',
 'Late to the party but here’s my two cents:\n• The final plots are good. I personally use a point for the median but it works either way. \n• Nudging the statistics on the x axis can help with readability. To &lt;@U028ZEAHA6S&gt;’s point, raincloud plots are a nice way to incorporate a density/violin. Tufte style boxplots are another nice option (for maximizing info/ink). This doesn’t work so well as well with beeswarms.\n',
 [])</code></pre>
</div>
</div>
<div id="756639e6" class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1"></a>res <span class="op">=</span> []</span>
<span id="cb37-2"><a href="#cb37-2"></a></span>
<span id="cb37-3"><a href="#cb37-3"></a><span class="cf">for</span> json_entry <span class="kw">in</span> entries:</span>
<span id="cb37-4"><a href="#cb37-4"></a></span>
<span id="cb37-5"><a href="#cb37-5"></a>    ts, name, text, files <span class="op">=</span> parse_entry(entry<span class="op">=</span>json_entry)</span>
<span id="cb37-6"><a href="#cb37-6"></a></span>
<span id="cb37-7"><a href="#cb37-7"></a>    quarto_text <span class="op">=</span> <span class="st">'</span><span class="ch">\n</span><span class="st">'</span>.join([</span>
<span id="cb37-8"><a href="#cb37-8"></a>    <span class="st">'---'</span>,</span>
<span id="cb37-9"><a href="#cb37-9"></a>    <span class="ss">f'title: "Placeholder"'</span>,</span>
<span id="cb37-10"><a href="#cb37-10"></a>    <span class="ss">f'author: "</span><span class="sc">{</span>name<span class="sc">}</span><span class="ss">"'</span>,</span>
<span id="cb37-11"><a href="#cb37-11"></a>    <span class="ss">f'date: "</span><span class="sc">{</span>ts<span class="sc">}</span><span class="ss">"'</span>,</span>
<span id="cb37-12"><a href="#cb37-12"></a>    <span class="st">'image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_</span><span class="sc">%281880%</span><span class="st">29_p.41.jpg"'</span>,</span>
<span id="cb37-13"><a href="#cb37-13"></a>    <span class="st">'categories: '</span>,</span>
<span id="cb37-14"><a href="#cb37-14"></a>    <span class="st">'  - code'</span>,</span>
<span id="cb37-15"><a href="#cb37-15"></a>    <span class="st">'  - beginner'</span>,</span>
<span id="cb37-16"><a href="#cb37-16"></a>    <span class="st">'  - r'</span>,</span>
<span id="cb37-17"><a href="#cb37-17"></a>    <span class="st">'draft: false'</span>,</span>
<span id="cb37-18"><a href="#cb37-18"></a>    <span class="st">'editor: '</span>,</span>
<span id="cb37-19"><a href="#cb37-19"></a>    <span class="st">'  markdown: '</span>,</span>
<span id="cb37-20"><a href="#cb37-20"></a>    <span class="st">'    wrap: 72'</span>,</span>
<span id="cb37-21"><a href="#cb37-21"></a>    <span class="st">'---'</span>,</span>
<span id="cb37-22"><a href="#cb37-22"></a>    <span class="st">''</span>,</span>
<span id="cb37-23"><a href="#cb37-23"></a>    text,</span>
<span id="cb37-24"><a href="#cb37-24"></a>    <span class="st">''</span>, </span>
<span id="cb37-25"><a href="#cb37-25"></a>    (<span class="kw">lambda</span> x: <span class="st">''</span> <span class="cf">if</span> x <span class="op">==</span> [] <span class="cf">else</span> <span class="st">'</span><span class="ch">\n</span><span class="st">'</span>.join(x))(files)</span>
<span id="cb37-26"><a href="#cb37-26"></a>    ])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="737c6a56" class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1"></a>all_text <span class="op">=</span> [<span class="st">'</span><span class="ch">\n</span><span class="st">'</span>.join([</span>
<span id="cb38-2"><a href="#cb38-2"></a>    <span class="st">'---'</span>,</span>
<span id="cb38-3"><a href="#cb38-3"></a>    <span class="ss">f'author: "</span><span class="sc">{</span>name<span class="sc">}</span><span class="ss">"'</span>,</span>
<span id="cb38-4"><a href="#cb38-4"></a>    <span class="ss">f'date: "</span><span class="sc">{</span>ts<span class="sc">}</span><span class="ss">"'</span>,</span>
<span id="cb38-5"><a href="#cb38-5"></a>    <span class="st">'---'</span>,</span>
<span id="cb38-6"><a href="#cb38-6"></a>    <span class="st">''</span>,</span>
<span id="cb38-7"><a href="#cb38-7"></a>    text,</span>
<span id="cb38-8"><a href="#cb38-8"></a>    <span class="st">''</span>, </span>
<span id="cb38-9"><a href="#cb38-9"></a>    (<span class="kw">lambda</span> x: <span class="st">''</span> <span class="cf">if</span> x <span class="op">==</span> [] <span class="cf">else</span> <span class="st">'</span><span class="ch">\n</span><span class="st">'</span>.join(x))(files)</span>
<span id="cb38-10"><a href="#cb38-10"></a>    ]) <span class="cf">for</span> ts, name, text, files <span class="kw">in</span> [parse_entry(entry<span class="op">=</span>json_entry) <span class="cf">for</span> json_entry <span class="kw">in</span> entries]]</span>
<span id="cb38-11"><a href="#cb38-11"></a></span>
<span id="cb38-12"><a href="#cb38-12"></a>all_text <span class="op">=</span> <span class="st">'</span><span class="ch">\n</span><span class="st">'</span>.join(all_text)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="72f13cd7" class="cell" data-execution_count="25">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1"></a></span>
<span id="cb39-2"><a href="#cb39-2"></a></span>
<span id="cb39-3"><a href="#cb39-3"></a>all_text <span class="op">=</span> <span class="st">"""</span></span>
<span id="cb39-4"><a href="#cb39-4"></a><span class="st">2020-4-1</span></span>
<span id="cb39-5"><a href="#cb39-5"></a><span class="st">Daniel</span></span>
<span id="cb39-6"><a href="#cb39-6"></a><span class="st">You can keep your working `Rmd` easier to navigate and less buggy by 1) packaging code into functions and 2) adding them to a companion `R` file. Load your functions with `source()` in the same block you load your libraries with a relative path, full path, or ideally with `here()`.</span></span>
<span id="cb39-7"><a href="#cb39-7"></a></span>
<span id="cb39-8"><a href="#cb39-8"></a><span class="st">``` r</span></span>
<span id="cb39-9"><a href="#cb39-9"></a><span class="st">library(here)</span></span>
<span id="cb39-10"><a href="#cb39-10"></a><span class="st">source(here("R", "02MoniterGapJunction.R")) #here's output is effectively ../R/02MoniterGapJunction.R</span></span>
<span id="cb39-11"><a href="#cb39-11"></a><span class="st">```</span></span>
<span id="cb39-12"><a href="#cb39-12"></a></span>
<span id="cb39-13"><a href="#cb39-13"></a></span>
<span id="cb39-14"><a href="#cb39-14"></a></span>
<span id="cb39-15"><a href="#cb39-15"></a><span class="st">-xcutx-</span></span>
<span id="cb39-16"><a href="#cb39-16"></a></span>
<span id="cb39-17"><a href="#cb39-17"></a></span>
<span id="cb39-18"><a href="#cb39-18"></a><span class="st">2020-4-10</span></span>
<span id="cb39-19"><a href="#cb39-19"></a><span class="st">If you're iteratively making plots, resampling, or doing another task that would have your reach for a for loop or `lapply()` use `furrr::future_map()` instead. `furrr` gives parallel processing ready versions of tidyverse's `purrr` functions (e.g. `map()`,  `walk()`). It's easy to install the dependencies and takes a lot of the headache out of parallel processing.</span></span>
<span id="cb39-20"><a href="#cb39-20"></a><span class="st">![cpu_use_1.PNG](cpu_use_1.PNG)</span></span>
<span id="cb39-21"><a href="#cb39-21"></a><span class="st">![cpu_use_0.PNG](cpu_use_0.PNG)</span></span>
<span id="cb39-22"><a href="#cb39-22"></a></span>
<span id="cb39-23"><a href="#cb39-23"></a></span>
<span id="cb39-24"><a href="#cb39-24"></a><span class="st">-xcutx-</span></span>
<span id="cb39-25"><a href="#cb39-25"></a></span>
<span id="cb39-26"><a href="#cb39-26"></a></span>
<span id="cb39-27"><a href="#cb39-27"></a><span class="st">2020-4-20</span></span>
<span id="cb39-28"><a href="#cb39-28"></a><span class="st">Daniel</span></span>
<span id="cb39-29"><a href="#cb39-29"></a><span class="st">If you're running low on memory, you can retain specific objects by running this:</span></span>
<span id="cb39-30"><a href="#cb39-30"></a></span>
<span id="cb39-31"><a href="#cb39-31"></a><span class="st">``` r</span></span>
<span id="cb39-32"><a href="#cb39-32"></a><span class="st"> </span></span>
<span id="cb39-33"><a href="#cb39-33"></a><span class="st"># get rid of everything</span></span>
<span id="cb39-34"><a href="#cb39-34"></a><span class="st">rm(list=ls())</span></span>
<span id="cb39-35"><a href="#cb39-35"></a></span>
<span id="cb39-36"><a href="#cb39-36"></a><span class="st"># get rid of everything except specific objects and all loaded functions</span></span>
<span id="cb39-37"><a href="#cb39-37"></a><span class="st">rm(list = </span></span>
<span id="cb39-38"><a href="#cb39-38"></a><span class="st">   ls()[!(ls() </span><span class="sc">%i</span><span class="st">n</span><span class="sc">% c</span><span class="st">(</span></span>
<span id="cb39-39"><a href="#cb39-39"></a><span class="st">   # objects</span></span>
<span id="cb39-40"><a href="#cb39-40"></a><span class="st">   c("data1", "data2", "bool1", "list1"), </span></span>
<span id="cb39-41"><a href="#cb39-41"></a><span class="st">   # functions</span></span>
<span id="cb39-42"><a href="#cb39-42"></a><span class="st">   lsf.str())</span></span>
<span id="cb39-43"><a href="#cb39-43"></a><span class="st">   ) ])</span></span>
<span id="cb39-44"><a href="#cb39-44"></a><span class="st">```</span></span>
<span id="cb39-45"><a href="#cb39-45"></a></span>
<span id="cb39-46"><a href="#cb39-46"></a><span class="st">-xcutx-</span></span>
<span id="cb39-47"><a href="#cb39-47"></a></span>
<span id="cb39-48"><a href="#cb39-48"></a><span class="st">2020-4-20</span></span>
<span id="cb39-49"><a href="#cb39-49"></a><span class="st">Daniel</span></span>
<span id="cb39-50"><a href="#cb39-50"></a><span class="st">`lsf.str()` returns a character string of the functions you've written. This doesnt' include functions from libraries so it may not be needed if you have used `source()` to pull your custom functions from `/R/`.</span></span>
<span id="cb39-51"><a href="#cb39-51"></a></span>
<span id="cb39-52"><a href="#cb39-52"></a><span class="st">2020-4-20</span></span>
<span id="cb39-53"><a href="#cb39-53"></a><span class="st">If you're tempted to use `rm(list = ls())` consider restarting your r session (`ctrl+shift+F10` on windows). Overreliance on `rm(list = ls())` is poor form and the Rstudio devs will know and find you.</span></span>
<span id="cb39-54"><a href="#cb39-54"></a><span class="st">![image.png](image.png)</span></span>
<span id="cb39-55"><a href="#cb39-55"></a></span>
<span id="cb39-56"><a href="#cb39-56"></a></span>
<span id="cb39-57"><a href="#cb39-57"></a><span class="st">-xcutx-</span></span>
<span id="cb39-58"><a href="#cb39-58"></a></span>
<span id="cb39-59"><a href="#cb39-59"></a></span>
<span id="cb39-60"><a href="#cb39-60"></a><span class="st">2020-4-21</span></span>
<span id="cb39-61"><a href="#cb39-61"></a><span class="st">Daniel</span></span>
<span id="cb39-62"><a href="#cb39-62"></a><span class="st">If you're applying the same theme to all your graphs, set it globally instead e.g. `theme_set(ggplot2::theme_minimal())`.  If you have a lot of custom changes to your theme, throw those into a function and set that to the global theme.</span></span>
<span id="cb39-63"><a href="#cb39-63"></a></span>
<span id="cb39-64"><a href="#cb39-64"></a></span>
<span id="cb39-65"><a href="#cb39-65"></a><span class="st">-xcutx-</span></span>
<span id="cb39-66"><a href="#cb39-66"></a></span>
<span id="cb39-67"><a href="#cb39-67"></a></span>
<span id="cb39-68"><a href="#cb39-68"></a><span class="st">2020-5-4</span></span>
<span id="cb39-69"><a href="#cb39-69"></a><span class="st">If you're on windows, `installr` should allow you to copy over the libraries from previous versions.</span></span>
<span id="cb39-70"><a href="#cb39-70"></a><span class="st">![image (1).png](image%20(1).png)</span></span>
<span id="cb39-71"><a href="#cb39-71"></a></span>
<span id="cb39-72"><a href="#cb39-72"></a></span>
<span id="cb39-73"><a href="#cb39-73"></a><span class="st">-xcutx-</span></span>
<span id="cb39-74"><a href="#cb39-74"></a></span>
<span id="cb39-75"><a href="#cb39-75"></a></span>
<span id="cb39-76"><a href="#cb39-76"></a><span class="st">2020-5-6</span></span>
<span id="cb39-77"><a href="#cb39-77"></a><span class="st">Don't get suckered when converting factors! Numeric data can be assigned to a factor type which will throw a wrench in a plot or analysis (1/3)</span></span>
<span id="cb39-78"><a href="#cb39-78"></a><span class="st">![image (2).png](image%20(2).png)</span></span>
<span id="cb39-79"><a href="#cb39-79"></a></span>
<span id="cb39-80"><a href="#cb39-80"></a><span class="st">2020-5-6</span></span>
<span id="cb39-81"><a href="#cb39-81"></a><span class="st">A knee jerk reaction would be to convert it to a numeric with `as.numeric()`. That doesn't work either.  (2/3)</span></span>
<span id="cb39-82"><a href="#cb39-82"></a><span class="st">![image (3).png](image%20(3).png)</span></span>
<span id="cb39-83"><a href="#cb39-83"></a></span>
<span id="cb39-84"><a href="#cb39-84"></a><span class="st">2020-5-6</span></span>
<span id="cb39-85"><a href="#cb39-85"></a><span class="st">However if you use `as.numeric(as.character())` then it works. That's because factors are ordinal and named so if you convert the type to character first to ensure R is working with the factor names instead of the ranks. (3/3)</span></span>
<span id="cb39-86"><a href="#cb39-86"></a><span class="st">![image (4).png](image%20(4).png)</span></span>
<span id="cb39-87"><a href="#cb39-87"></a></span>
<span id="cb39-88"><a href="#cb39-88"></a></span>
<span id="cb39-89"><a href="#cb39-89"></a><span class="st">-xcutx-</span></span>
<span id="cb39-90"><a href="#cb39-90"></a></span>
<span id="cb39-91"><a href="#cb39-91"></a></span>
<span id="cb39-92"><a href="#cb39-92"></a><span class="st">2020-5-11</span></span>
<span id="cb39-93"><a href="#cb39-93"></a><span class="st">Daniel</span></span>
<span id="cb39-94"><a href="#cb39-94"></a><span class="st">Need a quick base `R` syntax lookup? Check out &lt;https://learnxinyminutes.com/docs/r/&gt; .  There are even examples with `lm()` and `glm()`.</span></span>
<span id="cb39-95"><a href="#cb39-95"></a></span>
<span id="cb39-96"><a href="#cb39-96"></a><span class="st">-xcutx-</span></span>
<span id="cb39-97"><a href="#cb39-97"></a></span>
<span id="cb39-98"><a href="#cb39-98"></a><span class="st">2020-5-11</span></span>
<span id="cb39-99"><a href="#cb39-99"></a><span class="st">Daniel</span></span>
<span id="cb39-100"><a href="#cb39-100"></a><span class="st">tldr: `tidyverse` function variants take slightly different input. Testing out a variant or two can save you a lot of debugging time.</span></span>
<span id="cb39-101"><a href="#cb39-101"></a></span>
<span id="cb39-102"><a href="#cb39-102"></a><span class="st">In `tidyverse` watch out for inconsistencies in function versions. There are variants of common functions (e.g. `mutate()` `mutate_all()` `mutate_at()`) don't necessarily behave the same way (or how you would expect).</span></span>
<span id="cb39-103"><a href="#cb39-103"></a></span>
<span id="cb39-104"><a href="#cb39-104"></a><span class="st">Here, I was applying an operation to a grouped df where each `Experiment` contains several `FileName`s with multiple observations in each.  To keep everything reusable I'm using `exp` instead of `Experiment` to select the right col.</span></span>
<span id="cb39-105"><a href="#cb39-105"></a></span>
<span id="cb39-106"><a href="#cb39-106"></a><span class="st">``` r</span></span>
<span id="cb39-107"><a href="#cb39-107"></a><span class="st">&gt; exp = "Experiment"</span></span>
<span id="cb39-108"><a href="#cb39-108"></a><span class="st">&gt; rec = "FileName"</span></span>
<span id="cb39-109"><a href="#cb39-109"></a></span>
<span id="cb39-110"><a href="#cb39-110"></a><span class="st">&gt; df %&gt;%</span></span>
<span id="cb39-111"><a href="#cb39-111"></a><span class="st">+     dplyr::select(</span></span>
<span id="cb39-112"><a href="#cb39-112"></a><span class="st">+       exp, rec, r11, r12</span></span>
<span id="cb39-113"><a href="#cb39-113"></a><span class="st">+     )</span></span>
<span id="cb39-114"><a href="#cb39-114"></a><span class="st"># A tibble: 564 x 4</span></span>
<span id="cb39-115"><a href="#cb39-115"></a><span class="st">   Experiment FileName           r11   r12</span></span>
<span id="cb39-116"><a href="#cb39-116"></a><span class="st">   &lt;chr&gt;      &lt;chr&gt;            &lt;dbl&gt; &lt;dbl&gt;</span></span>
<span id="cb39-117"><a href="#cb39-117"></a><span class="st"> 1 190808a    190808a_0020.abf  5.06 0.150</span></span>
<span id="cb39-118"><a href="#cb39-118"></a><span class="st"> 2 190808a    190808a_0020.abf  5.13 0.160</span></span>
<span id="cb39-119"><a href="#cb39-119"></a><span class="st"> 3 190808a    190808a_0020.abf  5.11 0.122</span></span>
<span id="cb39-120"><a href="#cb39-120"></a><span class="st"> 4 190808a    190808a_0020.abf  2.49 0.152</span></span>
<span id="cb39-121"><a href="#cb39-121"></a><span class="st"> 5 190808a    190808a_0020.abf  2.49 0.195</span></span>
<span id="cb39-122"><a href="#cb39-122"></a><span class="st"># ... with 559 more rows</span></span>
<span id="cb39-123"><a href="#cb39-123"></a><span class="st">```</span></span>
<span id="cb39-124"><a href="#cb39-124"></a></span>
<span id="cb39-125"><a href="#cb39-125"></a><span class="st">As soon as we do the same thing with `group_by()` we don't get the right column even though `select()` didn't have an issue with `exp`.</span></span>
<span id="cb39-126"><a href="#cb39-126"></a></span>
<span id="cb39-127"><a href="#cb39-127"></a><span class="st">``` r</span></span>
<span id="cb39-128"><a href="#cb39-128"></a><span class="st">&gt; df %&gt;%</span></span>
<span id="cb39-129"><a href="#cb39-129"></a><span class="st">+     dplyr::select(</span></span>
<span id="cb39-130"><a href="#cb39-130"></a><span class="st">+       exp, rec, r11, r12</span></span>
<span id="cb39-131"><a href="#cb39-131"></a><span class="st">+     ) %&gt;%</span></span>
<span id="cb39-132"><a href="#cb39-132"></a><span class="st">+     group_by(exp, rec)</span></span>
<span id="cb39-133"><a href="#cb39-133"></a><span class="st">Error: Column `exp` is unknown</span></span>
<span id="cb39-134"><a href="#cb39-134"></a><span class="st">```</span></span>
<span id="cb39-135"><a href="#cb39-135"></a></span>
<span id="cb39-136"><a href="#cb39-136"></a><span class="st">So we can try explicitly selecting the columns we want as groupings.</span></span>
<span id="cb39-137"><a href="#cb39-137"></a></span>
<span id="cb39-138"><a href="#cb39-138"></a><span class="st">``` r</span></span>
<span id="cb39-139"><a href="#cb39-139"></a><span class="st">&gt;     df %&gt;%</span></span>
<span id="cb39-140"><a href="#cb39-140"></a><span class="st">+     dplyr::select(</span></span>
<span id="cb39-141"><a href="#cb39-141"></a><span class="st">+       exp, rec, r11, r12</span></span>
<span id="cb39-142"><a href="#cb39-142"></a><span class="st">+     ) %&gt;%</span></span>
<span id="cb39-143"><a href="#cb39-143"></a><span class="st">+     group_by(vars(exp, rec))</span></span>
<span id="cb39-144"><a href="#cb39-144"></a><span class="st">Error: Column `vars(exp, rec)` must be length 564 (the number of rows) or one, not 2</span></span>
<span id="cb39-145"><a href="#cb39-145"></a><span class="st">```</span></span>
<span id="cb39-146"><a href="#cb39-146"></a></span>
<span id="cb39-147"><a href="#cb39-147"></a><span class="st">No dice there. `vars()` is designed to work with the `_at` variants so we can try that.  _et voilà!_</span></span>
<span id="cb39-148"><a href="#cb39-148"></a></span>
<span id="cb39-149"><a href="#cb39-149"></a><span class="st">``` r</span></span>
<span id="cb39-150"><a href="#cb39-150"></a><span class="st">&gt; df %&gt;%</span></span>
<span id="cb39-151"><a href="#cb39-151"></a><span class="st">+     dplyr::select(</span></span>
<span id="cb39-152"><a href="#cb39-152"></a><span class="st">+       exp, rec, r11, r12</span></span>
<span id="cb39-153"><a href="#cb39-153"></a><span class="st">+     ) %&gt;%</span></span>
<span id="cb39-154"><a href="#cb39-154"></a><span class="st">+     group_by_at(vars(exp, rec))</span></span>
<span id="cb39-155"><a href="#cb39-155"></a></span>
<span id="cb39-156"><a href="#cb39-156"></a><span class="st"># A tibble: 564 x 4</span></span>
<span id="cb39-157"><a href="#cb39-157"></a><span class="st"># Groups:   Experiment, FileName [69]</span></span>
<span id="cb39-158"><a href="#cb39-158"></a><span class="st">   Experiment FileName           r11   r12</span></span>
<span id="cb39-159"><a href="#cb39-159"></a><span class="st">   &lt;chr&gt;      &lt;chr&gt;            &lt;dbl&gt; &lt;dbl&gt;</span></span>
<span id="cb39-160"><a href="#cb39-160"></a><span class="st"> 1 190808a    190808a_0020.abf  5.06 0.150</span></span>
<span id="cb39-161"><a href="#cb39-161"></a><span class="st"> 2 190808a    190808a_0020.abf  5.13 0.160</span></span>
<span id="cb39-162"><a href="#cb39-162"></a><span class="st"> 3 190808a    190808a_0020.abf  5.11 0.122</span></span>
<span id="cb39-163"><a href="#cb39-163"></a><span class="st"> 4 190808a    190808a_0020.abf  2.49 0.152</span></span>
<span id="cb39-164"><a href="#cb39-164"></a><span class="st"> 5 190808a    190808a_0020.abf  2.49 0.195</span></span>
<span id="cb39-165"><a href="#cb39-165"></a><span class="st"># ... with 559 more rows</span></span>
<span id="cb39-166"><a href="#cb39-166"></a><span class="st">```</span></span>
<span id="cb39-167"><a href="#cb39-167"></a></span>
<span id="cb39-168"><a href="#cb39-168"></a></span>
<span id="cb39-169"><a href="#cb39-169"></a></span>
<span id="cb39-170"><a href="#cb39-170"></a></span>
<span id="cb39-171"><a href="#cb39-171"></a><span class="st">-xcutx-</span></span>
<span id="cb39-172"><a href="#cb39-172"></a></span>
<span id="cb39-173"><a href="#cb39-173"></a></span>
<span id="cb39-174"><a href="#cb39-174"></a><span class="st">2020-5-12</span></span>
<span id="cb39-175"><a href="#cb39-175"></a><span class="st">Daniel</span></span>
<span id="cb39-176"><a href="#cb39-176"></a><span class="st">Save `R` objects as rds. These could be works in progress or items that take a while to generate (e.g. re-sampled results). This gives the utility of saving a work space without the dangers.</span></span>
<span id="cb39-177"><a href="#cb39-177"></a></span>
<span id="cb39-178"><a href="#cb39-178"></a><span class="st">``` r</span></span>
<span id="cb39-179"><a href="#cb39-179"></a><span class="st">library(here)</span></span>
<span id="cb39-180"><a href="#cb39-180"></a><span class="st">save(df, file = here("data", "df.rds"))</span></span>
<span id="cb39-181"><a href="#cb39-181"></a><span class="st"># df is now saved at ./data/df.rds</span></span>
<span id="cb39-182"><a href="#cb39-182"></a></span>
<span id="cb39-183"><a href="#cb39-183"></a><span class="st">load(here("data", "df.rds"))</span></span>
<span id="cb39-184"><a href="#cb39-184"></a><span class="st"># df is now loaded from ./data/df.rds as df</span></span>
<span id="cb39-185"><a href="#cb39-185"></a><span class="st">```</span></span>
<span id="cb39-186"><a href="#cb39-186"></a></span>
<span id="cb39-187"><a href="#cb39-187"></a></span>
<span id="cb39-188"><a href="#cb39-188"></a></span>
<span id="cb39-189"><a href="#cb39-189"></a><span class="st">-xcutx-</span></span>
<span id="cb39-190"><a href="#cb39-190"></a></span>
<span id="cb39-191"><a href="#cb39-191"></a></span>
<span id="cb39-192"><a href="#cb39-192"></a><span class="st">2020-5-15</span></span>
<span id="cb39-193"><a href="#cb39-193"></a><span class="st">Daniel</span></span>
<span id="cb39-194"><a href="#cb39-194"></a><span class="st">If you're using git for version control, don't forget about the `.gitignore` file. Anything large and static (like .abfs) or procedurally generated (e.g. plots) you can toss in the `.gitignore` and you'll not see it when you commit.</span></span>
<span id="cb39-195"><a href="#cb39-195"></a></span>
<span id="cb39-196"><a href="#cb39-196"></a><span class="st">I have tabular data that lives in `./inst/extdata/` is processed by a script and then saved as a .rds in `./data/`.  Here's my `.gitignore`.</span></span>
<span id="cb39-197"><a href="#cb39-197"></a></span>
<span id="cb39-198"><a href="#cb39-198"></a></span>
<span id="cb39-199"><a href="#cb39-199"></a><span class="st">``` r</span></span>
<span id="cb39-200"><a href="#cb39-200"></a><span class="st">.Rproj.user</span></span>
<span id="cb39-201"><a href="#cb39-201"></a><span class="st">.Rhistory</span></span>
<span id="cb39-202"><a href="#cb39-202"></a><span class="st">.RData</span></span>
<span id="cb39-203"><a href="#cb39-203"></a><span class="st">.Ruserdata</span></span>
<span id="cb39-204"><a href="#cb39-204"></a><span class="st"># Don't track ABFs -- large and static</span></span>
<span id="cb39-205"><a href="#cb39-205"></a><span class="st">*.abf</span></span>
<span id="cb39-206"><a href="#cb39-206"></a><span class="st"># Don't track files that are generated from the scripts</span></span>
<span id="cb39-207"><a href="#cb39-207"></a><span class="st">/data/*</span></span>
<span id="cb39-208"><a href="#cb39-208"></a><span class="st">```</span></span>
<span id="cb39-209"><a href="#cb39-209"></a></span>
<span id="cb39-210"><a href="#cb39-210"></a><span class="st">For more check out &lt;https://git-scm.com/docs/gitignore&gt;</span></span>
<span id="cb39-211"><a href="#cb39-211"></a></span>
<span id="cb39-212"><a href="#cb39-212"></a></span>
<span id="cb39-213"><a href="#cb39-213"></a><span class="st">-xcutx-</span></span>
<span id="cb39-214"><a href="#cb39-214"></a></span>
<span id="cb39-215"><a href="#cb39-215"></a></span>
<span id="cb39-216"><a href="#cb39-216"></a><span class="st">2020-5-19</span></span>
<span id="cb39-217"><a href="#cb39-217"></a><span class="st">Daniel</span></span>
<span id="cb39-218"><a href="#cb39-218"></a><span class="st">If you need to include greek letters or special characters in a plot use `annotate()` instead of `geom_text()`. On my machine it preforms a *lot* faster.</span></span>
<span id="cb39-219"><a href="#cb39-219"></a></span>
<span id="cb39-220"><a href="#cb39-220"></a></span>
<span id="cb39-221"><a href="#cb39-221"></a><span class="st">``` r</span></span>
<span id="cb39-222"><a href="#cb39-222"></a><span class="st"># 57.62977 secs</span></span>
<span id="cb39-223"><a href="#cb39-223"></a><span class="st">geom_text(aes(x = 15, y = 25, label = "phi~22.5"), parse=TRUE)</span></span>
<span id="cb39-224"><a href="#cb39-224"></a><span class="st"># 1.988642 secs</span></span>
<span id="cb39-225"><a href="#cb39-225"></a><span class="st">annotate("text", x = 15, y = 25, parse = TRUE, label = as.character(expression(paste(phi, " 22.5"))))</span></span>
<span id="cb39-226"><a href="#cb39-226"></a><span class="st">```</span></span>
<span id="cb39-227"><a href="#cb39-227"></a></span>
<span id="cb39-228"><a href="#cb39-228"></a><span class="st">-xcutx-</span></span>
<span id="cb39-229"><a href="#cb39-229"></a></span>
<span id="cb39-230"><a href="#cb39-230"></a><span class="st">2020-5-19</span></span>
<span id="cb39-231"><a href="#cb39-231"></a><span class="st">Use `geom_rect()` with the min/max set to `-Inf` and `Inf` to add a pleasant shading to your facets.</span></span>
<span id="cb39-232"><a href="#cb39-232"></a></span>
<span id="cb39-233"><a href="#cb39-233"></a><span class="st">For example by passing it a data frame with the faceting variables and a column to use for the color (green if positive, red if negative) we can make facets behave like cells in  a heatmap!</span></span>
<span id="cb39-234"><a href="#cb39-234"></a></span>
<span id="cb39-235"><a href="#cb39-235"></a><span class="st">``` r</span></span>
<span id="cb39-236"><a href="#cb39-236"></a><span class="st"># &gt; tp2</span></span>
<span id="cb39-237"><a href="#cb39-237"></a><span class="st"># # A tibble: 23 x 5</span></span>
<span id="cb39-238"><a href="#cb39-238"></a><span class="st">#    Condition   Trace             Time  Change StimId</span></span>
<span id="cb39-239"><a href="#cb39-239"></a><span class="st">#    &lt;fct&gt;       &lt;chr&gt;             &lt;chr&gt;  &lt;dbl&gt; &lt;chr&gt; </span></span>
<span id="cb39-240"><a href="#cb39-240"></a><span class="st">#  1 PS.0.orig   170623b_0007.abf  60         1 a     </span></span>
<span id="cb39-241"><a href="#cb39-241"></a><span class="st">#  2 PS.45.orig  170825a_0008.abf  60        -1 c     </span></span>
<span id="cb39-242"><a href="#cb39-242"></a></span>
<span id="cb39-243"><a href="#cb39-243"></a><span class="st">ggplot(df)+</span></span>
<span id="cb39-244"><a href="#cb39-244"></a><span class="st">  geom_rect(data = tp2, aes(fill = Change),xmin = -Inf,xmax = Inf, ymin = -Inf,ymax = Inf,alpha = 0.3) +</span></span>
<span id="cb39-245"><a href="#cb39-245"></a><span class="st">  geom_hline(yintercept = 0, color = "cornflowerblue")+</span></span>
<span id="cb39-246"><a href="#cb39-246"></a><span class="st">  geom_pointline(aes(x = Time, y = rc, group = Experiment), shape = 1, color = "black")+</span></span>
<span id="cb39-247"><a href="#cb39-247"></a><span class="st">  ylim(-2, 3.5)+</span></span>
<span id="cb39-248"><a href="#cb39-248"></a><span class="st">  scale_fill_gradientn(colors = c("Red", "Grey", "Green"))+</span></span>
<span id="cb39-249"><a href="#cb39-249"></a><span class="st">  theme_base()+</span></span>
<span id="cb39-250"><a href="#cb39-250"></a><span class="st">  theme(legend.position = "")</span></span>
<span id="cb39-251"><a href="#cb39-251"></a><span class="st">```</span></span>
<span id="cb39-252"><a href="#cb39-252"></a></span>
<span id="cb39-253"><a href="#cb39-253"></a><span class="st">![image (5).png](image%20(5).png)</span></span>
<span id="cb39-254"><a href="#cb39-254"></a></span>
<span id="cb39-255"><a href="#cb39-255"></a></span>
<span id="cb39-256"><a href="#cb39-256"></a><span class="st">-xcutx-</span></span>
<span id="cb39-257"><a href="#cb39-257"></a></span>
<span id="cb39-258"><a href="#cb39-258"></a><span class="st">2020-5-26</span></span>
<span id="cb39-259"><a href="#cb39-259"></a><span class="st">Handy trick:</span></span>
<span id="cb39-260"><a href="#cb39-260"></a><span class="st">Write your test results into a summary csv. Then while writing you can insert a link to the cell value for a result into your .doc. After that, changing post hoc corrections, re-sampling iterations, or data QC just requires you to re-run your code and let word refresh all the links.</span></span>
<span id="cb39-261"><a href="#cb39-261"></a></span>
<span id="cb39-262"><a href="#cb39-262"></a><span class="st">You can also do this from excel, but you'll have to update the formula for the cell used instead.</span></span>
<span id="cb39-263"><a href="#cb39-263"></a></span>
<span id="cb39-264"><a href="#cb39-264"></a><span class="st">&lt;#C010AS4PSUX|data_analysis&gt;</span></span>
<span id="cb39-265"><a href="#cb39-265"></a><span class="st">![AutoUpdateStats.gif](AutoUpdateStats.gif)</span></span>
<span id="cb39-266"><a href="#cb39-266"></a></span>
<span id="cb39-267"><a href="#cb39-267"></a><span class="st">-xcutx-</span></span>
<span id="cb39-268"><a href="#cb39-268"></a></span>
<span id="cb39-269"><a href="#cb39-269"></a><span class="st">2020-6-4</span></span>
<span id="cb39-270"><a href="#cb39-270"></a><span class="st">If you're looking at linear relationships, give this code a spin. It'll automatically flag outliers based on 1.5*IQR and show you the fit with and without the outliers.  Since the output is a ggplot, you and add to it or tweak the aesthetics of the output (see example below). If anyone has a favorite outlier detection method and ends up modifying this, please post it here so we can see it! &lt;@U0103JSQYLB&gt; &lt;@U010CH4JLKE&gt; &lt;@U010M1M3N9E&gt;</span></span>
<span id="cb39-271"><a href="#cb39-271"></a></span>
<span id="cb39-272"><a href="#cb39-272"></a><span class="st">Here's the function.</span></span>
<span id="cb39-273"><a href="#cb39-273"></a></span>
<span id="cb39-274"><a href="#cb39-274"></a><span class="st">``` r</span></span>
<span id="cb39-275"><a href="#cb39-275"></a><span class="st">scatter_w_wo_outliers &lt;- function(temp = filter(M, Time == "Baseline"),</span></span>
<span id="cb39-276"><a href="#cb39-276"></a><span class="st">                                  X = "bkkca", </span></span>
<span id="cb39-277"><a href="#cb39-277"></a><span class="st">                                  Y = "Ihtk.0"){</span></span>
<span id="cb39-278"><a href="#cb39-278"></a><span class="st">  </span></span>
<span id="cb39-279"><a href="#cb39-279"></a><span class="st">  # flag outliers based on 1.5xIQR from median</span></span>
<span id="cb39-280"><a href="#cb39-280"></a><span class="st">  X_low  &lt;- median(temp[[X]], na.rm = T) - (1.5 * IQR(temp[[X]], na.rm = T))</span></span>
<span id="cb39-281"><a href="#cb39-281"></a><span class="st">  X_high &lt;- median(temp[[X]], na.rm = T) + (1.5 * IQR(temp[[X]], na.rm = T))</span></span>
<span id="cb39-282"><a href="#cb39-282"></a><span class="st">  Y_low  &lt;- median(temp[[Y]], na.rm = T) - (1.5 * IQR(temp[[Y]], na.rm = T))</span></span>
<span id="cb39-283"><a href="#cb39-283"></a><span class="st">  Y_high &lt;- median(temp[[Y]], na.rm = T) + (1.5 * IQR(temp[[Y]], na.rm = T))</span></span>
<span id="cb39-284"><a href="#cb39-284"></a><span class="st">  </span></span>
<span id="cb39-285"><a href="#cb39-285"></a><span class="st">  X_pass &lt;- (temp[[X]] &gt; X_low) * (temp[[X]] &lt; X_high)</span></span>
<span id="cb39-286"><a href="#cb39-286"></a><span class="st">  Y_pass &lt;- (temp[[Y]] &gt; Y_low) * (temp[[Y]] &lt; Y_high)</span></span>
<span id="cb39-287"><a href="#cb39-287"></a></span>
<span id="cb39-288"><a href="#cb39-288"></a><span class="st">  temp$flag &lt;- ifelse((X_pass * Y_pass) == 1, T, F)</span></span>
<span id="cb39-289"><a href="#cb39-289"></a><span class="st">  </span></span>
<span id="cb39-290"><a href="#cb39-290"></a><span class="st">  </span></span>
<span id="cb39-291"><a href="#cb39-291"></a><span class="st">  # Duplicate so we have dataset 1, 2 (introduces duplicates)</span></span>
<span id="cb39-292"><a href="#cb39-292"></a><span class="st">  temp &lt;- rbind(temp[temp$flag == T, ], mutate(temp, flag = F))</span></span>
<span id="cb39-293"><a href="#cb39-293"></a><span class="st">  </span></span>
<span id="cb39-294"><a href="#cb39-294"></a><span class="st">  formula1 &lt;- y ~ x</span></span>
<span id="cb39-295"><a href="#cb39-295"></a><span class="st">  </span></span>
<span id="cb39-296"><a href="#cb39-296"></a><span class="st">  plt &lt;- ggplot(temp, aes_string(X, Y, color = "flag"))+</span></span>
<span id="cb39-297"><a href="#cb39-297"></a><span class="st">    geom_smooth(data = temp, method = lm, se = F, fullrange = T)+</span></span>
<span id="cb39-298"><a href="#cb39-298"></a><span class="st">    geom_point(data = temp)+</span></span>
<span id="cb39-299"><a href="#cb39-299"></a><span class="st">    geom_point(data = temp, color = "black", shape = 1)+</span></span>
<span id="cb39-300"><a href="#cb39-300"></a><span class="st">    geom_point(data = temp[temp$flag, ])+</span></span>
<span id="cb39-301"><a href="#cb39-301"></a><span class="st">    ggpmisc::stat_poly_eq(aes(label =  paste(stat(eq.label), "*</span><span class="ch">\"</span><span class="st"> with </span><span class="ch">\"</span><span class="st">*", </span></span>
<span id="cb39-302"><a href="#cb39-302"></a><span class="st">                                             stat(rr.label), "*</span><span class="ch">\"</span><span class="st">, </span><span class="ch">\"</span><span class="st">*", </span></span>
<span id="cb39-303"><a href="#cb39-303"></a><span class="st">                                             stat(f.value.label), "*</span><span class="ch">\"</span><span class="st">, and </span><span class="ch">\"</span><span class="st">*",</span></span>
<span id="cb39-304"><a href="#cb39-304"></a><span class="st">                                             stat(p.value.label), "*</span><span class="ch">\"</span><span class="st">.</span><span class="ch">\"</span><span class="st">",</span></span>
<span id="cb39-305"><a href="#cb39-305"></a><span class="st">                                             sep = "")),</span></span>
<span id="cb39-306"><a href="#cb39-306"></a><span class="st">                          formula = formula1, parse = TRUE, size = 4)+</span></span>
<span id="cb39-307"><a href="#cb39-307"></a><span class="st">    </span></span>
<span id="cb39-308"><a href="#cb39-308"></a><span class="st">    scale_color_manual(values = c("darkgray", "black"))+</span></span>
<span id="cb39-309"><a href="#cb39-309"></a><span class="st">    theme_bw()+</span></span>
<span id="cb39-310"><a href="#cb39-310"></a><span class="st">    theme(legend.position = "")</span></span>
<span id="cb39-311"><a href="#cb39-311"></a><span class="st">  </span></span>
<span id="cb39-312"><a href="#cb39-312"></a><span class="st">  return(plt)</span></span>
<span id="cb39-313"><a href="#cb39-313"></a><span class="st">}</span></span>
<span id="cb39-314"><a href="#cb39-314"></a><span class="st">```</span></span>
<span id="cb39-315"><a href="#cb39-315"></a></span>
<span id="cb39-316"><a href="#cb39-316"></a><span class="st">Here's a reproducible example. We're creating a Simpson's paradox by giving the "outliers" a negative slope and the real data a positive slope. I've added a red line showing the true relationship.</span></span>
<span id="cb39-317"><a href="#cb39-317"></a></span>
<span id="cb39-318"><a href="#cb39-318"></a><span class="st">``` r</span></span>
<span id="cb39-319"><a href="#cb39-319"></a><span class="st">set.seed(45645684)</span></span>
<span id="cb39-320"><a href="#cb39-320"></a><span class="st">df &lt;- data.frame(x = rnorm(30, mean = 10, sd = 4),</span></span>
<span id="cb39-321"><a href="#cb39-321"></a><span class="st">                 noise = runif(30, min = -2, max = 2),</span></span>
<span id="cb39-322"><a href="#cb39-322"></a><span class="st">                 y = NA,</span></span>
<span id="cb39-323"><a href="#cb39-323"></a><span class="st">                 is_outlier = rbinom(30, 1, prob = 0.2))</span></span>
<span id="cb39-324"><a href="#cb39-324"></a></span>
<span id="cb39-325"><a href="#cb39-325"></a></span>
<span id="cb39-326"><a href="#cb39-326"></a><span class="st">df$y &lt;- ifelse(df$is_outlier, </span></span>
<span id="cb39-327"><a href="#cb39-327"></a><span class="st">               -5*df$x+df$noise,</span></span>
<span id="cb39-328"><a href="#cb39-328"></a><span class="st">               2*df$x+df$noise)</span></span>
<span id="cb39-329"><a href="#cb39-329"></a></span>
<span id="cb39-330"><a href="#cb39-330"></a><span class="st">scatter_w_wo_outliers(temp = df,</span></span>
<span id="cb39-331"><a href="#cb39-331"></a><span class="st">                      X = "x",</span></span>
<span id="cb39-332"><a href="#cb39-332"></a><span class="st">                      Y = "y")+</span></span>
<span id="cb39-333"><a href="#cb39-333"></a><span class="st">  geom_abline(slope = 2, intercept = 0, color = "firebrick")</span></span>
<span id="cb39-334"><a href="#cb39-334"></a><span class="st">```</span></span>
<span id="cb39-335"><a href="#cb39-335"></a></span>
<span id="cb39-336"><a href="#cb39-336"></a><span class="st">![image (7).png](image%20(7).png)</span></span>
<span id="cb39-337"><a href="#cb39-337"></a></span>
<span id="cb39-338"><a href="#cb39-338"></a></span>
<span id="cb39-339"><a href="#cb39-339"></a><span class="st">2020-6-4</span></span>
<span id="cb39-340"><a href="#cb39-340"></a><span class="st">Daniel</span></span>
<span id="cb39-341"><a href="#cb39-341"></a><span class="st">Here's a related utility function. For a given column it'll return a logical vector where outliers are `FALSE`.</span></span>
<span id="cb39-342"><a href="#cb39-342"></a></span>
<span id="cb39-343"><a href="#cb39-343"></a><span class="st">``` r</span></span>
<span id="cb39-344"><a href="#cb39-344"></a><span class="st">w_in_x_iqr &lt;- function(col_in, multiplier = 1.5){</span></span>
<span id="cb39-345"><a href="#cb39-345"></a><span class="st">  col_in &lt;- as.vector(col_in)</span></span>
<span id="cb39-346"><a href="#cb39-346"></a><span class="st">  </span></span>
<span id="cb39-347"><a href="#cb39-347"></a><span class="st">  X_low  &lt;- median(col_in, na.rm = T) - (multiplier * IQR(col_in, na.rm = T))</span></span>
<span id="cb39-348"><a href="#cb39-348"></a><span class="st">  X_high &lt;- median(col_in, na.rm = T) + (multiplier * IQR(col_in, na.rm = T))</span></span>
<span id="cb39-349"><a href="#cb39-349"></a><span class="st">  X_pass &lt;- (col_in &gt; X_low) * (col_in &lt; X_high)</span></span>
<span id="cb39-350"><a href="#cb39-350"></a><span class="st">  </span></span>
<span id="cb39-351"><a href="#cb39-351"></a><span class="st">  return(as.logical(X_pass))</span></span>
<span id="cb39-352"><a href="#cb39-352"></a><span class="st">}</span></span>
<span id="cb39-353"><a href="#cb39-353"></a><span class="st">```</span></span>
<span id="cb39-354"><a href="#cb39-354"></a></span>
<span id="cb39-355"><a href="#cb39-355"></a><span class="st">e.g.</span></span>
<span id="cb39-356"><a href="#cb39-356"></a></span>
<span id="cb39-357"><a href="#cb39-357"></a><span class="st">``` r</span></span>
<span id="cb39-358"><a href="#cb39-358"></a><span class="st">&gt; mutate(M, ex = w_in_x_iqr(bkkca)) %&gt;</span><span class="sc">% s</span><span class="st">elect(bkkca, ex) %&gt;% arrange(bkkca) %&gt;% tail()</span></span>
<span id="cb39-359"><a href="#cb39-359"></a></span>
<span id="cb39-360"><a href="#cb39-360"></a><span class="st">#  A tibble: 6 x 2</span></span>
<span id="cb39-361"><a href="#cb39-361"></a><span class="st">#   bkkca ex   </span></span>
<span id="cb39-362"><a href="#cb39-362"></a><span class="st">#   &lt;dbl&gt; &lt;lgl&gt;</span></span>
<span id="cb39-363"><a href="#cb39-363"></a><span class="st"># 1 3056. TRUE </span></span>
<span id="cb39-364"><a href="#cb39-364"></a><span class="st"># 2 3222. TRUE </span></span>
<span id="cb39-365"><a href="#cb39-365"></a><span class="st"># 3 3255. TRUE   # Within bounds</span></span>
<span id="cb39-366"><a href="#cb39-366"></a><span class="st"># 4 3552. FALSE  # Outside bounds </span></span>
<span id="cb39-367"><a href="#cb39-367"></a><span class="st"># 5 3817. FALSE</span></span>
<span id="cb39-368"><a href="#cb39-368"></a><span class="st"># 6 6740. FALSE</span></span>
<span id="cb39-369"><a href="#cb39-369"></a><span class="st">```</span></span>
<span id="cb39-370"><a href="#cb39-370"></a></span>
<span id="cb39-371"><a href="#cb39-371"></a></span>
<span id="cb39-372"><a href="#cb39-372"></a><span class="st">-xcutx-</span></span>
<span id="cb39-373"><a href="#cb39-373"></a></span>
<span id="cb39-374"><a href="#cb39-374"></a></span>
<span id="cb39-375"><a href="#cb39-375"></a><span class="st">2020-6-5</span></span>
<span id="cb39-376"><a href="#cb39-376"></a><span class="st">ggplot's, `element_text()` comes with a color argument. Why does that matter? It doesn't just accept atomics, you can hand it a _vector_! This effectively gives one access to conditional formatting. This works on `ggplot2_3.3.0` but "vectorized input to `element_text()` is not officially supported" so YMMV with newer versions.</span></span>
<span id="cb39-377"><a href="#cb39-377"></a></span>
<span id="cb39-378"><a href="#cb39-378"></a><span class="st">Here's an example I think it makes an otherwise unbearable figure a little more so without requiring duplicated labels.</span></span>
<span id="cb39-379"><a href="#cb39-379"></a></span>
<span id="cb39-380"><a href="#cb39-380"></a><span class="st">One thing to be aware of is the ordering of a character/ factor may differ between a data.frame and the plot. In the visualized example, I had `mRNA` as type character instead of factor so it got alphabetized when it was plotted rather than by the order appearing in the data.frame.</span></span>
<span id="cb39-381"><a href="#cb39-381"></a><span class="st">![image (8).png](image%20(8).png)</span></span>
<span id="cb39-382"><a href="#cb39-382"></a></span>
<span id="cb39-383"><a href="#cb39-383"></a></span>
<span id="cb39-384"><a href="#cb39-384"></a><span class="st">-xcutx-</span></span>
<span id="cb39-385"><a href="#cb39-385"></a></span>
<span id="cb39-386"><a href="#cb39-386"></a></span>
<span id="cb39-387"><a href="#cb39-387"></a><span class="st">2020-6-12</span></span>
<span id="cb39-388"><a href="#cb39-388"></a><span class="st">Daniel</span></span>
<span id="cb39-389"><a href="#cb39-389"></a><span class="st">Trying something new here-- "Function Friday". If you've written a function you're rather pleased with, used one in a creative way, or know one that you think deserves it's time in the sun please share it here.</span></span>
<span id="cb39-390"><a href="#cb39-390"></a></span>
<span id="cb39-391"><a href="#cb39-391"></a><span class="st">I'll start. Y'all should be using `unlist()`. `unlist` is a crazy handy function when you pair it with the `purrr` library. It'll take a list and try to give you a vector. Strictly speaking it's probably best to be using `map_dbl()` or `map_chr()` instead but let's not worry about that at the moment.</span></span>
<span id="cb39-392"><a href="#cb39-392"></a></span>
<span id="cb39-393"><a href="#cb39-393"></a><span class="st">Why do I love `unlist` so much? Because `unlist(map( ))` gives you a flexible, effective way to iterate (that is parallel friendly with minor changes).</span></span>
<span id="cb39-394"><a href="#cb39-394"></a></span>
<span id="cb39-395"><a href="#cb39-395"></a><span class="st">Here's an example: I have a bunch of traces in a folder and I need to know if there are any experiments that didn't copy. Additionally, I'd like to know what kind of data is in the file. I _could_ look through them one by one, but that's no fun. Thankfully, R has functions that perform similarly to shell functions. (More on these some other time.)</span></span>
<span id="cb39-396"><a href="#cb39-396"></a></span>
<span id="cb39-397"><a href="#cb39-397"></a><span class="st">So we start by defining a data.frame with all file names and their sizes. Note that calling `unlist(map())` _inside_ `data.frame()` lets us do a lot work very quickly. We find all the files, get information about each one, and then selectively return the size.</span></span>
<span id="cb39-398"><a href="#cb39-398"></a></span>
<span id="cb39-399"><a href="#cb39-399"></a><span class="st">``` r</span></span>
<span id="cb39-400"><a href="#cb39-400"></a><span class="st">traces_dir &lt;- "C:/Users/Daniel/Documents/Trace_Holding"</span></span>
<span id="cb39-401"><a href="#cb39-401"></a></span>
<span id="cb39-402"><a href="#cb39-402"></a><span class="st">files_df &lt;- data.frame(files = list.files(traces_dir),</span></span>
<span id="cb39-403"><a href="#cb39-403"></a><span class="st">                       bytes = unlist(map(list.files(traces_dir), </span></span>
<span id="cb39-404"><a href="#cb39-404"></a><span class="st">                                          function(abf){ &lt;http://file.info|file.info&gt;(paste0(traces_dir, "/",abf))$size })))</span></span>
<span id="cb39-405"><a href="#cb39-405"></a></span>
<span id="cb39-406"><a href="#cb39-406"></a><span class="st">#              files    bytes</span></span>
<span id="cb39-407"><a href="#cb39-407"></a><span class="st"># 1 190808a_0000.abf 15006208 &lt;- long gap free recording</span></span>
<span id="cb39-408"><a href="#cb39-408"></a><span class="st"># 2 190808a_0001.abf 15006208</span></span>
<span id="cb39-409"><a href="#cb39-409"></a><span class="st"># 3 190808a_0002.abf 15006208</span></span>
<span id="cb39-410"><a href="#cb39-410"></a><span class="st">```</span></span>
<span id="cb39-411"><a href="#cb39-411"></a></span>
<span id="cb39-412"><a href="#cb39-412"></a><span class="st">The experiment is embedded in the file name for each. Once again , with a little help from `unlist` we can split the file names into a list of list (i.e. `"190808a_0000.abf"` becomes `[[1]] [[1]] "190808a" [[2]] "0000.abf"` select only the first part and populate a new column.</span></span>
<span id="cb39-413"><a href="#cb39-413"></a></span>
<span id="cb39-414"><a href="#cb39-414"></a><span class="st">``` r</span></span>
<span id="cb39-415"><a href="#cb39-415"></a><span class="st">files_df$Experiment &lt;- unlist(transpose(str_split(files_df$files, pattern = "_"))[[1]])</span></span>
<span id="cb39-416"><a href="#cb39-416"></a><span class="st">```</span></span>
<span id="cb39-417"><a href="#cb39-417"></a></span>
<span id="cb39-418"><a href="#cb39-418"></a></span>
<span id="cb39-419"><a href="#cb39-419"></a><span class="st">Okay, now we can make use of this. I've defined a data.frame for metadata about the experiments.</span></span>
<span id="cb39-420"><a href="#cb39-420"></a></span>
<span id="cb39-421"><a href="#cb39-421"></a><span class="st">``` r</span></span>
<span id="cb39-422"><a href="#cb39-422"></a><span class="st">file_groups</span></span>
<span id="cb39-423"><a href="#cb39-423"></a><span class="st">#    Experiment       Group</span></span>
<span id="cb39-424"><a href="#cb39-424"></a><span class="st"># 1      190924    Baseline</span></span>
<span id="cb39-425"><a href="#cb39-425"></a><span class="st"># ... </span></span>
<span id="cb39-426"><a href="#cb39-426"></a><span class="st"># 19     190918     Delayed</span></span>
<span id="cb39-427"><a href="#cb39-427"></a><span class="st">```</span></span>
<span id="cb39-428"><a href="#cb39-428"></a></span>
<span id="cb39-429"><a href="#cb39-429"></a><span class="st">We can join these data frames and apply a little `tidyverse` magic to see what experiments are missing (we could also compare the sets of experiments directly).</span></span>
<span id="cb39-430"><a href="#cb39-430"></a></span>
<span id="cb39-431"><a href="#cb39-431"></a><span class="st">``` r</span></span>
<span id="cb39-432"><a href="#cb39-432"></a><span class="st">full_join(file_groups, files_df) %&gt;% </span></span>
<span id="cb39-433"><a href="#cb39-433"></a><span class="st">  filter(&lt;http://is.na|is.na&gt;(files)) %&gt;% </span></span>
<span id="cb39-434"><a href="#cb39-434"></a><span class="st">  group_by(Group, Experiment) %&gt;% </span></span>
<span id="cb39-435"><a href="#cb39-435"></a><span class="st">  tally()</span></span>
<span id="cb39-436"><a href="#cb39-436"></a></span>
<span id="cb39-437"><a href="#cb39-437"></a><span class="st">#  Experiments that didn't transfer:</span></span>
<span id="cb39-438"><a href="#cb39-438"></a></span>
<span id="cb39-439"><a href="#cb39-439"></a><span class="st">#   Group       Experiment     n</span></span>
<span id="cb39-440"><a href="#cb39-440"></a><span class="st"># 1 Baseline    190924         1</span></span>
<span id="cb39-441"><a href="#cb39-441"></a><span class="st"># 2 Baseline    190924a        1</span></span>
<span id="cb39-442"><a href="#cb39-442"></a><span class="st"># ...</span></span>
<span id="cb39-443"><a href="#cb39-443"></a><span class="st">```</span></span>
<span id="cb39-444"><a href="#cb39-444"></a></span>
<span id="cb39-445"><a href="#cb39-445"></a><span class="st">We can repeat the same strategy to programattically look at the protocol types (e.g. based on file size or channel number via `&lt;http://file.info|file.info&gt;` | `readABF()`). Moral of the story, you should so stop `apply`ing yourself and give `unlist` and `purrr` functions a try.</span></span>
<span id="cb39-446"><a href="#cb39-446"></a></span>
<span id="cb39-447"><a href="#cb39-447"></a></span>
<span id="cb39-448"><a href="#cb39-448"></a><span class="st">-xcutx-</span></span>
<span id="cb39-449"><a href="#cb39-449"></a></span>
<span id="cb39-450"><a href="#cb39-450"></a></span>
<span id="cb39-451"><a href="#cb39-451"></a><span class="st">2020-6-15</span></span>
<span id="cb39-452"><a href="#cb39-452"></a><span class="st">Daniel</span></span>
<span id="cb39-453"><a href="#cb39-453"></a><span class="st">If you run into a situation where you're using a ton of memory (e.g. manipulating transcriptomic data, resampling, or working with electrophysiology traces)  use `rm()` to selectively get rid off objects in the environment. A useful pattern is to write out large objects you'll need in the future, remove them, and then read them back in when you need them. This is usually not important, but when it is, it is.</span></span>
<span id="cb39-454"><a href="#cb39-454"></a></span>
<span id="cb39-455"><a href="#cb39-455"></a><span class="st">A side note -- unlike listing items where the function matching the unix command acts on the environment and a new command acts on the files system (`ls()` and `list.files()`) the functions for removing items don't follow this logic. `rm()` acts on objects in your environment whereas `unlink()` acts on system files.</span></span>
<span id="cb39-456"><a href="#cb39-456"></a></span>
<span id="cb39-457"><a href="#cb39-457"></a></span>
<span id="cb39-458"><a href="#cb39-458"></a><span class="st">-xcutx-</span></span>
<span id="cb39-459"><a href="#cb39-459"></a></span>
<span id="cb39-460"><a href="#cb39-460"></a></span>
<span id="cb39-461"><a href="#cb39-461"></a><span class="st">2020-6-25</span></span>
<span id="cb39-462"><a href="#cb39-462"></a><span class="st">Daniel</span></span>
<span id="cb39-463"><a href="#cb39-463"></a><span class="st">The `tictoc` library has convenience functions for timing code.  Here’s the basic usage relative to timing with base R.</span></span>
<span id="cb39-464"><a href="#cb39-464"></a></span>
<span id="cb39-465"><a href="#cb39-465"></a><span class="st">``` r</span></span>
<span id="cb39-466"><a href="#cb39-466"></a><span class="st">library(tictoc)</span></span>
<span id="cb39-467"><a href="#cb39-467"></a><span class="st">tic()</span></span>
<span id="cb39-468"><a href="#cb39-468"></a><span class="st"># code here</span></span>
<span id="cb39-469"><a href="#cb39-469"></a><span class="st">toc()</span></span>
<span id="cb39-470"><a href="#cb39-470"></a></span>
<span id="cb39-471"><a href="#cb39-471"></a><span class="st">tic &lt;- Sys.time()</span></span>
<span id="cb39-472"><a href="#cb39-472"></a><span class="st"># code here</span></span>
<span id="cb39-473"><a href="#cb39-473"></a><span class="st">toc &lt;- Sys.time()</span></span>
<span id="cb39-474"><a href="#cb39-474"></a><span class="st">print(toc - tic)</span></span>
<span id="cb39-475"><a href="#cb39-475"></a><span class="st">```</span></span>
<span id="cb39-476"><a href="#cb39-476"></a></span>
<span id="cb39-477"><a href="#cb39-477"></a><span class="st">Where this library excels is when you want to time multiple parts of your code. Each `tic` pushes the time onto a stack and each `toc` pops the most recent time from said stack. That means you don’t have to worry about assigning several timing variables even if you want to time nested code.</span></span>
<span id="cb39-478"><a href="#cb39-478"></a></span>
<span id="cb39-479"><a href="#cb39-479"></a><span class="st">``` r</span></span>
<span id="cb39-480"><a href="#cb39-480"></a><span class="st">tic()</span></span>
<span id="cb39-481"><a href="#cb39-481"></a><span class="st"># stack is 1 deep</span></span>
<span id="cb39-482"><a href="#cb39-482"></a><span class="st">for (i in 1:10) {</span></span>
<span id="cb39-483"><a href="#cb39-483"></a><span class="st">     tic()</span></span>
<span id="cb39-484"><a href="#cb39-484"></a><span class="st">     # stack is now 2 deep</span></span>
<span id="cb39-485"><a href="#cb39-485"></a><span class="st">     for (j in 1:10){</span></span>
<span id="cb39-486"><a href="#cb39-486"></a><span class="st">          tic()</span></span>
<span id="cb39-487"><a href="#cb39-487"></a><span class="st">          # stack is now 3 deep</span></span>
<span id="cb39-488"><a href="#cb39-488"></a><span class="st">          toc()</span></span>
<span id="cb39-489"><a href="#cb39-489"></a><span class="st">     }</span></span>
<span id="cb39-490"><a href="#cb39-490"></a><span class="st">     toc()</span></span>
<span id="cb39-491"><a href="#cb39-491"></a><span class="st">}</span></span>
<span id="cb39-492"><a href="#cb39-492"></a><span class="st">toc()</span></span>
<span id="cb39-493"><a href="#cb39-493"></a><span class="st">```</span></span>
<span id="cb39-494"><a href="#cb39-494"></a></span>
<span id="cb39-495"><a href="#cb39-495"></a></span>
<span id="cb39-496"><a href="#cb39-496"></a></span>
<span id="cb39-497"><a href="#cb39-497"></a><span class="st">-xcutx-</span></span>
<span id="cb39-498"><a href="#cb39-498"></a></span>
<span id="cb39-499"><a href="#cb39-499"></a></span>
<span id="cb39-500"><a href="#cb39-500"></a><span class="st">2020-7-28</span></span>
<span id="cb39-501"><a href="#cb39-501"></a><span class="st">In tidyverse you can use `.` to reference the current dataframe. This is really useful for plotting. For example, I'd like to plot channels In7 and In12 in that order but by default In12 will come first. We could save an intermediate dataframe and re-level the factors, but by referencing the dataframe piped into mutate we can skip this step.</span></span>
<span id="cb39-502"><a href="#cb39-502"></a></span>
<span id="cb39-503"><a href="#cb39-503"></a></span>
<span id="cb39-504"><a href="#cb39-504"></a><span class="st">``` r</span></span>
<span id="cb39-505"><a href="#cb39-505"></a><span class="st">temp %&gt;% # temp is a down-sampled ephys recording</span></span>
<span id="cb39-506"><a href="#cb39-506"></a><span class="st">  ungroup() %&gt;% </span></span>
<span id="cb39-507"><a href="#cb39-507"></a><span class="st">  gather(key, value, c("In7", "In12")) %&gt;% </span></span>
<span id="cb39-508"><a href="#cb39-508"></a><span class="st">  mutate(key = factor(.$key, levels = c("In7", "In12"))) %&gt;%      # Note that if you have groupings, you'll need to get rid of them or supply a column of the same length as the group. </span></span>
<span id="cb39-509"><a href="#cb39-509"></a><span class="st">  ggplot(aes(Time, value, color = key, group = interaction(key, Sweep)))+</span></span>
<span id="cb39-510"><a href="#cb39-510"></a><span class="st">  geom_path()</span></span>
<span id="cb39-511"><a href="#cb39-511"></a><span class="st">```</span></span>
<span id="cb39-512"><a href="#cb39-512"></a></span>
<span id="cb39-513"><a href="#cb39-513"></a></span>
<span id="cb39-514"><a href="#cb39-514"></a><span class="st">![image (11).png](image%20(11).png)</span></span>
<span id="cb39-515"><a href="#cb39-515"></a></span>
<span id="cb39-516"><a href="#cb39-516"></a></span>
<span id="cb39-517"><a href="#cb39-517"></a><span class="st">-xcutx-</span></span>
<span id="cb39-518"><a href="#cb39-518"></a></span>
<span id="cb39-519"><a href="#cb39-519"></a></span>
<span id="cb39-520"><a href="#cb39-520"></a><span class="st">2020-7-30</span></span>
<span id="cb39-521"><a href="#cb39-521"></a><span class="st">If you want to make future you more positively disposed to present you, then organize your work with `R` packages, save your custom functions in the `/R/` directory. The function `roxygen2::roxygenise()` will documentation comments for your functions into help pages. For example running `roxygenise()` with the below function saved produces the attached help page accessible via `?shrug`.</span></span>
<span id="cb39-522"><a href="#cb39-522"></a></span>
<span id="cb39-523"><a href="#cb39-523"></a></span>
<span id="cb39-524"><a href="#cb39-524"></a><span class="st">``` r</span></span>
<span id="cb39-525"><a href="#cb39-525"></a><span class="st">#' @title Print Shrug</span></span>
<span id="cb39-526"><a href="#cb39-526"></a><span class="st">#' @description This function prints a shrug emoji a specified number of times, provided the input value is a numeric greater than zero.</span></span>
<span id="cb39-527"><a href="#cb39-527"></a><span class="st">#' @param n how many shrugs should be printed</span></span>
<span id="cb39-528"><a href="#cb39-528"></a><span class="st">#' @author Daniel Kick (\email{daniel.r.kick@@gmail.com})</span></span>
<span id="cb39-529"><a href="#cb39-529"></a><span class="st">#' @export</span></span>
<span id="cb39-530"><a href="#cb39-530"></a><span class="st">#' @examples</span></span>
<span id="cb39-531"><a href="#cb39-531"></a><span class="st">#' shrug(5)</span></span>
<span id="cb39-532"><a href="#cb39-532"></a></span>
<span id="cb39-533"><a href="#cb39-533"></a><span class="st">shrug &lt;- function(n = 1, ...){</span></span>
<span id="cb39-534"><a href="#cb39-534"></a><span class="st">  if (is.numeric(n) &amp; n&gt;0){</span></span>
<span id="cb39-535"><a href="#cb39-535"></a><span class="st">    for (i in seq(1, n)){</span></span>
<span id="cb39-536"><a href="#cb39-536"></a><span class="st">        cat("¯</span><span class="ch">\\</span><span class="st">_(ツ)_/¯</span><span class="ch">\n</span><span class="st">")</span></span>
<span id="cb39-537"><a href="#cb39-537"></a><span class="st">    }</span></span>
<span id="cb39-538"><a href="#cb39-538"></a><span class="st">  }</span></span>
<span id="cb39-539"><a href="#cb39-539"></a><span class="st">}</span></span>
<span id="cb39-540"><a href="#cb39-540"></a><span class="st">```</span></span>
<span id="cb39-541"><a href="#cb39-541"></a></span>
<span id="cb39-542"><a href="#cb39-542"></a></span>
<span id="cb39-543"><a href="#cb39-543"></a><span class="st">![image (12).png](image%20(12).png)</span></span>
<span id="cb39-544"><a href="#cb39-544"></a></span>
<span id="cb39-545"><a href="#cb39-545"></a></span>
<span id="cb39-546"><a href="#cb39-546"></a><span class="st">-xcutx-</span></span>
<span id="cb39-547"><a href="#cb39-547"></a></span>
<span id="cb39-548"><a href="#cb39-548"></a><span class="st">2020-8-22</span></span>
<span id="cb39-549"><a href="#cb39-549"></a><span class="st">Be mindful of your data types. Sometimes `T == 1 == 1.0` (logical, int, double) but assuming these are equivalent can get you into trouble. For example, in these animations, the only difference is the data is coerced to logicals or integers.</span></span>
<span id="cb39-550"><a href="#cb39-550"></a><span class="st">![conway_bool.gif](conway_bool.gif)</span></span>
<span id="cb39-551"><a href="#cb39-551"></a><span class="st">![conway_int.gif](conway_int.gif)</span></span>
<span id="cb39-552"><a href="#cb39-552"></a></span>
<span id="cb39-553"><a href="#cb39-553"></a></span>
<span id="cb39-554"><a href="#cb39-554"></a><span class="st">-xcutx-</span></span>
<span id="cb39-555"><a href="#cb39-555"></a></span>
<span id="cb39-556"><a href="#cb39-556"></a></span>
<span id="cb39-557"><a href="#cb39-557"></a><span class="st">2020-9-19</span></span>
<span id="cb39-558"><a href="#cb39-558"></a><span class="st">Daniel</span></span>
<span id="cb39-559"><a href="#cb39-559"></a><span class="st">If you're working on data that takes a long time to process, consider adding a the following to your analysis.</span></span>
<span id="cb39-560"><a href="#cb39-560"></a></span>
<span id="cb39-561"><a href="#cb39-561"></a><span class="st">``` r</span></span>
<span id="cb39-562"><a href="#cb39-562"></a><span class="st">save.image(file='myEnvironment.RData')</span></span>
<span id="cb39-563"><a href="#cb39-563"></a><span class="st">load('myEnvironment.RData')</span></span>
<span id="cb39-564"><a href="#cb39-564"></a><span class="st">```</span></span>
<span id="cb39-565"><a href="#cb39-565"></a></span>
<span id="cb39-566"><a href="#cb39-566"></a><span class="st">This will let you reload your environment. One can also by default save your environment when closing Rstudio but that may make it easy to reference objects that no longer are generated in the document itself thereby speeding software rot.</span></span>
<span id="cb39-567"><a href="#cb39-567"></a></span>
<span id="cb39-568"><a href="#cb39-568"></a></span>
<span id="cb39-569"><a href="#cb39-569"></a><span class="st">-xcutx-</span></span>
<span id="cb39-570"><a href="#cb39-570"></a></span>
<span id="cb39-571"><a href="#cb39-571"></a><span class="st">2020-11-4</span></span>
<span id="cb39-572"><a href="#cb39-572"></a><span class="st">Daniel</span></span>
<span id="cb39-573"><a href="#cb39-573"></a><span class="st">&lt;@U010CH4JLKE&gt; As an example to show how accessible this is, here's bit of code that resamples an anova and computes an empirical p value.</span></span>
<span id="cb39-574"><a href="#cb39-574"></a></span>
<span id="cb39-575"><a href="#cb39-575"></a><span class="st">`temp` is a dataframe containing the data</span></span>
<span id="cb39-576"><a href="#cb39-576"></a><span class="st">`Condition` is a column with exactly that</span></span>
<span id="cb39-577"><a href="#cb39-577"></a><span class="st">`temp_col` is the name of a dependent variable. It's a string to make this easy to reuse.</span></span>
<span id="cb39-578"><a href="#cb39-578"></a><span class="st">if you haven't used `map` before it's basically a for loop that returns a list. When the output get's passed into `unlist` it becomes an array.</span></span>
<span id="cb39-579"><a href="#cb39-579"></a></span>
<span id="cb39-580"><a href="#cb39-580"></a></span>
<span id="cb39-581"><a href="#cb39-581"></a><span class="st">``` r</span></span>
<span id="cb39-582"><a href="#cb39-582"></a><span class="st">temp_shuffle &lt;- temp</span></span>
<span id="cb39-583"><a href="#cb39-583"></a><span class="st">resample_array &lt;- map(1:1000, function(i){</span></span>
<span id="cb39-584"><a href="#cb39-584"></a><span class="st">     temp_shuffle$Condition &lt;- sample(temp_shuffle$Condition, replace = F)</span></span>
<span id="cb39-585"><a href="#cb39-585"></a><span class="st">     fm &lt;- lm(as.formula(paste0(temp_col, " ~ Condition")), data = temp_shuffle)      </span></span>
<span id="cb39-586"><a href="#cb39-586"></a><span class="st">     return(car::Anova(fm)[1,3])</span></span>
<span id="cb39-587"><a href="#cb39-587"></a><span class="st">}) %&gt;</span><span class="sc">% u</span><span class="st">nlist()</span></span>
<span id="cb39-588"><a href="#cb39-588"></a><span class="st">ep &lt;- mean(resample_array &gt;= car::Anova(fm)[1,3])</span></span>
<span id="cb39-589"><a href="#cb39-589"></a><span class="st">```</span></span>
<span id="cb39-590"><a href="#cb39-590"></a></span>
<span id="cb39-591"><a href="#cb39-591"></a><span class="st">The down side is that it takes orders of magnitude more time to run because you're running the same code hundreds or thousands of times. This is only a problem if you need crazy high precision or have a really complex/hard to fit model. For reference using the code above took about ~2 seconds/dv for 1000 iterations on my machine.</span></span>
<span id="cb39-592"><a href="#cb39-592"></a></span>
<span id="cb39-593"><a href="#cb39-593"></a><span class="st">2020-11-4</span></span>
<span id="cb39-594"><a href="#cb39-594"></a><span class="st">Daniel</span></span>
<span id="cb39-595"><a href="#cb39-595"></a><span class="st">&lt;@U0103JSQYLB&gt;  A handy pattern is to use map to summarize data and then bind it.</span></span>
<span id="cb39-596"><a href="#cb39-596"></a></span>
<span id="cb39-597"><a href="#cb39-597"></a><span class="st">``` r</span></span>
<span id="cb39-598"><a href="#cb39-598"></a><span class="st">map_res &lt;- map(names(M)[names(M) != "Sample"], function(i){</span></span>
<span id="cb39-599"><a href="#cb39-599"></a><span class="st">  res &lt;- shapiro.test(M[[i]])</span></span>
<span id="cb39-600"><a href="#cb39-600"></a></span>
<span id="cb39-601"><a href="#cb39-601"></a><span class="st">  return(</span></span>
<span id="cb39-602"><a href="#cb39-602"></a><span class="st">    list(</span></span>
<span id="cb39-603"><a href="#cb39-603"></a><span class="st">    mrna = i,</span></span>
<span id="cb39-604"><a href="#cb39-604"></a><span class="st">    p = res$p.value</span></span>
<span id="cb39-605"><a href="#cb39-605"></a><span class="st">    )</span></span>
<span id="cb39-606"><a href="#cb39-606"></a><span class="st">  )</span></span>
<span id="cb39-607"><a href="#cb39-607"></a><span class="st">})</span></span>
<span id="cb39-608"><a href="#cb39-608"></a></span>
<span id="cb39-609"><a href="#cb39-609"></a><span class="st">shapiro_res &lt;- do.call(rbind, map_res)</span></span>
<span id="cb39-610"><a href="#cb39-610"></a><span class="st">```</span></span>
<span id="cb39-611"><a href="#cb39-611"></a><span class="st">-xcutx-</span></span>
<span id="cb39-612"><a href="#cb39-612"></a></span>
<span id="cb39-613"><a href="#cb39-613"></a></span>
<span id="cb39-614"><a href="#cb39-614"></a><span class="st">2020-11-20</span></span>
<span id="cb39-615"><a href="#cb39-615"></a><span class="st">Daniel</span></span>
<span id="cb39-616"><a href="#cb39-616"></a><span class="st">For all you classics fiends out there, R has a convenience function just for you! (As long as your numbers aren't too big)</span></span>
<span id="cb39-617"><a href="#cb39-617"></a></span>
<span id="cb39-618"><a href="#cb39-618"></a><span class="st">``` r</span></span>
<span id="cb39-619"><a href="#cb39-619"></a><span class="st">&gt; as.numeric(as.roman("MCXXIII"))</span></span>
<span id="cb39-620"><a href="#cb39-620"></a><span class="st">[1] 1123</span></span>
<span id="cb39-621"><a href="#cb39-621"></a><span class="st">&gt; as.roman(1123)</span></span>
<span id="cb39-622"><a href="#cb39-622"></a><span class="st">[1] MCXXIII</span></span>
<span id="cb39-623"><a href="#cb39-623"></a><span class="st">&gt; as.roman(11234)</span></span>
<span id="cb39-624"><a href="#cb39-624"></a><span class="st">[1] &lt;NA&gt;</span></span>
<span id="cb39-625"><a href="#cb39-625"></a><span class="st">```</span></span>
<span id="cb39-626"><a href="#cb39-626"></a></span>
<span id="cb39-627"><a href="#cb39-627"></a><span class="st">&lt;https://twitter.com/geokaramanis/status/1327286201443905538|source&gt;</span></span>
<span id="cb39-628"><a href="#cb39-628"></a></span>
<span id="cb39-629"><a href="#cb39-629"></a></span>
<span id="cb39-630"><a href="#cb39-630"></a><span class="st">-xcutx-</span></span>
<span id="cb39-631"><a href="#cb39-631"></a></span>
<span id="cb39-632"><a href="#cb39-632"></a></span>
<span id="cb39-633"><a href="#cb39-633"></a><span class="st">2020-12-7</span></span>
<span id="cb39-634"><a href="#cb39-634"></a><span class="st">Daniel</span></span>
<span id="cb39-635"><a href="#cb39-635"></a><span class="st">Get ready to celebrate in the streets: native pipes and terse lambdas are coming to R!</span></span>
<span id="cb39-636"><a href="#cb39-636"></a></span>
<span id="cb39-637"><a href="#cb39-637"></a><span class="st">&gt; *R* now provides a simple native pipe syntax `|&gt;` as well as a shorthand notation for creating functions, e.g. `\(x) x + 1` is parsed as `function(x) x + 1`. </span></span>
<span id="cb39-638"><a href="#cb39-638"></a><span class="st">&lt;https://cran.r-project.org/doc/manuals/r-devel/NEWS.html&gt;</span></span>
<span id="cb39-639"><a href="#cb39-639"></a></span>
<span id="cb39-640"><a href="#cb39-640"></a></span>
<span id="cb39-641"><a href="#cb39-641"></a><span class="st">2020-12-7</span></span>
<span id="cb39-642"><a href="#cb39-642"></a><span class="st">David Schulz</span></span>
<span id="cb39-643"><a href="#cb39-643"></a><span class="st">I am a little torn though.  I LOVE me a terse lambda.</span></span>
<span id="cb39-644"><a href="#cb39-644"></a></span>
<span id="cb39-645"><a href="#cb39-645"></a></span>
<span id="cb39-646"><a href="#cb39-646"></a><span class="st">-xcutx-</span></span>
<span id="cb39-647"><a href="#cb39-647"></a></span>
<span id="cb39-648"><a href="#cb39-648"></a><span class="st">2020-12-9</span></span>
<span id="cb39-649"><a href="#cb39-649"></a><span class="st">Check out `DiagrammeR` (or mermaid) if you need clean easy flow diagrams. I find they're not to hard to make and _even easier_ to forget that you've made them.</span></span>
<span id="cb39-650"><a href="#cb39-650"></a></span>
<span id="cb39-651"><a href="#cb39-651"></a></span>
<span id="cb39-652"><a href="#cb39-652"></a><span class="st">``` r</span></span>
<span id="cb39-653"><a href="#cb39-653"></a><span class="st">DiagrammeR::grViz("digraph {</span></span>
<span id="cb39-654"><a href="#cb39-654"></a><span class="st">graph [layout = dot]</span></span>
<span id="cb39-655"><a href="#cb39-655"></a></span>
<span id="cb39-656"><a href="#cb39-656"></a><span class="st"># define the global styles of the nodes. We can override these in box if we wish</span></span>
<span id="cb39-657"><a href="#cb39-657"></a><span class="st">node [shape = circle, style = filled, fillcolor = LightSteelBlue]</span></span>
<span id="cb39-658"><a href="#cb39-658"></a></span>
<span id="cb39-659"><a href="#cb39-659"></a><span class="st">data1 [label = 'Input', shape = folder, fillcolor = Beige]</span></span>
<span id="cb39-660"><a href="#cb39-660"></a></span>
<span id="cb39-661"><a href="#cb39-661"></a><span class="st">glmnet [label = 'Lasso </span><span class="ch">\n</span><span class="st">Regression', shape = box, fillcolor = Linen]</span></span>
<span id="cb39-662"><a href="#cb39-662"></a><span class="st">mnnet [label = 'Mulitnomial </span><span class="ch">\n</span><span class="st">Neural </span><span class="ch">\n</span><span class="st">Network', shape = box, fillcolor = Linen]</span></span>
<span id="cb39-663"><a href="#cb39-663"></a><span class="st">nnet [label = 'Neural </span><span class="ch">\n</span><span class="st">Network', shape = box, fillcolor = Linen]</span></span>
<span id="cb39-664"><a href="#cb39-664"></a><span class="st">knn [label = 'k-Nearest </span><span class="ch">\n</span><span class="st">Neighbor', shape = box, fillcolor = Linen]</span></span>
<span id="cb39-665"><a href="#cb39-665"></a><span class="st">ranger [label = 'Random </span><span class="ch">\n</span><span class="st">Forest', shape = box, fillcolor = Linen]</span></span>
<span id="cb39-666"><a href="#cb39-666"></a><span class="st">svml [label = 'SVM </span><span class="ch">\n</span><span class="st">Linear', shape = box, fillcolor = Linen]</span></span>
<span id="cb39-667"><a href="#cb39-667"></a><span class="st">svmr [label = 'SVM </span><span class="ch">\n</span><span class="st">Radial', shape = box, fillcolor = Linen]</span></span>
<span id="cb39-668"><a href="#cb39-668"></a></span>
<span id="cb39-669"><a href="#cb39-669"></a><span class="st">data2 [label = '5-fold CV </span><span class="ch">\n</span><span class="st">Accuracy', shape = folder, fillcolor = Beige]</span></span>
<span id="cb39-670"><a href="#cb39-670"></a></span>
<span id="cb39-671"><a href="#cb39-671"></a><span class="st"># edge definitions with the node IDs</span></span>
<span id="cb39-672"><a href="#cb39-672"></a><span class="st">data1 -&gt; {glmnet mnnet nnet knn ranger svml svmr} -&gt; data2</span></span>
<span id="cb39-673"><a href="#cb39-673"></a><span class="st">{alpha lambda} -&gt; glmnet</span></span>
<span id="cb39-674"><a href="#cb39-674"></a><span class="st">decay -&gt; mnnet</span></span>
<span id="cb39-675"><a href="#cb39-675"></a><span class="st">{size decay} -&gt; nnet</span></span>
<span id="cb39-676"><a href="#cb39-676"></a><span class="st">k -&gt; knn</span></span>
<span id="cb39-677"><a href="#cb39-677"></a><span class="st">{mtry splitrule minNodeSize} -&gt; ranger</span></span>
<span id="cb39-678"><a href="#cb39-678"></a><span class="st">cost -&gt; svml</span></span>
<span id="cb39-679"><a href="#cb39-679"></a><span class="st">sigma -&gt; svmr</span></span>
<span id="cb39-680"><a href="#cb39-680"></a><span class="st">}")</span></span>
<span id="cb39-681"><a href="#cb39-681"></a><span class="st">```</span></span>
<span id="cb39-682"><a href="#cb39-682"></a></span>
<span id="cb39-683"><a href="#cb39-683"></a><span class="st">![image (15).png](image%20(15).png)</span></span>
<span id="cb39-684"><a href="#cb39-684"></a></span>
<span id="cb39-685"><a href="#cb39-685"></a><span class="st">-xcutx-</span></span>
<span id="cb39-686"><a href="#cb39-686"></a></span>
<span id="cb39-687"><a href="#cb39-687"></a><span class="st">2020-12-18</span></span>
<span id="cb39-688"><a href="#cb39-688"></a><span class="st">Daniel</span></span>
<span id="cb39-689"><a href="#cb39-689"></a><span class="st">I just ran into the same issue &lt;@U010N31RYMV&gt; described after updating R on OSX —  none of the libraries are associated with the new version. I tested and abandoned directly copying the libraries over in favor programmatic reinstallation. The downside to this is that it’s slower. The upside is that if you just copy them, R will ask you to update the packages anyway.</span></span>
<span id="cb39-690"><a href="#cb39-690"></a></span>
<span id="cb39-691"><a href="#cb39-691"></a><span class="st">This won’t work for libraries that aren’t on CRAN but drastically reduces the number of libraries you’re installing by hand. In my case this took care of all but about 3</span><span class="sc">% o</span><span class="st">f the libraries I had installed for 3.6.</span></span>
<span id="cb39-692"><a href="#cb39-692"></a></span>
<span id="cb39-693"><a href="#cb39-693"></a></span>
<span id="cb39-694"><a href="#cb39-694"></a><span class="st">``` r</span></span>
<span id="cb39-695"><a href="#cb39-695"></a><span class="st">all_packages &lt;- list.files("/Library/Frameworks/R.framework/Versions/3.6/Resources/library/")</span></span>
<span id="cb39-696"><a href="#cb39-696"></a><span class="st">installed_packages &lt;- list.files("/Library/Frameworks/R.framework/Versions/4.0/Resources/library/")</span></span>
<span id="cb39-697"><a href="#cb39-697"></a><span class="st">all_packages &lt;- all_packages[!(all_packages </span><span class="sc">%i</span><span class="st">n</span><span class="sc">% i</span><span class="st">nstalled_packages)]</span></span>
<span id="cb39-698"><a href="#cb39-698"></a></span>
<span id="cb39-699"><a href="#cb39-699"></a><span class="st">options(install.packages.compile.from.source = "always")</span></span>
<span id="cb39-700"><a href="#cb39-700"></a></span>
<span id="cb39-701"><a href="#cb39-701"></a><span class="st">for (package in all_packages){</span></span>
<span id="cb39-702"><a href="#cb39-702"></a><span class="st">  try(install.packages(package))</span></span>
<span id="cb39-703"><a href="#cb39-703"></a><span class="st">}</span></span>
<span id="cb39-704"><a href="#cb39-704"></a></span>
<span id="cb39-705"><a href="#cb39-705"></a><span class="st">options(install.packages.compile.from.source = "interactive")</span></span>
<span id="cb39-706"><a href="#cb39-706"></a><span class="st">```</span></span>
<span id="cb39-707"><a href="#cb39-707"></a></span>
<span id="cb39-708"><a href="#cb39-708"></a><span class="st">-xcutx-</span></span>
<span id="cb39-709"><a href="#cb39-709"></a></span>
<span id="cb39-710"><a href="#cb39-710"></a></span>
<span id="cb39-711"><a href="#cb39-711"></a><span class="st">2020-12-22</span></span>
<span id="cb39-712"><a href="#cb39-712"></a><span class="st">You can set ggplot's font using the `theme` function. Particularly if combined with functions from `ggthemes` or `ggsci` you can get very pleasing visualizations quickly. Beyond accessing fonts already on your system you can import and fonts with minimal hassle.</span></span>
<span id="cb39-713"><a href="#cb39-713"></a></span>
<span id="cb39-714"><a href="#cb39-714"></a><span class="st">e.g. to get the font &lt;https://fonts.google.com/specimen/Metal+Mania?sidebar.open=true&amp;amp;selection.family=JetBrains+Mono:wght@100&amp;amp;query=metal|Metal Mania&gt; :metal: ready to use one might run:</span></span>
<span id="cb39-715"><a href="#cb39-715"></a></span>
<span id="cb39-716"><a href="#cb39-716"></a><span class="st">``` r</span></span>
<span id="cb39-717"><a href="#cb39-717"></a><span class="st">library(showtext)</span></span>
<span id="cb39-718"><a href="#cb39-718"></a><span class="st">font_add_google(name = "Metal Mania", family = "Metal+Mania")</span></span>
<span id="cb39-719"><a href="#cb39-719"></a><span class="st">```</span></span>
<span id="cb39-720"><a href="#cb39-720"></a></span>
<span id="cb39-721"><a href="#cb39-721"></a></span>
<span id="cb39-722"><a href="#cb39-722"></a><span class="st">``` r</span></span>
<span id="cb39-723"><a href="#cb39-723"></a><span class="st">library(palmerpenguins)</span></span>
<span id="cb39-724"><a href="#cb39-724"></a><span class="st">library(tidyverse)</span></span>
<span id="cb39-725"><a href="#cb39-725"></a><span class="st">library(ggthemes)</span></span>
<span id="cb39-726"><a href="#cb39-726"></a><span class="st">library(patchwork) # for adding together plots at the end</span></span>
<span id="cb39-727"><a href="#cb39-727"></a><span class="st">library(extrafont)</span></span>
<span id="cb39-728"><a href="#cb39-728"></a><span class="st"># font_import()) # &lt;--- run this once</span></span>
<span id="cb39-729"><a href="#cb39-729"></a><span class="st">loadfonts(device = "win", quiet = TRUE) # &lt;--- run this once per session</span></span>
<span id="cb39-730"><a href="#cb39-730"></a><span class="st"># extrafont::fonts() # see fonts that are available</span></span>
<span id="cb39-731"><a href="#cb39-731"></a></span>
<span id="cb39-732"><a href="#cb39-732"></a><span class="st">plt1 &lt;- palmerpenguins::penguins %&gt;%</span></span>
<span id="cb39-733"><a href="#cb39-733"></a><span class="st">  filter(!&lt;http://is.na|is.na&gt;(sex)) %&gt;%</span></span>
<span id="cb39-734"><a href="#cb39-734"></a><span class="st">  mutate(sex = case_when(sex == "male" ~ "m",</span></span>
<span id="cb39-735"><a href="#cb39-735"></a><span class="st">                         sex == "female" ~ "f")) %&gt;% </span></span>
<span id="cb39-736"><a href="#cb39-736"></a><span class="st">  ggplot(aes(sex, body_mass_g, fill = species, group = interaction(species, sex)))+</span></span>
<span id="cb39-737"><a href="#cb39-737"></a><span class="st">  geom_boxplot()+</span></span>
<span id="cb39-738"><a href="#cb39-738"></a><span class="st">  ggthemes::scale_fill_colorblind()+</span></span>
<span id="cb39-739"><a href="#cb39-739"></a><span class="st">  ggthemes::theme_clean()+</span></span>
<span id="cb39-740"><a href="#cb39-740"></a><span class="st">  theme(legend.position = "")+</span></span>
<span id="cb39-741"><a href="#cb39-741"></a><span class="st">  facet_grid(.~species)+</span></span>
<span id="cb39-742"><a href="#cb39-742"></a><span class="st">  labs(title = "Default")</span></span>
<span id="cb39-743"><a href="#cb39-743"></a></span>
<span id="cb39-744"><a href="#cb39-744"></a><span class="st">plt2 &lt;- plt1+</span></span>
<span id="cb39-745"><a href="#cb39-745"></a><span class="st">  theme(text = element_text(family = "Consolas"))+</span></span>
<span id="cb39-746"><a href="#cb39-746"></a><span class="st">  labs(title = "Consolas")</span></span>
<span id="cb39-747"><a href="#cb39-747"></a></span>
<span id="cb39-748"><a href="#cb39-748"></a><span class="st">plt3 &lt;- plt1+</span></span>
<span id="cb39-749"><a href="#cb39-749"></a><span class="st">  theme(text = element_text(family = "Garamond"))+</span></span>
<span id="cb39-750"><a href="#cb39-750"></a><span class="st">  labs(title = "Garamond")</span></span>
<span id="cb39-751"><a href="#cb39-751"></a></span>
<span id="cb39-752"><a href="#cb39-752"></a><span class="st">plt1 + plt2 + plt3</span></span>
<span id="cb39-753"><a href="#cb39-753"></a><span class="st">```</span></span>
<span id="cb39-754"><a href="#cb39-754"></a></span>
<span id="cb39-755"><a href="#cb39-755"></a></span>
<span id="cb39-756"><a href="#cb39-756"></a><span class="st">![image (17).png](image%20(17).png)</span></span>
<span id="cb39-757"><a href="#cb39-757"></a></span>
<span id="cb39-758"><a href="#cb39-758"></a></span>
<span id="cb39-759"><a href="#cb39-759"></a><span class="st">-xcutx-</span></span>
<span id="cb39-760"><a href="#cb39-760"></a></span>
<span id="cb39-761"><a href="#cb39-761"></a></span>
<span id="cb39-762"><a href="#cb39-762"></a><span class="st">2021-2-2</span></span>
<span id="cb39-763"><a href="#cb39-763"></a><span class="st">Anyone run into issues with `corrplot::corrplot()`  cutting off the title?</span></span>
<span id="cb39-764"><a href="#cb39-764"></a><span class="st">![image (18).png](image%20(18).png)</span></span>
<span id="cb39-765"><a href="#cb39-765"></a></span>
<span id="cb39-766"><a href="#cb39-766"></a></span>
<span id="cb39-767"><a href="#cb39-767"></a></span>
<span id="cb39-768"><a href="#cb39-768"></a><span class="st">2021-2-2</span></span>
<span id="cb39-769"><a href="#cb39-769"></a><span class="st">`ggcorrplot` makes okay plots with ggplot2's logic. Not as clean as the above but it'll work with `patchwork`  and `cowplot` . Unfortunately, `scale_colour_stepsn`  doesn't override the  scaling.</span></span>
<span id="cb39-770"><a href="#cb39-770"></a></span>
<span id="cb39-771"><a href="#cb39-771"></a><span class="st">``` r</span></span>
<span id="cb39-772"><a href="#cb39-772"></a><span class="st">library(ggcorrplot)</span></span>
<span id="cb39-773"><a href="#cb39-773"></a></span>
<span id="cb39-774"><a href="#cb39-774"></a><span class="st">p.mat &lt;- cor_pmat(cor_df)</span></span>
<span id="cb39-775"><a href="#cb39-775"></a><span class="st">ggcorrplot(cor(cor_df, use = "pairwise.complete.obs"), </span></span>
<span id="cb39-776"><a href="#cb39-776"></a><span class="st">           p.mat = p.mat,</span></span>
<span id="cb39-777"><a href="#cb39-777"></a><span class="st">           insig = "blank",</span></span>
<span id="cb39-778"><a href="#cb39-778"></a><span class="st">           type = "upper",</span></span>
<span id="cb39-779"><a href="#cb39-779"></a><span class="st">           outline.col = "white",</span></span>
<span id="cb39-780"><a href="#cb39-780"></a><span class="st">           colors = RColorBrewer::brewer.pal(n = 9, name = "PuOr")[c(1,5,9)]</span></span>
<span id="cb39-781"><a href="#cb39-781"></a><span class="st">           )+</span></span>
<span id="cb39-782"><a href="#cb39-782"></a><span class="st">  labs(title = "Brian_AP_Delayed")</span></span>
<span id="cb39-783"><a href="#cb39-783"></a><span class="st">```</span></span>
<span id="cb39-784"><a href="#cb39-784"></a></span>
<span id="cb39-785"><a href="#cb39-785"></a></span>
<span id="cb39-786"><a href="#cb39-786"></a><span class="st">![image (19).png](image%20(19).png)</span></span>
<span id="cb39-787"><a href="#cb39-787"></a></span>
<span id="cb39-788"><a href="#cb39-788"></a><span class="st">2021-2-2</span></span>
<span id="cb39-789"><a href="#cb39-789"></a><span class="st">`ggcorrplot` appears to call internal functions which makes modifying it quickly impractical (one would probably be best forking the package and modifying that). I think I have a workaround that gets the same binning behavior:</span></span>
<span id="cb39-790"><a href="#cb39-790"></a></span>
<span id="cb39-791"><a href="#cb39-791"></a><span class="st">After the significance matrix (`p.mat`) is generated overwrite the correlation matrix with the middle value of each desired bin.</span></span>
<span id="cb39-792"><a href="#cb39-792"></a></span>
<span id="cb39-793"><a href="#cb39-793"></a><span class="st">``` r</span></span>
<span id="cb39-794"><a href="#cb39-794"></a><span class="st">          bkkca      cav1      cav2</span></span>
<span id="cb39-795"><a href="#cb39-795"></a><span class="st">bkkca 1.0000000 0.3452702 0.5603564</span></span>
<span id="cb39-796"><a href="#cb39-796"></a><span class="st">cav1  0.3452702 1.0000000 0.7880727</span></span>
<span id="cb39-797"><a href="#cb39-797"></a><span class="st">cav2  0.5603564 0.7880727 1.0000000</span></span>
<span id="cb39-798"><a href="#cb39-798"></a><span class="st">&gt; # bin the correlations so there are fewer colors used in the figure</span></span>
<span id="cb39-799"><a href="#cb39-799"></a><span class="st">&gt; cor_bins &lt;- seq(-1, 1, length.out = 9)</span></span>
<span id="cb39-800"><a href="#cb39-800"></a><span class="st">&gt; for (i in 1:(length(cor_bins)-1)){</span></span>
<span id="cb39-801"><a href="#cb39-801"></a><span class="st">+   test[test &gt; cor_bins[i] &amp;amp; test &lt; cor_bins[i+1]] &lt;- ((cor_bins[i] + cor_bins[i+1])/2)</span></span>
<span id="cb39-802"><a href="#cb39-802"></a><span class="st">+ }</span></span>
<span id="cb39-803"><a href="#cb39-803"></a><span class="st">&gt; test</span></span>
<span id="cb39-804"><a href="#cb39-804"></a><span class="st">      bkkca  cav1  cav2</span></span>
<span id="cb39-805"><a href="#cb39-805"></a><span class="st">bkkca 1.000 0.375 0.625</span></span>
<span id="cb39-806"><a href="#cb39-806"></a><span class="st">cav1  0.375 1.000 0.875</span></span>
<span id="cb39-807"><a href="#cb39-807"></a><span class="st">cav2  0.625 0.875 1.000</span></span>
<span id="cb39-808"><a href="#cb39-808"></a><span class="st">```</span></span>
<span id="cb39-809"><a href="#cb39-809"></a></span>
<span id="cb39-810"><a href="#cb39-810"></a><span class="st">Here this makes very slight changes to the plot. (legend dropped to not imply a continuous fill)</span></span>
<span id="cb39-811"><a href="#cb39-811"></a><span class="st">![image (20).png](image%20(20).png)</span></span>
<span id="cb39-812"><a href="#cb39-812"></a></span>
<span id="cb39-813"><a href="#cb39-813"></a><span class="st">2021-2-2</span></span>
<span id="cb39-814"><a href="#cb39-814"></a><span class="st">Last update, this is harder to read up will use the more extreme value to get closer to `corrplot`</span></span>
<span id="cb39-815"><a href="#cb39-815"></a></span>
<span id="cb39-816"><a href="#cb39-816"></a></span>
<span id="cb39-817"><a href="#cb39-817"></a><span class="st">``` r</span></span>
<span id="cb39-818"><a href="#cb39-818"></a><span class="st">test &lt;- seq(-1, 1, length.out = 5)+.0000001</span></span>
<span id="cb39-819"><a href="#cb39-819"></a><span class="st">test</span></span>
<span id="cb39-820"><a href="#cb39-820"></a><span class="st">#-0.9999999 -0.4999999  0.0000001  0.5000001  1.0000001</span></span>
<span id="cb39-821"><a href="#cb39-821"></a><span class="st">for (i in 1:(length(cor_bins)-1)){</span></span>
<span id="cb39-822"><a href="#cb39-822"></a><span class="st">  # test[test &gt; cor_bins[i] &amp;amp; test &lt; cor_bins[i+1]] &lt;- (cor_bins[i] + cor_bins[i+1])</span></span>
<span id="cb39-823"><a href="#cb39-823"></a><span class="st">  test[test &gt; cor_bins[i] &amp;amp; test &lt; cor_bins[i+1]] &lt;-   cor_bins[c(i, (i+1))[which.max(abs(cor_bins[i:(i+1)]))]]</span></span>
<span id="cb39-824"><a href="#cb39-824"></a><span class="st">}</span></span>
<span id="cb39-825"><a href="#cb39-825"></a><span class="st">test</span></span>
<span id="cb39-826"><a href="#cb39-826"></a><span class="st">#-1.00 -0.50  0.25  0.75  1.00</span></span>
<span id="cb39-827"><a href="#cb39-827"></a><span class="st">```</span></span>
<span id="cb39-828"><a href="#cb39-828"></a></span>
<span id="cb39-829"><a href="#cb39-829"></a></span>
<span id="cb39-830"><a href="#cb39-830"></a><span class="st">![image (21).png](image%20(21).png)</span></span>
<span id="cb39-831"><a href="#cb39-831"></a></span>
<span id="cb39-832"><a href="#cb39-832"></a></span>
<span id="cb39-833"><a href="#cb39-833"></a><span class="st">-xcutx-</span></span>
<span id="cb39-834"><a href="#cb39-834"></a></span>
<span id="cb39-835"><a href="#cb39-835"></a></span>
<span id="cb39-836"><a href="#cb39-836"></a><span class="st">2021-2-3</span></span>
<span id="cb39-837"><a href="#cb39-837"></a><span class="st">Daniel</span></span>
<span id="cb39-838"><a href="#cb39-838"></a><span class="st">&lt;@U0103JSQYLB&gt; and I just talked through how one might compare two sets of p values. We're thinking about the following approach and wanted to see if anyone can see a flaw in this approach or has other ideas.</span></span>
<span id="cb39-839"><a href="#cb39-839"></a></span>
<span id="cb39-840"><a href="#cb39-840"></a><span class="st">Setup:</span></span>
<span id="cb39-841"><a href="#cb39-841"></a><span class="st">You have two replicates of an experiment (`r1, r2`). In each experiment you measured three mRNAs (_`a, b, c`_) in control and treatment (`c, t`). You want to know if the same trends in `r1` show up in `r2` but there is a batch effect that will prevent comparing them directly (e.g. you can't run a t test on `r1 _a_ c` and `r2 _a_ c` )</span></span>
<span id="cb39-842"><a href="#cb39-842"></a></span>
<span id="cb39-843"><a href="#cb39-843"></a><span class="st">Proposed Solution:</span></span>
<span id="cb39-844"><a href="#cb39-844"></a><span class="st">1. Compare each mRNA within a replicate and note the sign of change and if the p value reached a pre-determined cutoff</span></span>
<span id="cb39-845"><a href="#cb39-845"></a></span>
<span id="cb39-846"><a href="#cb39-846"></a><span class="st">``` r</span></span>
<span id="cb39-847"><a href="#cb39-847"></a><span class="st">rep  |mrna |sign |pval |sig  |</span></span>
<span id="cb39-848"><a href="#cb39-848"></a><span class="st">r1   | a   | -   |0.04 |1    |</span></span>
<span id="cb39-849"><a href="#cb39-849"></a><span class="st">r1   | b   | +   |0.10 |0    |</span></span>
<span id="cb39-850"><a href="#cb39-850"></a><span class="st">...  |     |     |     |     |</span></span>
<span id="cb39-851"><a href="#cb39-851"></a><span class="st">r2   | c   | -   |0.02 |1    |</span></span>
<span id="cb39-852"><a href="#cb39-852"></a><span class="st">```</span></span>
<span id="cb39-853"><a href="#cb39-853"></a></span>
<span id="cb39-854"><a href="#cb39-854"></a><span class="st">2. Multiply the sign by the significance code so that -1 = "significant decrease", 0 = "no significant change", +1 = "significant increase"</span></span>
<span id="cb39-855"><a href="#cb39-855"></a></span>
<span id="cb39-856"><a href="#cb39-856"></a><span class="st">``` r</span></span>
<span id="cb39-857"><a href="#cb39-857"></a><span class="st">rep  |mrna |sign |pval |sig  |sxp  |</span></span>
<span id="cb39-858"><a href="#cb39-858"></a><span class="st">r1   | a   | -   |0.04 |1    |+1   |</span></span>
<span id="cb39-859"><a href="#cb39-859"></a><span class="st">r1   | b   | +   |0.10 |0    |0    |</span></span>
<span id="cb39-860"><a href="#cb39-860"></a><span class="st">...  |     |     |     |     |     |</span></span>
<span id="cb39-861"><a href="#cb39-861"></a><span class="st">r2   | c   | -   |0.02 |0    |-1   |</span></span>
<span id="cb39-862"><a href="#cb39-862"></a><span class="st">```</span></span>
<span id="cb39-863"><a href="#cb39-863"></a></span>
<span id="cb39-864"><a href="#cb39-864"></a><span class="st">3. Reshape these as two vectors and treat them as categorical data. Then compare the "assignment" between these two lists using a jaccard index as if we were comparing an assignment from clustering against reality.</span></span>
<span id="cb39-865"><a href="#cb39-865"></a></span>
<span id="cb39-866"><a href="#cb39-866"></a><span class="st">``` r</span></span>
<span id="cb39-867"><a href="#cb39-867"></a><span class="st">r1 &lt;- as.character(c(1, 0, 0))</span></span>
<span id="cb39-868"><a href="#cb39-868"></a><span class="st">r2 &lt;- as.character(c(0, 0, -1))</span></span>
<span id="cb39-869"><a href="#cb39-869"></a><span class="st">jaccard(r1, r2)</span></span>
<span id="cb39-870"><a href="#cb39-870"></a><span class="st">```</span></span>
<span id="cb39-871"><a href="#cb39-871"></a></span>
<span id="cb39-872"><a href="#cb39-872"></a><span class="st">4. Use resampling to find an empirical p value for this observed jaccard index.</span></span>
<span id="cb39-873"><a href="#cb39-873"></a></span>
<span id="cb39-874"><a href="#cb39-874"></a><span class="st">Does that seem reasonable? Is there another way you would go about it?</span></span>
<span id="cb39-875"><a href="#cb39-875"></a></span>
<span id="cb39-876"><a href="#cb39-876"></a></span>
<span id="cb39-877"><a href="#cb39-877"></a><span class="st">-xcutx-</span></span>
<span id="cb39-878"><a href="#cb39-878"></a></span>
<span id="cb39-879"><a href="#cb39-879"></a></span>
<span id="cb39-880"><a href="#cb39-880"></a><span class="st">2021-2-10</span></span>
<span id="cb39-881"><a href="#cb39-881"></a><span class="st">Here's an implementation for two sets of correlations. Here we bin the correlations into 5 bins use a jaccard index to assess whether the bin assignments are the same for both datasets (Brian's and mine). To confirm that the measured jaccard index (0.23) isn't anything to write home about we can generate  an empirical p value (ep = 0.13).</span></span>
<span id="cb39-882"><a href="#cb39-882"></a></span>
<span id="cb39-883"><a href="#cb39-883"></a></span>
<span id="cb39-884"><a href="#cb39-884"></a><span class="st">``` r</span></span>
<span id="cb39-885"><a href="#cb39-885"></a><span class="st"># cor_comp_df</span></span>
<span id="cb39-886"><a href="#cb39-886"></a><span class="st">#</span></span>
<span id="cb39-887"><a href="#cb39-887"></a><span class="st">#    Source Time     x     y        Corr</span></span>
<span id="cb39-888"><a href="#cb39-888"></a><span class="st">#    &lt;fct&gt;  &lt;fct&gt;    &lt;chr&gt; &lt;chr&gt;   &lt;dbl&gt;</span></span>
<span id="cb39-889"><a href="#cb39-889"></a><span class="st">#  1 Brian  Baseline bkkca cav1    0.441</span></span>
<span id="cb39-890"><a href="#cb39-890"></a><span class="st">#  2 Brian  Baseline bkkca cav2    0.476</span></span>
<span id="cb39-891"><a href="#cb39-891"></a><span class="st">#  3 Brian  Baseline bkkca inx1    0.435</span></span>
<span id="cb39-892"><a href="#cb39-892"></a><span class="st">#  4 Brian  Baseline bkkca inx2    0.159</span></span>
<span id="cb39-893"><a href="#cb39-893"></a><span class="st">#  5 Brian  Baseline bkkca inx3   -0.174</span></span>
<span id="cb39-894"><a href="#cb39-894"></a><span class="st">#  6 Brian  Baseline bkkca inx4   -0.167</span></span>
<span id="cb39-895"><a href="#cb39-895"></a></span>
<span id="cb39-896"><a href="#cb39-896"></a><span class="st">n_bins &lt;- 5</span></span>
<span id="cb39-897"><a href="#cb39-897"></a></span>
<span id="cb39-898"><a href="#cb39-898"></a><span class="st">temp &lt;- cor_comp_df %&gt;% </span></span>
<span id="cb39-899"><a href="#cb39-899"></a><span class="st">  mutate(Bins = cut(Corr, breaks = seq(-1, 1, length.out = n_bins))) %&gt;% </span></span>
<span id="cb39-900"><a href="#cb39-900"></a><span class="st">  select(-Corr) %&gt;% </span></span>
<span id="cb39-901"><a href="#cb39-901"></a><span class="st">  pivot_wider(names_from = "Source", values_from = "Bins")</span></span>
<span id="cb39-902"><a href="#cb39-902"></a></span>
<span id="cb39-903"><a href="#cb39-903"></a><span class="st"># temp</span></span>
<span id="cb39-904"><a href="#cb39-904"></a><span class="st">#</span></span>
<span id="cb39-905"><a href="#cb39-905"></a><span class="st">#   Time     x     y     Brian    Daniel  </span></span>
<span id="cb39-906"><a href="#cb39-906"></a><span class="st">#   &lt;fct&gt;    &lt;chr&gt; &lt;chr&gt; &lt;fct&gt;    &lt;fct&gt;   </span></span>
<span id="cb39-907"><a href="#cb39-907"></a><span class="st"># 1 Baseline bkkca cav1  (0,0.5]  (0,0.5] </span></span>
<span id="cb39-908"><a href="#cb39-908"></a><span class="st"># 2 Baseline bkkca cav2  (0,0.5]  (0,0.5] </span></span>
<span id="cb39-909"><a href="#cb39-909"></a><span class="st"># 3 Baseline bkkca inx1  (0,0.5]  (0.5,1] </span></span>
<span id="cb39-910"><a href="#cb39-910"></a><span class="st"># 4 Baseline bkkca inx2  (0,0.5]  (0.5,1] </span></span>
<span id="cb39-911"><a href="#cb39-911"></a><span class="st"># 5 Baseline bkkca inx3  (-0.5,0] (0.5,1] </span></span>
<span id="cb39-912"><a href="#cb39-912"></a><span class="st"># 6 Baseline bkkca inx4  (-0.5,0] (-0.5,0]</span></span>
<span id="cb39-913"><a href="#cb39-913"></a></span>
<span id="cb39-914"><a href="#cb39-914"></a></span>
<span id="cb39-915"><a href="#cb39-915"></a><span class="st">library('clusteval')</span></span>
<span id="cb39-916"><a href="#cb39-916"></a><span class="st">obs_jaccard &lt;- cluster_similarity(temp$Brian, temp$Daniel, similarity="jaccard")  </span></span>
<span id="cb39-917"><a href="#cb39-917"></a></span>
<span id="cb39-918"><a href="#cb39-918"></a><span class="st"># 0.2304234</span></span>
<span id="cb39-919"><a href="#cb39-919"></a></span>
<span id="cb39-920"><a href="#cb39-920"></a><span class="st">null_jaccard &lt;- map(1:10000, function(i){</span></span>
<span id="cb39-921"><a href="#cb39-921"></a><span class="st">  cluster_similarity(sample(temp$Brian, replace = F), </span></span>
<span id="cb39-922"><a href="#cb39-922"></a><span class="st">                     temp$Daniel, similarity="jaccard")</span></span>
<span id="cb39-923"><a href="#cb39-923"></a><span class="st">  }) %&gt;% </span></span>
<span id="cb39-924"><a href="#cb39-924"></a><span class="st">  unlist()</span></span>
<span id="cb39-925"><a href="#cb39-925"></a></span>
<span id="cb39-926"><a href="#cb39-926"></a></span>
<span id="cb39-927"><a href="#cb39-927"></a><span class="st">temp &lt;- with(density(null_jaccard), data.frame(x, y))</span></span>
<span id="cb39-928"><a href="#cb39-928"></a><span class="st">temp &lt;- temp %&gt;% mutate(xmax = max(x),</span></span>
<span id="cb39-929"><a href="#cb39-929"></a><span class="st">                obs = obs_jaccard)</span></span>
<span id="cb39-930"><a href="#cb39-930"></a></span>
<span id="cb39-931"><a href="#cb39-931"></a><span class="st">ggplot(data = temp, aes(x = x, y = y))+</span></span>
<span id="cb39-932"><a href="#cb39-932"></a><span class="st">  geom_line()+</span></span>
<span id="cb39-933"><a href="#cb39-933"></a><span class="st">  geom_vline(xintercept = obs_jaccard)+</span></span>
<span id="cb39-934"><a href="#cb39-934"></a><span class="st">  geom_ribbon(data = temp[temp$x &gt; temp$obs, ], </span></span>
<span id="cb39-935"><a href="#cb39-935"></a><span class="st">              aes(xmin = obs, xmax = xmax, ymin = 0, ymax = y))+</span></span>
<span id="cb39-936"><a href="#cb39-936"></a><span class="st">  labs(subtitle = paste("empirical p=", as.character(mean(null_jaccard &gt;= obs_jaccard))))</span></span>
<span id="cb39-937"><a href="#cb39-937"></a><span class="st">```</span></span>
<span id="cb39-938"><a href="#cb39-938"></a></span>
<span id="cb39-939"><a href="#cb39-939"></a></span>
<span id="cb39-940"><a href="#cb39-940"></a><span class="st">![image (22).png](image%20(22).png)</span></span>
<span id="cb39-941"><a href="#cb39-941"></a></span>
<span id="cb39-942"><a href="#cb39-942"></a><span class="st">2021-2-10</span></span>
<span id="cb39-943"><a href="#cb39-943"></a><span class="st">It's worth generating an empirical p value for each comparison you're making. For example here I'm comparing the results of an experiment replicate. Each dependent variable is assigned a group based on if one would conclude it there was a difference (0 or 1) between groups and the sign of that difference (+ or -). Seeing a Jaccard index of 0.61 (out of 1) we might conclude we replicated most of the findings. _However,_ the empirical p value is 1 because most of the comparisons were non-significant in both groups resulting in a high floor for the index.</span></span>
<span id="cb39-944"><a href="#cb39-944"></a><span class="st">![image (23).png](image%20(23).png)</span></span>
<span id="cb39-945"><a href="#cb39-945"></a></span>
<span id="cb39-946"><a href="#cb39-946"></a></span>
<span id="cb39-947"><a href="#cb39-947"></a><span class="st">-xcutx-</span></span>
<span id="cb39-948"><a href="#cb39-948"></a></span>
<span id="cb39-949"><a href="#cb39-949"></a></span>
<span id="cb39-950"><a href="#cb39-950"></a></span>
<span id="cb39-951"><a href="#cb39-951"></a><span class="st">2021-2-21</span></span>
<span id="cb39-952"><a href="#cb39-952"></a><span class="st">I think I have a solution to get decent enough dendrograms without fussing with base graphics.</span></span>
<span id="cb39-953"><a href="#cb39-953"></a></span>
<span id="cb39-954"><a href="#cb39-954"></a><span class="st">The overview of my workaround is to cluster with `pvclust`, extract `$hclust` , plot it as a dendrogram, coerce into a ggplot.  This makes it easy enough to replicate the functionality of the `colored_bars()`  function by making additional plots.  The function below makes a few plots in addition to the dendrogram. If you end up working with base graphics anyway, `dendextend` is still worth a look.</span></span>
<span id="cb39-955"><a href="#cb39-955"></a></span>
<span id="cb39-956"><a href="#cb39-956"></a><span class="st">Here's an example:</span></span>
<span id="cb39-957"><a href="#cb39-957"></a></span>
<span id="cb39-958"><a href="#cb39-958"></a><span class="st">``` r</span></span>
<span id="cb39-959"><a href="#cb39-959"></a><span class="st"># needed </span></span>
<span id="cb39-960"><a href="#cb39-960"></a><span class="st">library(pvclust)</span></span>
<span id="cb39-961"><a href="#cb39-961"></a><span class="st">library(tidyverse)</span></span>
<span id="cb39-962"><a href="#cb39-962"></a><span class="st">library(dendextend) # for color_labels</span></span>
<span id="cb39-963"><a href="#cb39-963"></a><span class="st">library(ggnewscale) # to accommodate two fill scales </span></span>
<span id="cb39-964"><a href="#cb39-964"></a></span>
<span id="cb39-965"><a href="#cb39-965"></a><span class="st"># recommended</span></span>
<span id="cb39-966"><a href="#cb39-966"></a><span class="st"># install.packages("palmerpenguins")</span></span>
<span id="cb39-967"><a href="#cb39-967"></a><span class="st">library(palmerpenguins)</span></span>
<span id="cb39-968"><a href="#cb39-968"></a><span class="st">library(patchwork)</span></span>
<span id="cb39-969"><a href="#cb39-969"></a><span class="st">library(scales) # for overiding scientific notation on dendrogram y axis </span></span>
<span id="cb39-970"><a href="#cb39-970"></a></span>
<span id="cb39-971"><a href="#cb39-971"></a><span class="st"># Make a demo dataset</span></span>
<span id="cb39-972"><a href="#cb39-972"></a><span class="st">tux &lt;- select(palmerpenguins::penguins, </span></span>
<span id="cb39-973"><a href="#cb39-973"></a><span class="st">              species,</span></span>
<span id="cb39-974"><a href="#cb39-974"></a><span class="st">              bill_length_mm, bill_depth_mm, </span></span>
<span id="cb39-975"><a href="#cb39-975"></a><span class="st">              flipper_length_mm, body_mass_g) </span></span>
<span id="cb39-976"><a href="#cb39-976"></a></span>
<span id="cb39-977"><a href="#cb39-977"></a><span class="st">tux &lt;- tux[complete.cases(tux), ]</span></span>
<span id="cb39-978"><a href="#cb39-978"></a></span>
<span id="cb39-979"><a href="#cb39-979"></a><span class="st">set.seed(54646)</span></span>
<span id="cb39-980"><a href="#cb39-980"></a><span class="st">tux &lt;- tux[sample(1:nrow(tux), 30), ] # for faster demo clustering</span></span>
<span id="cb39-981"><a href="#cb39-981"></a></span>
<span id="cb39-982"><a href="#cb39-982"></a><span class="st"># Example use</span></span>
<span id="cb39-983"><a href="#cb39-983"></a><span class="st">o &lt;- </span></span>
<span id="cb39-984"><a href="#cb39-984"></a><span class="st">  mk_hclust_plts(</span></span>
<span id="cb39-985"><a href="#cb39-985"></a><span class="st">    df = mutate(tux, uid = paste(seq(1, nrow(tux)), species, sep = "-")),</span></span>
<span id="cb39-986"><a href="#cb39-986"></a><span class="st">    cluster_by = c("bill_length_mm", "bill_depth_mm", </span></span>
<span id="cb39-987"><a href="#cb39-987"></a><span class="st">                   "flipper_length_mm", "body_mass_g"),</span></span>
<span id="cb39-988"><a href="#cb39-988"></a><span class="st">    uid_col = "uid",</span></span>
<span id="cb39-989"><a href="#cb39-989"></a><span class="st">    n_clusters = 3,</span></span>
<span id="cb39-990"><a href="#cb39-990"></a><span class="st">    true_groups = "species",</span></span>
<span id="cb39-991"><a href="#cb39-991"></a><span class="st">    true_colors = RColorBrewer::brewer.pal(3, "Set2"),</span></span>
<span id="cb39-992"><a href="#cb39-992"></a><span class="st">    cluster_colors = RColorBrewer::brewer.pal(3, "Set1") </span></span>
<span id="cb39-993"><a href="#cb39-993"></a><span class="st">  )</span></span>
<span id="cb39-994"><a href="#cb39-994"></a></span>
<span id="cb39-995"><a href="#cb39-995"></a></span>
<span id="cb39-996"><a href="#cb39-996"></a><span class="st"># Patchwork to arrange the output plots</span></span>
<span id="cb39-997"><a href="#cb39-997"></a><span class="st">(o$dendrogram_both+</span></span>
<span id="cb39-998"><a href="#cb39-998"></a><span class="st">    scale_y_continuous(limits = c(-.0001, 0.00022), labels = scales::comma)</span></span>
<span id="cb39-999"><a href="#cb39-999"></a><span class="st">)/ </span></span>
<span id="cb39-1000"><a href="#cb39-1000"></a><span class="st">  (o$group_compare_tile+</span></span>
<span id="cb39-1001"><a href="#cb39-1001"></a><span class="st">     # lims(y =  c(2, -2))+ # y axis can be flipped like so</span></span>
<span id="cb39-1002"><a href="#cb39-1002"></a><span class="st">     theme(legend.position = "")</span></span>
<span id="cb39-1003"><a href="#cb39-1003"></a><span class="st">  ) / </span></span>
<span id="cb39-1004"><a href="#cb39-1004"></a><span class="st">  (o$heatmap_raw + theme(legend.position = "right")) / </span></span>
<span id="cb39-1005"><a href="#cb39-1005"></a><span class="st">  (o$heatmap_z + theme(legend.position = "right")) + </span></span>
<span id="cb39-1006"><a href="#cb39-1006"></a><span class="st">  patchwork::plot_layout(heights = c(5, .3, 1.25, 1.25))</span></span>
<span id="cb39-1007"><a href="#cb39-1007"></a></span>
<span id="cb39-1008"><a href="#cb39-1008"></a></span>
<span id="cb39-1009"><a href="#cb39-1009"></a><span class="st"># example 2</span></span>
<span id="cb39-1010"><a href="#cb39-1010"></a></span>
<span id="cb39-1011"><a href="#cb39-1011"></a><span class="st"># o &lt;- </span></span>
<span id="cb39-1012"><a href="#cb39-1012"></a><span class="st"># mk_hclust_plts(</span></span>
<span id="cb39-1013"><a href="#cb39-1013"></a><span class="st">#   df = mutate(iris, uid = paste(seq(1, nrow(iris)), Species, sep = "-")),</span></span>
<span id="cb39-1014"><a href="#cb39-1014"></a><span class="st">#   cluster_by = c("Sepal.Length", "Sepal.Width", "Petal.Length", "Petal.Width"),</span></span>
<span id="cb39-1015"><a href="#cb39-1015"></a><span class="st">#   uid_col = "uid",</span></span>
<span id="cb39-1016"><a href="#cb39-1016"></a><span class="st">#   n_clusters = 3,</span></span>
<span id="cb39-1017"><a href="#cb39-1017"></a><span class="st">#   true_groups = "Species",</span></span>
<span id="cb39-1018"><a href="#cb39-1018"></a><span class="st">#   true_colors = RColorBrewer::brewer.pal(3, "Set2"),</span></span>
<span id="cb39-1019"><a href="#cb39-1019"></a><span class="st">#   cluster_colors = RColorBrewer::brewer.pal(3, "Set1") </span></span>
<span id="cb39-1020"><a href="#cb39-1020"></a><span class="st"># )</span></span>
<span id="cb39-1021"><a href="#cb39-1021"></a><span class="st">```</span></span>
<span id="cb39-1022"><a href="#cb39-1022"></a></span>
<span id="cb39-1023"><a href="#cb39-1023"></a><span class="st">ps, it's worth checking your code on sample datasets (`penguins`, `iris`, `mpg`, etc. ). That'll help iron out weird behavior sooner rather than later.</span></span>
<span id="cb39-1024"><a href="#cb39-1024"></a><span class="st">![image (25).png](image%20(25).png)</span></span>
<span id="cb39-1025"><a href="#cb39-1025"></a></span>
<span id="cb39-1026"><a href="#cb39-1026"></a><span class="st">2021-2-21</span></span>
<span id="cb39-1027"><a href="#cb39-1027"></a><span class="st">Daniel</span></span>
<span id="cb39-1028"><a href="#cb39-1028"></a><span class="st">And here's the source for the function itself.</span></span>
<span id="cb39-1029"><a href="#cb39-1029"></a></span>
<span id="cb39-1030"><a href="#cb39-1030"></a></span>
<span id="cb39-1031"><a href="#cb39-1031"></a><span class="st">2021-2-21</span></span>
<span id="cb39-1032"><a href="#cb39-1032"></a><span class="st">Daniel</span></span>
<span id="cb39-1033"><a href="#cb39-1033"></a><span class="st">Edit: The original function here depended on the factor levels of the clusters and true groups to make the color's consistent between plots (e.g. factors ordered abcdABCD not aAbBcCdD). This breaks down with some cases (e.g. if you start group names with a number (e.g. 0hours)). The below edit uses `ggnewscale`  to fix this.</span></span>
<span id="cb39-1034"><a href="#cb39-1034"></a></span>
<span id="cb39-1035"><a href="#cb39-1035"></a><span class="st">&lt;https://github.com/eliocamp/ggnewscale&gt;</span></span>
<span id="cb39-1036"><a href="#cb39-1036"></a></span>
<span id="cb39-1037"><a href="#cb39-1037"></a></span>
<span id="cb39-1038"><a href="#cb39-1038"></a><span class="st">``` r</span></span>
<span id="cb39-1039"><a href="#cb39-1039"></a><span class="st">mk_hclust_plts &lt;- function(</span></span>
<span id="cb39-1040"><a href="#cb39-1040"></a><span class="st">  df = unite(M_winxiqr, uid, Experiment, Cell, sep = "-"),</span></span>
<span id="cb39-1041"><a href="#cb39-1041"></a><span class="st">  cluster_by = c("vrest", "r11", "r1", "Ihtk.0", "Ihtk.Slope", "Ia.0", "Ia.Slope"),</span></span>
<span id="cb39-1042"><a href="#cb39-1042"></a><span class="st">  uid_col = "uid",</span></span>
<span id="cb39-1043"><a href="#cb39-1043"></a><span class="st">  n_clusters = 3,</span></span>
<span id="cb39-1044"><a href="#cb39-1044"></a><span class="st">  true_groups = "Condition",</span></span>
<span id="cb39-1045"><a href="#cb39-1045"></a><span class="st">  true_colors = RColorBrewer::brewer.pal(3, "Set2"),</span></span>
<span id="cb39-1046"><a href="#cb39-1046"></a><span class="st">  cluster_colors = RColorBrewer::brewer.pal(3, "Set1")</span></span>
<span id="cb39-1047"><a href="#cb39-1047"></a><span class="st">){</span></span>
<span id="cb39-1048"><a href="#cb39-1048"></a><span class="st">  df &lt;- as.data.frame(df) </span></span>
<span id="cb39-1049"><a href="#cb39-1049"></a><span class="st">  </span></span>
<span id="cb39-1050"><a href="#cb39-1050"></a><span class="st">  if (!exists("cluster_colors")){</span></span>
<span id="cb39-1051"><a href="#cb39-1051"></a><span class="st">    cluster_colors = rainbow(n_clusters)</span></span>
<span id="cb39-1052"><a href="#cb39-1052"></a><span class="st">  }</span></span>
<span id="cb39-1053"><a href="#cb39-1053"></a><span class="st">  ## prep</span></span>
<span id="cb39-1054"><a href="#cb39-1054"></a><span class="st">  # move uid to rowname</span></span>
<span id="cb39-1055"><a href="#cb39-1055"></a><span class="st">  row.names(df) &lt;- df[[uid_col]]</span></span>
<span id="cb39-1056"><a href="#cb39-1056"></a><span class="st">  df_groups &lt;- select(df, all_of(true_groups))</span></span>
<span id="cb39-1057"><a href="#cb39-1057"></a><span class="st">  df &lt;- df[, cluster_by]</span></span>
<span id="cb39-1058"><a href="#cb39-1058"></a><span class="st">  </span></span>
<span id="cb39-1059"><a href="#cb39-1059"></a><span class="st">  ## Cluster</span></span>
<span id="cb39-1060"><a href="#cb39-1060"></a><span class="st">  cluster &lt;- pvclust(t(df),</span></span>
<span id="cb39-1061"><a href="#cb39-1061"></a><span class="st">                     method.hclust = "ward.D2",</span></span>
<span id="cb39-1062"><a href="#cb39-1062"></a><span class="st">                     method.dist = "correlation",</span></span>
<span id="cb39-1063"><a href="#cb39-1063"></a><span class="st">                     use.cor = "pairwise.complete.obs")</span></span>
<span id="cb39-1064"><a href="#cb39-1064"></a><span class="st">  </span></span>
<span id="cb39-1065"><a href="#cb39-1065"></a><span class="st">  </span></span>
<span id="cb39-1066"><a href="#cb39-1066"></a><span class="st">  ## make dendrogram  ####</span></span>
<span id="cb39-1067"><a href="#cb39-1067"></a><span class="st">  dend &lt;- cluster$hclust %&gt;% </span></span>
<span id="cb39-1068"><a href="#cb39-1068"></a><span class="st">    as.dendrogram() </span></span>
<span id="cb39-1069"><a href="#cb39-1069"></a><span class="st">  </span></span>
<span id="cb39-1070"><a href="#cb39-1070"></a><span class="st">  # iteratively coloring the labels is a workaround to get the "true" groups shown</span></span>
<span id="cb39-1071"><a href="#cb39-1071"></a><span class="st">  dend_labs &lt;- rownames_to_column(df_groups, var = "rownames")[, ]</span></span>
<span id="cb39-1072"><a href="#cb39-1072"></a><span class="st">  dend_labs &lt;- full_join(data.frame(rownames = labels(dend)), dend_labs)</span></span>
<span id="cb39-1073"><a href="#cb39-1073"></a><span class="st">  for(i in seq_along(unique(df_groups[[true_groups]]))){</span></span>
<span id="cb39-1074"><a href="#cb39-1074"></a><span class="st">    true_group &lt;- unique(df_groups[[true_groups]])[i]</span></span>
<span id="cb39-1075"><a href="#cb39-1075"></a><span class="st">    true_color &lt;- true_colors[i]</span></span>
<span id="cb39-1076"><a href="#cb39-1076"></a><span class="st">    </span></span>
<span id="cb39-1077"><a href="#cb39-1077"></a><span class="st">    dend &lt;- dend %&gt;% </span></span>
<span id="cb39-1078"><a href="#cb39-1078"></a><span class="st">      dendextend::color_labels(</span></span>
<span id="cb39-1079"><a href="#cb39-1079"></a><span class="st">        col = true_color, </span></span>
<span id="cb39-1080"><a href="#cb39-1080"></a><span class="st">        labels = dend_labs[dend_labs[[true_groups]] == true_group, "rownames"]) </span></span>
<span id="cb39-1081"><a href="#cb39-1081"></a><span class="st">    </span></span>
<span id="cb39-1082"><a href="#cb39-1082"></a><span class="st">    </span></span>
<span id="cb39-1083"><a href="#cb39-1083"></a><span class="st">  }</span></span>
<span id="cb39-1084"><a href="#cb39-1084"></a><span class="st">  </span></span>
<span id="cb39-1085"><a href="#cb39-1085"></a><span class="st">  dend &lt;- dend %&gt;% </span></span>
<span id="cb39-1086"><a href="#cb39-1086"></a><span class="st">    set("branches_k_color", </span></span>
<span id="cb39-1087"><a href="#cb39-1087"></a><span class="st">        k = n_clusters, </span></span>
<span id="cb39-1088"><a href="#cb39-1088"></a><span class="st">        value = cluster_colors</span></span>
<span id="cb39-1089"><a href="#cb39-1089"></a><span class="st">    ) %&gt;% </span></span>
<span id="cb39-1090"><a href="#cb39-1090"></a><span class="st">    set("branches_lwd", 0.7) %&gt;%</span></span>
<span id="cb39-1091"><a href="#cb39-1091"></a><span class="st">    set("labels_cex", 0.6) </span></span>
<span id="cb39-1092"><a href="#cb39-1092"></a></span>
<span id="cb39-1093"><a href="#cb39-1093"></a><span class="st">  dend_cluster_only &lt;- dend %&gt;% </span></span>
<span id="cb39-1094"><a href="#cb39-1094"></a><span class="st">    set("labels_colors",</span></span>
<span id="cb39-1095"><a href="#cb39-1095"></a><span class="st">        k = n_clusters,</span></span>
<span id="cb39-1096"><a href="#cb39-1096"></a><span class="st">        value = cluster_colors) %&gt;%</span></span>
<span id="cb39-1097"><a href="#cb39-1097"></a><span class="st">    as.ggdend()</span></span>
<span id="cb39-1098"><a href="#cb39-1098"></a><span class="st">  </span></span>
<span id="cb39-1099"><a href="#cb39-1099"></a><span class="st">  </span></span>
<span id="cb39-1100"><a href="#cb39-1100"></a><span class="st">  dend &lt;- dend %&gt;% </span></span>
<span id="cb39-1101"><a href="#cb39-1101"></a><span class="st">    as.ggdend()</span></span>
<span id="cb39-1102"><a href="#cb39-1102"></a><span class="st">  </span></span>
<span id="cb39-1103"><a href="#cb39-1103"></a><span class="st">  </span></span>
<span id="cb39-1104"><a href="#cb39-1104"></a><span class="st">  </span></span>
<span id="cb39-1105"><a href="#cb39-1105"></a><span class="st">  plt_dend_cluster_only &lt;- ggplot(dend_cluster_only)+</span></span>
<span id="cb39-1106"><a href="#cb39-1106"></a><span class="st">    theme(axis.ticks.y = element_line(),</span></span>
<span id="cb39-1107"><a href="#cb39-1107"></a><span class="st">          axis.text.y = element_text(),</span></span>
<span id="cb39-1108"><a href="#cb39-1108"></a><span class="st">          axis.line.y = element_line())</span></span>
<span id="cb39-1109"><a href="#cb39-1109"></a><span class="st">  </span></span>
<span id="cb39-1110"><a href="#cb39-1110"></a><span class="st">  </span></span>
<span id="cb39-1111"><a href="#cb39-1111"></a><span class="st">  plt_dend &lt;- ggplot(dend)+</span></span>
<span id="cb39-1112"><a href="#cb39-1112"></a><span class="st">    theme(axis.ticks.y = element_line(),</span></span>
<span id="cb39-1113"><a href="#cb39-1113"></a><span class="st">          axis.text.y = element_text(),</span></span>
<span id="cb39-1114"><a href="#cb39-1114"></a><span class="st">          axis.line.y = element_line())  </span></span>
<span id="cb39-1115"><a href="#cb39-1115"></a><span class="st">  </span></span>
<span id="cb39-1116"><a href="#cb39-1116"></a></span>
<span id="cb39-1117"><a href="#cb39-1117"></a><span class="st">  ## Add reality ribbon with or without clustering result ####</span></span>
<span id="cb39-1118"><a href="#cb39-1118"></a><span class="st">  groups_to_plt &lt;- full_join(</span></span>
<span id="cb39-1119"><a href="#cb39-1119"></a><span class="st">    as.data.frame(dend$labels),</span></span>
<span id="cb39-1120"><a href="#cb39-1120"></a><span class="st">    rownames_to_column(var = "label", df_groups))</span></span>
<span id="cb39-1121"><a href="#cb39-1121"></a><span class="st">  </span></span>
<span id="cb39-1122"><a href="#cb39-1122"></a><span class="st">  plt_grouping &lt;- groups_to_plt %&gt;% </span></span>
<span id="cb39-1123"><a href="#cb39-1123"></a><span class="st">    ggplot(aes_string(x="x", y="0", fill = true_groups))+</span></span>
<span id="cb39-1124"><a href="#cb39-1124"></a><span class="st">    geom_tile()+</span></span>
<span id="cb39-1125"><a href="#cb39-1125"></a><span class="st">    scale_fill_manual(values = true_colors)+</span></span>
<span id="cb39-1126"><a href="#cb39-1126"></a><span class="st">    theme_void()+</span></span>
<span id="cb39-1127"><a href="#cb39-1127"></a><span class="st">    labs(x = "", y = "")+</span></span>
<span id="cb39-1128"><a href="#cb39-1128"></a><span class="st">    theme(legend.position = "left")</span></span>
<span id="cb39-1129"><a href="#cb39-1129"></a><span class="st">  </span></span>
<span id="cb39-1130"><a href="#cb39-1130"></a><span class="st">  plt_grouping_contrast &lt;- ggplot()+</span></span>
<span id="cb39-1131"><a href="#cb39-1131"></a><span class="st">    geom_tile(data = groups_to_plt, aes_string(x="x", y="0.5", fill = true_groups))+</span></span>
<span id="cb39-1132"><a href="#cb39-1132"></a><span class="st">    scale_fill_manual(values = true_colors)+</span></span>
<span id="cb39-1133"><a href="#cb39-1133"></a><span class="st">    </span></span>
<span id="cb39-1134"><a href="#cb39-1134"></a><span class="st">    ggnewscale::new_scale("fill") +</span></span>
<span id="cb39-1135"><a href="#cb39-1135"></a><span class="st">    geom_tile(data = data.frame(x = seq_along(dend_cluster_only$labels$col),</span></span>
<span id="cb39-1136"><a href="#cb39-1136"></a><span class="st">                                cluster_groups = as.character(as.numeric(as.factor(dend_cluster_only$labels$col)))</span></span>
<span id="cb39-1137"><a href="#cb39-1137"></a><span class="st">    ),</span></span>
<span id="cb39-1138"><a href="#cb39-1138"></a><span class="st">    aes_string(x="x", y= "-0.5", fill = "cluster_groups"),</span></span>
<span id="cb39-1139"><a href="#cb39-1139"></a><span class="st">    )+</span></span>
<span id="cb39-1140"><a href="#cb39-1140"></a><span class="st">    scale_fill_manual(values = cluster_colors)+</span></span>
<span id="cb39-1141"><a href="#cb39-1141"></a><span class="st">    </span></span>
<span id="cb39-1142"><a href="#cb39-1142"></a><span class="st">    theme_void()+</span></span>
<span id="cb39-1143"><a href="#cb39-1143"></a><span class="st">    labs(x = "", y = "")+</span></span>
<span id="cb39-1144"><a href="#cb39-1144"></a><span class="st">    theme(legend.position = "left")</span></span>
<span id="cb39-1145"><a href="#cb39-1145"></a><span class="st">  </span></span>
<span id="cb39-1146"><a href="#cb39-1146"></a><span class="st">  </span></span>
<span id="cb39-1147"><a href="#cb39-1147"></a><span class="st">  ## Add heatmap  ####</span></span>
<span id="cb39-1148"><a href="#cb39-1148"></a><span class="st">  data_to_plt &lt;- full_join(</span></span>
<span id="cb39-1149"><a href="#cb39-1149"></a><span class="st">    as.data.frame(dend$labels),</span></span>
<span id="cb39-1150"><a href="#cb39-1150"></a><span class="st">    rownames_to_column(var = "label", df)) </span></span>
<span id="cb39-1151"><a href="#cb39-1151"></a><span class="st">  </span></span>
<span id="cb39-1152"><a href="#cb39-1152"></a><span class="st">  data_to_plt &lt;- </span></span>
<span id="cb39-1153"><a href="#cb39-1153"></a><span class="st">    data_to_plt %&gt;% </span></span>
<span id="cb39-1154"><a href="#cb39-1154"></a><span class="st">    gather("key", "value", </span></span>
<span id="cb39-1155"><a href="#cb39-1155"></a><span class="st">           names(data_to_plt)[</span></span>
<span id="cb39-1156"><a href="#cb39-1156"></a><span class="st">             !(names(data_to_plt) </span><span class="sc">%i</span><span class="st">n</span><span class="sc">% c</span><span class="st">("x", "y", </span></span>
<span id="cb39-1157"><a href="#cb39-1157"></a><span class="st">                                         "label", "col", "cex", </span></span>
<span id="cb39-1158"><a href="#cb39-1158"></a><span class="st">                                         true_groups))</span></span>
<span id="cb39-1159"><a href="#cb39-1159"></a><span class="st">           ])</span></span>
<span id="cb39-1160"><a href="#cb39-1160"></a><span class="st">  </span></span>
<span id="cb39-1161"><a href="#cb39-1161"></a><span class="st">  plt_heatmap_raw &lt;- data_to_plt %&gt;% </span></span>
<span id="cb39-1162"><a href="#cb39-1162"></a><span class="st">    ggplot(aes(x, </span></span>
<span id="cb39-1163"><a href="#cb39-1163"></a><span class="st">               y = key, </span></span>
<span id="cb39-1164"><a href="#cb39-1164"></a><span class="st">               fill = value))+</span></span>
<span id="cb39-1165"><a href="#cb39-1165"></a><span class="st">    geom_tile()+</span></span>
<span id="cb39-1166"><a href="#cb39-1166"></a><span class="st">    scale_fill_viridis_c()+</span></span>
<span id="cb39-1167"><a href="#cb39-1167"></a><span class="st">    labs(x = "", y = "")+</span></span>
<span id="cb39-1168"><a href="#cb39-1168"></a><span class="st">    theme(panel.background = element_blank(),</span></span>
<span id="cb39-1169"><a href="#cb39-1169"></a><span class="st">          axis.ticks.x = element_blank(),</span></span>
<span id="cb39-1170"><a href="#cb39-1170"></a><span class="st">          axis.text.x = element_blank(),</span></span>
<span id="cb39-1171"><a href="#cb39-1171"></a><span class="st">          legend.position = "left")</span></span>
<span id="cb39-1172"><a href="#cb39-1172"></a><span class="st">  </span></span>
<span id="cb39-1173"><a href="#cb39-1173"></a><span class="st">  </span></span>
<span id="cb39-1174"><a href="#cb39-1174"></a><span class="st">  plt_heatmap_z &lt;- data_to_plt %&gt;% </span></span>
<span id="cb39-1175"><a href="#cb39-1175"></a><span class="st">    group_by(key) %&gt;% </span></span>
<span id="cb39-1176"><a href="#cb39-1176"></a><span class="st">    mutate(mean = mean(value, na.rm = T),</span></span>
<span id="cb39-1177"><a href="#cb39-1177"></a><span class="st">           sd = sd(value, na.rm = T)) %&gt;% </span></span>
<span id="cb39-1178"><a href="#cb39-1178"></a><span class="st">    mutate(value = ((value - mean)/sd)) %&gt;% # Now Z scores</span></span>
<span id="cb39-1179"><a href="#cb39-1179"></a><span class="st">    ggplot(aes(x, </span></span>
<span id="cb39-1180"><a href="#cb39-1180"></a><span class="st">               y = key, </span></span>
<span id="cb39-1181"><a href="#cb39-1181"></a><span class="st">               fill = value))+</span></span>
<span id="cb39-1182"><a href="#cb39-1182"></a><span class="st">    geom_tile()+</span></span>
<span id="cb39-1183"><a href="#cb39-1183"></a><span class="st">    scale_fill_viridis_c()+</span></span>
<span id="cb39-1184"><a href="#cb39-1184"></a><span class="st">    labs(x = "", y = "")+</span></span>
<span id="cb39-1185"><a href="#cb39-1185"></a><span class="st">    theme(panel.background = element_blank(),</span></span>
<span id="cb39-1186"><a href="#cb39-1186"></a><span class="st">          axis.ticks.x = element_blank(),</span></span>
<span id="cb39-1187"><a href="#cb39-1187"></a><span class="st">          axis.text.x = element_blank(),</span></span>
<span id="cb39-1188"><a href="#cb39-1188"></a><span class="st">          legend.position = "left")</span></span>
<span id="cb39-1189"><a href="#cb39-1189"></a><span class="st">  </span></span>
<span id="cb39-1190"><a href="#cb39-1190"></a><span class="st">  </span></span>
<span id="cb39-1191"><a href="#cb39-1191"></a><span class="st">  ## Return plots, manually tweak layout  ####</span></span>
<span id="cb39-1192"><a href="#cb39-1192"></a><span class="st">  return(</span></span>
<span id="cb39-1193"><a href="#cb39-1193"></a><span class="st">    list(</span></span>
<span id="cb39-1194"><a href="#cb39-1194"></a><span class="st">      pvclust_out = cluster,</span></span>
<span id="cb39-1195"><a href="#cb39-1195"></a><span class="st">      dendrogram_clusters = plt_dend_cluster_only,</span></span>
<span id="cb39-1196"><a href="#cb39-1196"></a><span class="st">      dendrogram_both = plt_dend,</span></span>
<span id="cb39-1197"><a href="#cb39-1197"></a><span class="st">      group_tile = plt_grouping,</span></span>
<span id="cb39-1198"><a href="#cb39-1198"></a><span class="st">      group_compare_tile = plt_grouping_contrast,</span></span>
<span id="cb39-1199"><a href="#cb39-1199"></a><span class="st">      heatmap_raw = plt_heatmap_raw,</span></span>
<span id="cb39-1200"><a href="#cb39-1200"></a><span class="st">      heatmap_z = plt_heatmap_z</span></span>
<span id="cb39-1201"><a href="#cb39-1201"></a><span class="st">    )</span></span>
<span id="cb39-1202"><a href="#cb39-1202"></a><span class="st">  )</span></span>
<span id="cb39-1203"><a href="#cb39-1203"></a><span class="st">}</span></span>
<span id="cb39-1204"><a href="#cb39-1204"></a><span class="st">```</span></span>
<span id="cb39-1205"><a href="#cb39-1205"></a></span>
<span id="cb39-1206"><a href="#cb39-1206"></a></span>
<span id="cb39-1207"><a href="#cb39-1207"></a><span class="st">2021-2-21</span></span>
<span id="cb39-1208"><a href="#cb39-1208"></a><span class="st">David Schulz</span></span>
<span id="cb39-1209"><a href="#cb39-1209"></a><span class="st">I am going to have to find a reason to play around with this, because it looks great.  Thanks!  How can we ensure the world can benefit? Is this new and exciting to others as well?</span></span>
<span id="cb39-1210"><a href="#cb39-1210"></a></span>
<span id="cb39-1211"><a href="#cb39-1211"></a></span>
<span id="cb39-1212"><a href="#cb39-1212"></a></span>
<span id="cb39-1213"><a href="#cb39-1213"></a></span>
<span id="cb39-1214"><a href="#cb39-1214"></a></span>
<span id="cb39-1215"><a href="#cb39-1215"></a><span class="st">2021-2-22</span></span>
<span id="cb39-1216"><a href="#cb39-1216"></a><span class="st">Daniel</span></span>
<span id="cb39-1217"><a href="#cb39-1217"></a><span class="st">It's potentially worth making available online but I don't know that it's all together too useful. The components of it make for a good starting place to write from but the function itself is fairly inflexible (e.g. in terms of clustering method and tree aesthetics). Where I think it could work nicely is in a sort of "lab reference" with recommended/first pass solutions to needs that come up with the sort of data we collect.</span></span>
<span id="cb39-1218"><a href="#cb39-1218"></a></span>
<span id="cb39-1219"><a href="#cb39-1219"></a></span>
<span id="cb39-1220"><a href="#cb39-1220"></a><span class="st">-xcutx-</span></span>
<span id="cb39-1221"><a href="#cb39-1221"></a></span>
<span id="cb39-1222"><a href="#cb39-1222"></a></span>
<span id="cb39-1223"><a href="#cb39-1223"></a><span class="st">2021-3-3</span></span>
<span id="cb39-1224"><a href="#cb39-1224"></a><span class="st">Does anyone have an effective way to deduplicate/shave correlation data frames prior to running a KS test? An inelegant workaround seems to be to order the data frame by correlation and retain every other row. Here's some example data either with or without these duplications.</span></span>
<span id="cb39-1225"><a href="#cb39-1225"></a><span class="st">![image (26).png](image%20(26).png)</span></span>
<span id="cb39-1226"><a href="#cb39-1226"></a><span class="st">![image (27).png](image%20(27).png)</span></span>
<span id="cb39-1227"><a href="#cb39-1227"></a></span>
<span id="cb39-1228"><a href="#cb39-1228"></a><span class="st">2021-3-3</span></span>
<span id="cb39-1229"><a href="#cb39-1229"></a><span class="st">Daniel</span></span>
<span id="cb39-1230"><a href="#cb39-1230"></a><span class="st">I also thought about using a mutate and ifelse to paste the x/y values into an alphabetical identifier (e.g. shal, bkkca, 0.## would become bkkca-shal, 0.##) but doing that seems like a huge pain.</span></span>
<span id="cb39-1231"><a href="#cb39-1231"></a></span>
<span id="cb39-1232"><a href="#cb39-1232"></a><span class="st">2021-3-3</span></span>
<span id="cb39-1233"><a href="#cb39-1233"></a><span class="st">David Schulz</span></span>
<span id="cb39-1234"><a href="#cb39-1234"></a><span class="st">I think this is code you generated for me, no?</span></span>
<span id="cb39-1235"><a href="#cb39-1235"></a></span>
<span id="cb39-1236"><a href="#cb39-1236"></a><span class="st">2021-3-3</span></span>
<span id="cb39-1237"><a href="#cb39-1237"></a><span class="st">David Schulz</span></span>
<span id="cb39-1238"><a href="#cb39-1238"></a></span>
<span id="cb39-1239"><a href="#cb39-1239"></a><span class="st">``` r</span></span>
<span id="cb39-1240"><a href="#cb39-1240"></a><span class="st">temp.list &lt;- map(unique(M$Cell), function(i){</span></span>
<span id="cb39-1241"><a href="#cb39-1241"></a><span class="st">  M %&gt;% </span></span>
<span id="cb39-1242"><a href="#cb39-1242"></a><span class="st">    filter(Cell == i) %&gt;% </span></span>
<span id="cb39-1243"><a href="#cb39-1243"></a><span class="st">    select(-c("Cell")) %&gt;% </span></span>
<span id="cb39-1244"><a href="#cb39-1244"></a><span class="st">    correlate(use = "pairwise.complete.obs", method = "pearson") %&gt;% </span></span>
<span id="cb39-1245"><a href="#cb39-1245"></a><span class="st">    shave() %&gt;% # Prevent double counting of correlations</span></span>
<span id="cb39-1246"><a href="#cb39-1246"></a><span class="st">    mutate(Cell = i)</span></span>
<span id="cb39-1247"><a href="#cb39-1247"></a><span class="st">})</span></span>
<span id="cb39-1248"><a href="#cb39-1248"></a><span class="st">```</span></span>
<span id="cb39-1249"><a href="#cb39-1249"></a></span>
<span id="cb39-1250"><a href="#cb39-1250"></a></span>
<span id="cb39-1251"><a href="#cb39-1251"></a><span class="st">2021-3-3</span></span>
<span id="cb39-1252"><a href="#cb39-1252"></a><span class="st">David Schulz</span></span>
<span id="cb39-1253"><a href="#cb39-1253"></a><span class="st">I'll put the whole thing in here, just to make sure.  Either you generated this for me (which it has your fingerprints on it), or I pilfered this from somewhere.</span></span>
<span id="cb39-1254"><a href="#cb39-1254"></a></span>
<span id="cb39-1255"><a href="#cb39-1255"></a><span class="st">2021-3-3</span></span>
<span id="cb39-1256"><a href="#cb39-1256"></a><span class="st">David Schulz</span></span>
<span id="cb39-1257"><a href="#cb39-1257"></a></span>
<span id="cb39-1258"><a href="#cb39-1258"></a><span class="st">``` r</span></span>
<span id="cb39-1259"><a href="#cb39-1259"></a></span>
<span id="cb39-1260"><a href="#cb39-1260"></a><span class="sc">{r}</span></span>
<span id="cb39-1261"><a href="#cb39-1261"></a><span class="st">library(corrr)</span></span>
<span id="cb39-1262"><a href="#cb39-1262"></a><span class="st">M &lt;- STGGOI</span></span>
<span id="cb39-1263"><a href="#cb39-1263"></a></span>
<span id="cb39-1264"><a href="#cb39-1264"></a><span class="st">temp.list &lt;- map(unique(M$Cell), function(i){</span></span>
<span id="cb39-1265"><a href="#cb39-1265"></a><span class="st">  M %&gt;% </span></span>
<span id="cb39-1266"><a href="#cb39-1266"></a><span class="st">    filter(Cell == i) %&gt;% </span></span>
<span id="cb39-1267"><a href="#cb39-1267"></a><span class="st">    select(-c("Cell")) %&gt;% </span></span>
<span id="cb39-1268"><a href="#cb39-1268"></a><span class="st">    correlate(use = "pairwise.complete.obs", method = "pearson") %&gt;% </span></span>
<span id="cb39-1269"><a href="#cb39-1269"></a><span class="st">    shave() %&gt;% # Prevent double counting of correlations</span></span>
<span id="cb39-1270"><a href="#cb39-1270"></a><span class="st">    mutate(Cell = i)</span></span>
<span id="cb39-1271"><a href="#cb39-1271"></a><span class="st">})</span></span>
<span id="cb39-1272"><a href="#cb39-1272"></a></span>
<span id="cb39-1273"><a href="#cb39-1273"></a></span>
<span id="cb39-1274"><a href="#cb39-1274"></a></span>
<span id="cb39-1275"><a href="#cb39-1275"></a><span class="st">M &lt;- do.call(rbind, temp.list)</span></span>
<span id="cb39-1276"><a href="#cb39-1276"></a><span class="st">M &lt;- M %&gt;% </span></span>
<span id="cb39-1277"><a href="#cb39-1277"></a><span class="st">  gather(colname, cor, names(select(M, -c("rowname", "Cell")))) </span></span>
<span id="cb39-1278"><a href="#cb39-1278"></a></span>
<span id="cb39-1279"><a href="#cb39-1279"></a><span class="st">#M &lt;- subset(M, cor &gt;= 0.5)</span></span>
<span id="cb39-1280"><a href="#cb39-1280"></a><span class="st">plot_ecdf_ks &lt;- function(</span></span>
<span id="cb39-1281"><a href="#cb39-1281"></a><span class="st">  df = temp.df,</span></span>
<span id="cb39-1282"><a href="#cb39-1282"></a><span class="st">  data.col = "cor",</span></span>
<span id="cb39-1283"><a href="#cb39-1283"></a><span class="st">  group.col = "data",</span></span>
<span id="cb39-1284"><a href="#cb39-1284"></a><span class="st">  group1 = "0",</span></span>
<span id="cb39-1285"><a href="#cb39-1285"></a><span class="st">  group2 = "24",</span></span>
<span id="cb39-1286"><a href="#cb39-1286"></a><span class="st">  colors = c("#4d4d4d", </span></span>
<span id="cb39-1287"><a href="#cb39-1287"></a><span class="st">             #"#67a9cf", </span></span>
<span id="cb39-1288"><a href="#cb39-1288"></a><span class="st">             "#1c9099")) {</span></span>
<span id="cb39-1289"><a href="#cb39-1289"></a><span class="st">  </span></span>
<span id="cb39-1290"><a href="#cb39-1290"></a><span class="st">  # Adapted from:</span></span>
<span id="cb39-1291"><a href="#cb39-1291"></a><span class="st">  # &lt;https://rpubs.com/mharris/KSplot&gt;</span></span>
<span id="cb39-1292"><a href="#cb39-1292"></a><span class="st">  df &lt;- filter(df, df[[group.col]] </span><span class="sc">%i</span><span class="st">n</span><span class="sc">% c</span><span class="st">(group1, group2))</span></span>
<span id="cb39-1293"><a href="#cb39-1293"></a><span class="st">  </span></span>
<span id="cb39-1294"><a href="#cb39-1294"></a><span class="st">  df[df[[group.col]] == group1, data.col]</span></span>
<span id="cb39-1295"><a href="#cb39-1295"></a><span class="st">  </span></span>
<span id="cb39-1296"><a href="#cb39-1296"></a><span class="st">  </span></span>
<span id="cb39-1297"><a href="#cb39-1297"></a><span class="st">  data1 &lt;- unlist(df[df[[group.col]] == group1, data.col])</span></span>
<span id="cb39-1298"><a href="#cb39-1298"></a><span class="st">  data2 &lt;- unlist(df[df[[group.col]] == group2, data.col])</span></span>
<span id="cb39-1299"><a href="#cb39-1299"></a><span class="st">  </span></span>
<span id="cb39-1300"><a href="#cb39-1300"></a><span class="st">  ecdf1 &lt;- ecdf(data1)</span></span>
<span id="cb39-1301"><a href="#cb39-1301"></a><span class="st">  ecdf2 &lt;- ecdf(data2)</span></span>
<span id="cb39-1302"><a href="#cb39-1302"></a><span class="st">  </span></span>
<span id="cb39-1303"><a href="#cb39-1303"></a><span class="st">  # used to get the most extreme difference between the two samples</span></span>
<span id="cb39-1304"><a href="#cb39-1304"></a><span class="st">  MostExtremeDiff &lt;- seq(min(data1, data2, na.rm = T), max(data1, data2, na.rm = T), length.out = length(data1))</span></span>
<span id="cb39-1305"><a href="#cb39-1305"></a><span class="st">  x0 &lt;- MostExtremeDiff[which(abs(ecdf1(MostExtremeDiff) - ecdf2(MostExtremeDiff)) == </span></span>
<span id="cb39-1306"><a href="#cb39-1306"></a><span class="st">                                max(abs(ecdf1(MostExtremeDiff) - ecdf2(MostExtremeDiff))))]</span></span>
<span id="cb39-1307"><a href="#cb39-1307"></a><span class="st">  y0 &lt;- ecdf1(x0)</span></span>
<span id="cb39-1308"><a href="#cb39-1308"></a><span class="st">  y1 &lt;- ecdf2(x0)</span></span>
<span id="cb39-1309"><a href="#cb39-1309"></a><span class="st">  </span></span>
<span id="cb39-1310"><a href="#cb39-1310"></a><span class="st">  graph.df &lt;- data.frame(data1, data2) %&gt;</span><span class="sc">% g</span><span class="st">ather(Condition, Value, 1:2)</span></span>
<span id="cb39-1311"><a href="#cb39-1311"></a><span class="st">  graph.df[graph.df$Condition == "data1", "Condition"] &lt;- group1</span></span>
<span id="cb39-1312"><a href="#cb39-1312"></a><span class="st">  graph.df[graph.df$Condition == "data2", "Condition"] &lt;- group2</span></span>
<span id="cb39-1313"><a href="#cb39-1313"></a><span class="st">  </span></span>
<span id="cb39-1314"><a href="#cb39-1314"></a><span class="st">  # Run two sided KS test on data</span></span>
<span id="cb39-1315"><a href="#cb39-1315"></a><span class="st">  test.res &lt;- ks.test(data1, data2)</span></span>
<span id="cb39-1316"><a href="#cb39-1316"></a><span class="st">  </span></span>
<span id="cb39-1317"><a href="#cb39-1317"></a><span class="st">  plt &lt;- </span></span>
<span id="cb39-1318"><a href="#cb39-1318"></a><span class="st">    ggplot(graph.df)+</span></span>
<span id="cb39-1319"><a href="#cb39-1319"></a><span class="st">    geom_segment(aes(x = x0[1], y = y0[1], xend = x0[1], yend = y1[1]),</span></span>
<span id="cb39-1320"><a href="#cb39-1320"></a><span class="st">                 linetype = "dashed", color = "black", size = 1)+</span></span>
<span id="cb39-1321"><a href="#cb39-1321"></a><span class="st">    geom_point(aes(x = x0[1] , y= y0[1]), color="black", size=2) +</span></span>
<span id="cb39-1322"><a href="#cb39-1322"></a><span class="st">    geom_point(aes(x = x0[1] , y= y1[1]), color="black", size=2) +</span></span>
<span id="cb39-1323"><a href="#cb39-1323"></a><span class="st">    stat_ecdf(aes(x = Value, group = Condition, color = Condition))+</span></span>
<span id="cb39-1324"><a href="#cb39-1324"></a><span class="st">    labs(x = "Sample", </span></span>
<span id="cb39-1325"><a href="#cb39-1325"></a><span class="st">         y = "ECDF", </span></span>
<span id="cb39-1326"><a href="#cb39-1326"></a><span class="st">         title = paste("K-S Test", as.character(group1), "vs", as.character(group2), </span></span>
<span id="cb39-1327"><a href="#cb39-1327"></a><span class="st">                       "</span><span class="ch">\n</span><span class="st">p-value:", as.character(test.res$p.value, digits = 4)))+</span></span>
<span id="cb39-1328"><a href="#cb39-1328"></a><span class="st">    theme_minimal()+</span></span>
<span id="cb39-1329"><a href="#cb39-1329"></a><span class="st">    theme(legend.position = "bottom")+</span></span>
<span id="cb39-1330"><a href="#cb39-1330"></a><span class="st">    scale_color_manual(values = colors)#+</span></span>
<span id="cb39-1331"><a href="#cb39-1331"></a><span class="st">    # theme(text=element_text(family="Calibri Light", size=14)) </span></span>
<span id="cb39-1332"><a href="#cb39-1332"></a><span class="st">  </span></span>
<span id="cb39-1333"><a href="#cb39-1333"></a><span class="st">  return(plt)</span></span>
<span id="cb39-1334"><a href="#cb39-1334"></a><span class="st">  </span></span>
<span id="cb39-1335"><a href="#cb39-1335"></a><span class="st">}</span></span>
<span id="cb39-1336"><a href="#cb39-1336"></a><span class="st">```</span></span>
<span id="cb39-1337"><a href="#cb39-1337"></a></span>
<span id="cb39-1338"><a href="#cb39-1338"></a></span>
<span id="cb39-1339"><a href="#cb39-1339"></a><span class="st">``` r</span></span>
<span id="cb39-1340"><a href="#cb39-1340"></a></span>
<span id="cb39-1341"><a href="#cb39-1341"></a></span>
<span id="cb39-1342"><a href="#cb39-1342"></a><span class="st">2021-3-3</span></span>
<span id="cb39-1343"><a href="#cb39-1343"></a><span class="st">Daniel</span></span>
<span id="cb39-1344"><a href="#cb39-1344"></a><span class="st">Yeah I think this is from me. `shave`  is from `library(corrr)`  and achieves this nicely. I was thinking after the data's in long format, but the solution may just be to do it on the front end instead of after it's long.</span></span>
<span id="cb39-1345"><a href="#cb39-1345"></a></span>
<span id="cb39-1346"><a href="#cb39-1346"></a><span class="st">2021-3-3</span></span>
<span id="cb39-1347"><a href="#cb39-1347"></a><span class="st">David Schulz</span></span>
<span id="cb39-1348"><a href="#cb39-1348"></a><span class="st">So in other words I answered your question by facilitating a conversation between you and yourself. You're welcome.  :slightly_smiling_face:</span></span>
<span id="cb39-1349"><a href="#cb39-1349"></a></span>
<span id="cb39-1350"><a href="#cb39-1350"></a><span class="st">2021-3-3</span></span>
<span id="cb39-1351"><a href="#cb39-1351"></a><span class="st">Daniel</span></span>
<span id="cb39-1352"><a href="#cb39-1352"></a><span class="st">I appreciate it.</span></span>
<span id="cb39-1353"><a href="#cb39-1353"></a></span>
<span id="cb39-1354"><a href="#cb39-1354"></a><span class="st">For anyone using this in the future, `shave`  doesn't work with dataframes or the output of `cor`  so you'll need to run `as.data.frame` after shaving it.</span></span>
<span id="cb39-1355"><a href="#cb39-1355"></a></span>
<span id="cb39-1356"><a href="#cb39-1356"></a><span class="st">2021-3-3</span></span>
<span id="cb39-1357"><a href="#cb39-1357"></a><span class="st">Daniel</span></span>
<span id="cb39-1358"><a href="#cb39-1358"></a><span class="st">Found some code I wrote to do this following `cor`.</span></span>
<span id="cb39-1359"><a href="#cb39-1359"></a><span class="st">This is probably only worth doing this way if you're calculating p values (for corplot) since the same strategy can be applied to a `p.mat`.</span></span>
<span id="cb39-1360"><a href="#cb39-1360"></a></span>
<span id="cb39-1361"><a href="#cb39-1361"></a></span>
<span id="cb39-1362"><a href="#cb39-1362"></a><span class="st">```</span></span>
<span id="cb39-1363"><a href="#cb39-1363"></a><span class="st">  cor_df &lt;- cor(cor_df, use = "pairwise.complete.obs")</span></span>
<span id="cb39-1364"><a href="#cb39-1364"></a><span class="st">  </span></span>
<span id="cb39-1365"><a href="#cb39-1365"></a><span class="st">  # Shave to diagonal</span></span>
<span id="cb39-1366"><a href="#cb39-1366"></a><span class="st">  for (j in seq(1, nrow(cor_df))){</span></span>
<span id="cb39-1367"><a href="#cb39-1367"></a><span class="st">    cor_df[j, 1:j] &lt;- NA</span></span>
<span id="cb39-1368"><a href="#cb39-1368"></a><span class="st">  }</span></span>
<span id="cb39-1369"><a href="#cb39-1369"></a><span class="st">``` r</span></span>
<span id="cb39-1370"><a href="#cb39-1370"></a></span>
<span id="cb39-1371"><a href="#cb39-1371"></a></span>
<span id="cb39-1372"><a href="#cb39-1372"></a></span>
<span id="cb39-1373"><a href="#cb39-1373"></a></span>
<span id="cb39-1374"><a href="#cb39-1374"></a><span class="st">-xcutx-</span></span>
<span id="cb39-1375"><a href="#cb39-1375"></a></span>
<span id="cb39-1376"><a href="#cb39-1376"></a></span>
<span id="cb39-1377"><a href="#cb39-1377"></a><span class="st">2021-3-5</span></span>
<span id="cb39-1378"><a href="#cb39-1378"></a><span class="st">Two tricks today:</span></span>
<span id="cb39-1379"><a href="#cb39-1379"></a><span class="st">1. Scale fill functions can accept breaks and limit arguments so you don't have to use hacky workarounds like binning the data before plotting (which is what I usually do).</span></span>
<span id="cb39-1380"><a href="#cb39-1380"></a><span class="st">2. `library(ggtext)` lets you render markdown within plots (e.g. for those pesky mRNAs)</span></span>
<span id="cb39-1381"><a href="#cb39-1381"></a></span>
<span id="cb39-1382"><a href="#cb39-1382"></a><span class="st">```</span></span>
<span id="cb39-1383"><a href="#cb39-1383"></a><span class="st">&gt; df</span></span>
<span id="cb39-1384"><a href="#cb39-1384"></a><span class="st"># A tibble: 540 x 4</span></span>
<span id="cb39-1385"><a href="#cb39-1385"></a><span class="st">   term   Condition term2    Cor</span></span>
<span id="cb39-1386"><a href="#cb39-1386"></a><span class="st">   &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt;  &lt;dbl&gt;</span></span>
<span id="cb39-1387"><a href="#cb39-1387"></a><span class="st"> 1 bkkca  0h        vrest -0.420</span></span>
<span id="cb39-1388"><a href="#cb39-1388"></a><span class="st"> 2 cav1   0h        vrest -0.252</span></span>
<span id="cb39-1389"><a href="#cb39-1389"></a><span class="st"> 3 cav2   0h        vrest -0.276</span></span>
<span id="cb39-1390"><a href="#cb39-1390"></a><span class="st"> 4 inx1   0h        vrest -0.313</span></span>
<span id="cb39-1391"><a href="#cb39-1391"></a><span class="st"> 5 inx2   0h        vrest -0.327</span></span>
<span id="cb39-1392"><a href="#cb39-1392"></a><span class="st">``` r</span></span>
<span id="cb39-1393"><a href="#cb39-1393"></a></span>
<span id="cb39-1394"><a href="#cb39-1394"></a><span class="st">Before:</span></span>
<span id="cb39-1395"><a href="#cb39-1395"></a></span>
<span id="cb39-1396"><a href="#cb39-1396"></a><span class="st">```</span></span>
<span id="cb39-1397"><a href="#cb39-1397"></a><span class="st">df %&gt;% </span></span>
<span id="cb39-1398"><a href="#cb39-1398"></a><span class="st">  ggplot(aes(term, term2, fill = Cor))+</span></span>
<span id="cb39-1399"><a href="#cb39-1399"></a><span class="st">  geom_tile()+</span></span>
<span id="cb39-1400"><a href="#cb39-1400"></a><span class="st">  labs(x= "mRNA", y = "")+</span></span>
<span id="cb39-1401"><a href="#cb39-1401"></a><span class="st">  scale_fill_distiller(palette = "PuOr")+</span></span>
<span id="cb39-1402"><a href="#cb39-1402"></a><span class="st">  coord_fixed()</span></span>
<span id="cb39-1403"><a href="#cb39-1403"></a><span class="st">``` r</span></span>
<span id="cb39-1404"><a href="#cb39-1404"></a></span>
<span id="cb39-1405"><a href="#cb39-1405"></a><span class="st">After:</span></span>
<span id="cb39-1406"><a href="#cb39-1406"></a></span>
<span id="cb39-1407"><a href="#cb39-1407"></a><span class="st">```</span></span>
<span id="cb39-1408"><a href="#cb39-1408"></a><span class="st">library(ggtext)</span></span>
<span id="cb39-1409"><a href="#cb39-1409"></a><span class="st">library(glue)</span></span>
<span id="cb39-1410"><a href="#cb39-1410"></a></span>
<span id="cb39-1411"><a href="#cb39-1411"></a><span class="st">df %&gt;% </span></span>
<span id="cb39-1412"><a href="#cb39-1412"></a><span class="st">  mutate(term = glue(("&lt;i&gt;</span><span class="sc">{term}</span><span class="st">&lt;/i&gt;"))) %&gt;%  </span></span>
<span id="cb39-1413"><a href="#cb39-1413"></a><span class="st">  ggplot(aes(term, term2, fill = Cor))+</span></span>
<span id="cb39-1414"><a href="#cb39-1414"></a><span class="st">  geom_tile()+</span></span>
<span id="cb39-1415"><a href="#cb39-1415"></a><span class="st">  labs(x= "mRNA", y = "")+</span></span>
<span id="cb39-1416"><a href="#cb39-1416"></a><span class="st">  </span></span>
<span id="cb39-1417"><a href="#cb39-1417"></a><span class="st">  theme(axis.text.x = element_markdown(angle = 45))+ # &lt;-- Note that we have element_markdown not element_text</span></span>
<span id="cb39-1418"><a href="#cb39-1418"></a><span class="st">  </span></span>
<span id="cb39-1419"><a href="#cb39-1419"></a><span class="st">  scale_fill_stepsn(</span></span>
<span id="cb39-1420"><a href="#cb39-1420"></a><span class="st">    colors=RColorBrewer::brewer.pal(n = 8, name = "PuOr"),</span></span>
<span id="cb39-1421"><a href="#cb39-1421"></a><span class="st">    na.value = "transparent",</span></span>
<span id="cb39-1422"><a href="#cb39-1422"></a><span class="st">    breaks=round(seq(-1, 1, length.out = 8), digits = 2),</span></span>
<span id="cb39-1423"><a href="#cb39-1423"></a><span class="st">    limits=c(-1,1)</span></span>
<span id="cb39-1424"><a href="#cb39-1424"></a><span class="st">  )+</span></span>
<span id="cb39-1425"><a href="#cb39-1425"></a><span class="st">  coord_fixed()</span></span>
<span id="cb39-1426"><a href="#cb39-1426"></a><span class="st">``` r</span></span>
<span id="cb39-1427"><a href="#cb39-1427"></a></span>
<span id="cb39-1428"><a href="#cb39-1428"></a><span class="st">![image (28).png](image%20(28).png)</span></span>
<span id="cb39-1429"><a href="#cb39-1429"></a><span class="st">![image (29).png](image%20(29).png)</span></span>
<span id="cb39-1430"><a href="#cb39-1430"></a></span>
<span id="cb39-1431"><a href="#cb39-1431"></a><span class="st">2021-3-5</span></span>
<span id="cb39-1432"><a href="#cb39-1432"></a><span class="st">Daniel</span></span>
<span id="cb39-1433"><a href="#cb39-1433"></a><span class="st">&lt;https://github.com/wilkelab/ggtext&gt;</span></span>
<span id="cb39-1434"><a href="#cb39-1434"></a></span>
<span id="cb39-1435"><a href="#cb39-1435"></a></span>
<span id="cb39-1436"><a href="#cb39-1436"></a><span class="st">-xcutx-</span></span>
<span id="cb39-1437"><a href="#cb39-1437"></a></span>
<span id="cb39-1438"><a href="#cb39-1438"></a></span>
<span id="cb39-1439"><a href="#cb39-1439"></a><span class="st">2021-3-9</span></span>
<span id="cb39-1440"><a href="#cb39-1440"></a><span class="st">Another useful trick is to pass expressions into ggplot. Here I've used the following as arguments in `labs()`.</span></span>
<span id="cb39-1441"><a href="#cb39-1441"></a></span>
<span id="cb39-1442"><a href="#cb39-1442"></a></span>
<span id="cb39-1443"><a href="#cb39-1443"></a><span class="st">```</span></span>
<span id="cb39-1444"><a href="#cb39-1444"></a><span class="st">c("r11", "r1", "Ihtk.0", "Ihtk.Slope", "Ia.0", "Ia.Slope", "vrest")</span></span>
<span id="cb39-1445"><a href="#cb39-1445"></a><span class="st">c(expression(M~Omega), expression(M~Omega), "nA", expression(frac(nA, mV)), "nA", expression(frac(nA, mV)), "mV" )</span></span>
<span id="cb39-1446"><a href="#cb39-1446"></a><span class="st">``` r</span></span>
<span id="cb39-1447"><a href="#cb39-1447"></a></span>
<span id="cb39-1448"><a href="#cb39-1448"></a></span>
<span id="cb39-1449"><a href="#cb39-1449"></a><span class="st">![image (30).png](image%20(30).png)</span></span>
<span id="cb39-1450"><a href="#cb39-1450"></a></span>
<span id="cb39-1451"><a href="#cb39-1451"></a><span class="st">2021-3-9</span></span>
<span id="cb39-1452"><a href="#cb39-1452"></a><span class="st">Daniel</span></span>
<span id="cb39-1453"><a href="#cb39-1453"></a><span class="st">You can also do something like this `theme(plot.title = element_text(face="italic"))`</span></span>
<span id="cb39-1454"><a href="#cb39-1454"></a></span>
<span id="cb39-1455"><a href="#cb39-1455"></a></span>
<span id="cb39-1456"><a href="#cb39-1456"></a><span class="st">---</span></span>
<span id="cb39-1457"><a href="#cb39-1457"></a></span>
<span id="cb39-1458"><a href="#cb39-1458"></a></span>
<span id="cb39-1459"><a href="#cb39-1459"></a><span class="st">2021-3-11</span></span>
<span id="cb39-1460"><a href="#cb39-1460"></a><span class="st">Reordering a discrete axis in ggplot after generation a lot simpler than one might expect.  Rather than converting a character column to a factor (what if the data gets pivoted?), or using one column for the position and one for the labels, you can use `xlim` or `ylim`.</span></span>
<span id="cb39-1461"><a href="#cb39-1461"></a></span>
<span id="cb39-1462"><a href="#cb39-1462"></a><span class="st">```</span></span>
<span id="cb39-1463"><a href="#cb39-1463"></a><span class="st">&gt; mrna_cols # desired order</span></span>
<span id="cb39-1464"><a href="#cb39-1464"></a><span class="st"># [1] "nav"    "cav1"   "cav2"   "bkkca"  "shaker" "shal"   "shab"   "shaw1" </span></span>
<span id="cb39-1465"><a href="#cb39-1465"></a><span class="st"># [9] "shaw2"  "inx1"   "inx2"   "inx3"  </span></span>
<span id="cb39-1466"><a href="#cb39-1466"></a></span>
<span id="cb39-1467"><a href="#cb39-1467"></a><span class="st">o_mrna$heatmap_z / # ggplot object within a list</span></span>
<span id="cb39-1468"><a href="#cb39-1468"></a><span class="st">o_mrna$heatmap_z+ylim(mrna_cols)</span></span>
<span id="cb39-1469"><a href="#cb39-1469"></a><span class="st">``` r</span></span>
<span id="cb39-1470"><a href="#cb39-1470"></a></span>
<span id="cb39-1471"><a href="#cb39-1471"></a></span>
<span id="cb39-1472"><a href="#cb39-1472"></a><span class="st">![image (31).png](image%20(31).png)</span></span>
<span id="cb39-1473"><a href="#cb39-1473"></a></span>
<span id="cb39-1474"><a href="#cb39-1474"></a></span>
<span id="cb39-1475"><a href="#cb39-1475"></a><span class="st">-xcutx-</span></span>
<span id="cb39-1476"><a href="#cb39-1476"></a></span>
<span id="cb39-1477"><a href="#cb39-1477"></a></span>
<span id="cb39-1478"><a href="#cb39-1478"></a><span class="st">2021-3-23</span></span>
<span id="cb39-1479"><a href="#cb39-1479"></a><span class="st">R's distribution simulation functions (e.g. `dbinom`, `runif`) make it quick and easy to double check one's intuitions. For example, I'd been thinking that under H0 the distribution of correlations from normal samples should drop off sharply as you go away from 0 such that a shift in correlation from 0 -&gt; 0.1 is much more likely than 0.8 -&gt; 0.9.</span></span>
<span id="cb39-1480"><a href="#cb39-1480"></a></span>
<span id="cb39-1481"><a href="#cb39-1481"></a></span>
<span id="cb39-1482"><a href="#cb39-1482"></a><span class="st">So I used `purrr::map()` to run a quick simulation. Here we simulate the null distribution based on 100,000 observations and compute the chance of a value being above 0.7. If it was uniform we would expect ~15% (.03/2) of the distribution to be here but end up with ~1.2% with the drop off.</span></span>
<span id="cb39-1483"><a href="#cb39-1483"></a></span>
<span id="cb39-1484"><a href="#cb39-1484"></a><span class="st">```</span></span>
<span id="cb39-1485"><a href="#cb39-1485"></a><span class="st">set.seed(89745)</span></span>
<span id="cb39-1486"><a href="#cb39-1486"></a><span class="st">cor_check &lt;- map(1:100000, function(i){ </span></span>
<span id="cb39-1487"><a href="#cb39-1487"></a><span class="st">  cor(rnorm(10), rnorm(10), method = "pearson")</span></span>
<span id="cb39-1488"><a href="#cb39-1488"></a><span class="st">})</span></span>
<span id="cb39-1489"><a href="#cb39-1489"></a><span class="st">cor_check &lt;- data.frame(cor = do.call(rbind,  cor_check))</span></span>
<span id="cb39-1490"><a href="#cb39-1490"></a><span class="st">ggplot(cor_check, aes(x = cor))+  geom_histogram(binwidth = 0.05)</span></span>
<span id="cb39-1491"><a href="#cb39-1491"></a><span class="st">mean(cor_check$cor &gt;= 0.7)*100</span></span>
<span id="cb39-1492"><a href="#cb39-1492"></a><span class="st"># 1.227 Percent</span></span>
<span id="cb39-1493"><a href="#cb39-1493"></a><span class="st">``` r</span></span>
<span id="cb39-1494"><a href="#cb39-1494"></a></span>
<span id="cb39-1495"><a href="#cb39-1495"></a></span>
<span id="cb39-1496"><a href="#cb39-1496"></a><span class="st">![image (32).png](image%20(32).png)</span></span>
<span id="cb39-1497"><a href="#cb39-1497"></a></span>
<span id="cb39-1498"><a href="#cb39-1498"></a></span>
<span id="cb39-1499"><a href="#cb39-1499"></a><span class="st">-xcutx-</span></span>
<span id="cb39-1500"><a href="#cb39-1500"></a></span>
<span id="cb39-1501"><a href="#cb39-1501"></a></span>
<span id="cb39-1502"><a href="#cb39-1502"></a><span class="st">2021-9-7</span></span>
<span id="cb39-1503"><a href="#cb39-1503"></a><span class="st">Daniel</span></span>
<span id="cb39-1504"><a href="#cb39-1504"></a><span class="st">&lt;https://github.com/seatgeek/fuzzywuzzy|https://github.com/seatgeek/fuzzywuzzy&gt; this is a tool that isn’t necessary most of the time but when it is it can save a ton of time. It lets you do approximate string matching. I’ve used it for handling typos and differences in white space/ punctuation/ naming conventions in entry labels and it’s worked nicely. There’s a port for R and a few other languages too. </span></span>
<span id="cb39-1505"><a href="#cb39-1505"></a></span>
<span id="cb39-1506"><a href="#cb39-1506"></a><span class="st">2021-9-7</span></span>
<span id="cb39-1507"><a href="#cb39-1507"></a><span class="st">Joe</span></span>
<span id="cb39-1508"><a href="#cb39-1508"></a><span class="st">Thanks zombie Daniel </span></span>
<span id="cb39-1509"><a href="#cb39-1509"></a></span>
<span id="cb39-1510"><a href="#cb39-1510"></a></span>
<span id="cb39-1511"><a href="#cb39-1511"></a><span class="st">-xcutx-</span></span>
<span id="cb39-1512"><a href="#cb39-1512"></a></span>
<span id="cb39-1513"><a href="#cb39-1513"></a></span>
<span id="cb39-1514"><a href="#cb39-1514"></a><span class="st">2022-5-31</span></span>
<span id="cb39-1515"><a href="#cb39-1515"></a><span class="st">Daniel</span></span>
<span id="cb39-1516"><a href="#cb39-1516"></a><span class="st">ggplot tip: if you use an eight digit hex code for specifying a color value, the first two control transparency. Thus, you can set `fill` `=` `“#00000000”` in ggplot to get a boxplot with no fill.</span></span>
<span id="cb39-1517"><a href="#cb39-1517"></a></span>
<span id="cb39-1518"><a href="#cb39-1518"></a><span class="st">2022-5-31</span></span>
<span id="cb39-1519"><a href="#cb39-1519"></a><span class="st">Here's the use case: points on top occlude the cross bar but box on top hides the observations. You could also do this by changing alpha but I think that will alter the lines as well.</span></span>
<span id="cb39-1520"><a href="#cb39-1520"></a><span class="st">![Image from iOS](Image from iOS.jpg)</span></span>
<span id="cb39-1521"><a href="#cb39-1521"></a></span>
<span id="cb39-1522"><a href="#cb39-1522"></a></span>
<span id="cb39-1523"><a href="#cb39-1523"></a></span>
<span id="cb39-1524"><a href="#cb39-1524"></a><span class="st">-xcutx-</span></span>
<span id="cb39-1525"><a href="#cb39-1525"></a></span>
<span id="cb39-1526"><a href="#cb39-1526"></a></span>
<span id="cb39-1527"><a href="#cb39-1527"></a><span class="st">2023-5-23</span></span>
<span id="cb39-1528"><a href="#cb39-1528"></a><span class="st">Daniel</span></span>
<span id="cb39-1529"><a href="#cb39-1529"></a><span class="st">R treats # as a comment in text files. </span></span>
<span id="cb39-1530"><a href="#cb39-1530"></a></span>
<span id="cb39-1531"><a href="#cb39-1531"></a><span class="st">A file I have (hapmap of snps) included one in a field of the header. If your look at the words on the first to lines with wc they will agree but R will fail to load the file. The solution is to pass in an explicit comment character like so `read.table(“snps.txt”, comment.char=‘’)`. </span></span>
<span id="cb39-1532"><a href="#cb39-1532"></a></span>
<span id="cb39-1533"><a href="#cb39-1533"></a></span>
<span id="cb39-1534"><a href="#cb39-1534"></a><span class="st">-xcutx-</span></span>
<span id="cb39-1535"><a href="#cb39-1535"></a></span>
<span id="cb39-1536"><a href="#cb39-1536"></a></span>
<span id="cb39-1537"><a href="#cb39-1537"></a><span class="st">2024-3-21</span></span>
<span id="cb39-1538"><a href="#cb39-1538"></a><span class="st">Daniel</span></span>
<span id="cb39-1539"><a href="#cb39-1539"></a><span class="st">&lt;https://blog.moertel.com/posts/2006-01-20-wondrous-oddities-rs-function-call-semantics.html|https://blog.moertel.com/posts/2006-01-20-wondrous-oddities-rs-function-call-semantics.html&gt;</span></span>
<span id="cb39-1540"><a href="#cb39-1540"></a></span>
<span id="cb39-1541"><a href="#cb39-1541"></a><span class="st">This is the sort of thing you don’t realize until it would be _really_ useful to access the name of a variable or run text as if it were code. I think the most accessible example of R’s wizardry is in plotting- you pass variables (time, mv) to plot or ggplot instead of strings (“time”, “mv”) and magically you get axis labels. R gets access to the _value_ of a variable _*and*_ its name and no one notices because it just works. :exploding_head:  ```</span></span>
<span id="cb39-1542"><a href="#cb39-1542"></a><span class="st">"""</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="5efbb912" class="cell" data-execution_count="32">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1"></a>posts <span class="op">=</span> [e.strip(<span class="st">'</span><span class="ch">\n</span><span class="st">'</span>) <span class="cf">for</span> e <span class="kw">in</span> all_text.split(<span class="st">'-xcutx-'</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="5d88a4cc" class="cell" data-execution_count="64">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1"></a>post <span class="op">=</span> posts[<span class="dv">0</span>]</span>
<span id="cb41-2"><a href="#cb41-2"></a></span>
<span id="cb41-3"><a href="#cb41-3"></a>out <span class="op">=</span> []</span>
<span id="cb41-4"><a href="#cb41-4"></a></span>
<span id="cb41-5"><a href="#cb41-5"></a><span class="cf">for</span> post <span class="kw">in</span> posts:</span>
<span id="cb41-6"><a href="#cb41-6"></a>  _ <span class="op">=</span> post.split(<span class="st">'</span><span class="ch">\n</span><span class="st">'</span>)</span>
<span id="cb41-7"><a href="#cb41-7"></a>  ts <span class="op">=</span> _[<span class="dv">0</span>]</span>
<span id="cb41-8"><a href="#cb41-8"></a>  res <span class="op">=</span> <span class="ss">f"""---</span></span>
<span id="cb41-9"><a href="#cb41-9"></a><span class="ss">title: "Placeholder"</span></span>
<span id="cb41-10"><a href="#cb41-10"></a><span class="ss">author: "Daniel Kick"</span></span>
<span id="cb41-11"><a href="#cb41-11"></a><span class="ss">date: </span><span class="sc">{</span>ts<span class="sc">}</span></span>
<span id="cb41-12"><a href="#cb41-12"></a><span class="ss">image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"</span></span>
<span id="cb41-13"><a href="#cb41-13"></a><span class="ss">categories: </span></span>
<span id="cb41-14"><a href="#cb41-14"></a><span class="ss">  - code</span></span>
<span id="cb41-15"><a href="#cb41-15"></a><span class="ss">  - beginner</span></span>
<span id="cb41-16"><a href="#cb41-16"></a><span class="ss">  - r</span></span>
<span id="cb41-17"><a href="#cb41-17"></a><span class="ss">draft: false</span></span>
<span id="cb41-18"><a href="#cb41-18"></a><span class="ss">editor: </span></span>
<span id="cb41-19"><a href="#cb41-19"></a><span class="ss">  markdown: </span></span>
<span id="cb41-20"><a href="#cb41-20"></a><span class="ss">    wrap: 72</span></span>
<span id="cb41-21"><a href="#cb41-21"></a><span class="ss">---</span></span>
<span id="cb41-22"><a href="#cb41-22"></a><span class="ss">  """</span><span class="op">+</span><span class="st">'</span><span class="ch">\n</span><span class="st">'</span>.join(_[<span class="dv">2</span>:] <span class="cf">if</span> _[<span class="dv">1</span>] <span class="op">==</span> <span class="st">'Daniel'</span> <span class="cf">else</span> _[<span class="dv">1</span>:])</span>
<span id="cb41-23"><a href="#cb41-23"></a>  out.append(res)</span>
<span id="cb41-24"><a href="#cb41-24"></a></span>
<span id="cb41-25"><a href="#cb41-25"></a><span class="bu">print</span>(<span class="st">'</span><span class="ch">\n\n</span><span class="st">'</span>.join(out))</span>
<span id="cb41-26"><a href="#cb41-26"></a></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>---
title: "Placeholder"
author: "Daniel Kick"
date: 2020-4-1
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  You can keep your working `Rmd` easier to navigate and less buggy by 1) packaging code into functions and 2) adding them to a companion `R` file. Load your functions with `source()` in the same block you load your libraries with a relative path, full path, or ideally with `here()`.

``` r
library(here)
source(here("R", "02MoniterGapJunction.R")) #here's output is effectively ../R/02MoniterGapJunction.R
```

---
title: "Placeholder"
author: "Daniel Kick"
date: 2020-4-10
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  If you're iteratively making plots, resampling, or doing another task that would have your reach for a for loop or `lapply()` use `furrr::future_map()` instead. `furrr` gives parallel processing ready versions of tidyverse's `purrr` functions (e.g. `map()`,  `walk()`). It's easy to install the dependencies and takes a lot of the headache out of parallel processing.
![cpu_use_1.PNG](cpu_use_1.PNG)
![cpu_use_0.PNG](cpu_use_0.PNG)

---
title: "Placeholder"
author: "Daniel Kick"
date: 2020-4-20
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  If you're running low on memory, you can retain specific objects by running this:

``` r
 
# get rid of everything
rm(list=ls())

# get rid of everything except specific objects and all loaded functions
rm(list = 
   ls()[!(ls() %in% c(
   # objects
   c("data1", "data2", "bool1", "list1"), 
   # functions
   lsf.str())
   ) ])
```

---
title: "Placeholder"
author: "Daniel Kick"
date: 2020-4-20
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  `lsf.str()` returns a character string of the functions you've written. This doesnt' include functions from libraries so it may not be needed if you have used `source()` to pull your custom functions from `/R/`.

2020-4-20
If you're tempted to use `rm(list = ls())` consider restarting your r session (`ctrl+shift+F10` on windows). Overreliance on `rm(list = ls())` is poor form and the Rstudio devs will know and find you.
![image.png](image.png)

---
title: "Placeholder"
author: "Daniel Kick"
date: 2020-4-21
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  If you're applying the same theme to all your graphs, set it globally instead e.g. `theme_set(ggplot2::theme_minimal())`.  If you have a lot of custom changes to your theme, throw those into a function and set that to the global theme.

---
title: "Placeholder"
author: "Daniel Kick"
date: 2020-5-4
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  If you're on windows, `installr` should allow you to copy over the libraries from previous versions.
![image (1).png](image%20(1).png)

---
title: "Placeholder"
author: "Daniel Kick"
date: 2020-5-6
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  Don't get suckered when converting factors! Numeric data can be assigned to a factor type which will throw a wrench in a plot or analysis (1/3)
![image (2).png](image%20(2).png)

2020-5-6
A knee jerk reaction would be to convert it to a numeric with `as.numeric()`. That doesn't work either.  (2/3)
![image (3).png](image%20(3).png)

2020-5-6
However if you use `as.numeric(as.character())` then it works. That's because factors are ordinal and named so if you convert the type to character first to ensure R is working with the factor names instead of the ranks. (3/3)
![image (4).png](image%20(4).png)

---
title: "Placeholder"
author: "Daniel Kick"
date: 2020-5-11
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  Need a quick base `R` syntax lookup? Check out &lt;https://learnxinyminutes.com/docs/r/&gt; .  There are even examples with `lm()` and `glm()`.

---
title: "Placeholder"
author: "Daniel Kick"
date: 2020-5-11
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  tldr: `tidyverse` function variants take slightly different input. Testing out a variant or two can save you a lot of debugging time.

In `tidyverse` watch out for inconsistencies in function versions. There are variants of common functions (e.g. `mutate()` `mutate_all()` `mutate_at()`) don't necessarily behave the same way (or how you would expect).

Here, I was applying an operation to a grouped df where each `Experiment` contains several `FileName`s with multiple observations in each.  To keep everything reusable I'm using `exp` instead of `Experiment` to select the right col.

``` r
&gt; exp = "Experiment"
&gt; rec = "FileName"

&gt; df %&gt;%
+     dplyr::select(
+       exp, rec, r11, r12
+     )
# A tibble: 564 x 4
   Experiment FileName           r11   r12
   &lt;chr&gt;      &lt;chr&gt;            &lt;dbl&gt; &lt;dbl&gt;
 1 190808a    190808a_0020.abf  5.06 0.150
 2 190808a    190808a_0020.abf  5.13 0.160
 3 190808a    190808a_0020.abf  5.11 0.122
 4 190808a    190808a_0020.abf  2.49 0.152
 5 190808a    190808a_0020.abf  2.49 0.195
# ... with 559 more rows
```

As soon as we do the same thing with `group_by()` we don't get the right column even though `select()` didn't have an issue with `exp`.

``` r
&gt; df %&gt;%
+     dplyr::select(
+       exp, rec, r11, r12
+     ) %&gt;%
+     group_by(exp, rec)
Error: Column `exp` is unknown
```

So we can try explicitly selecting the columns we want as groupings.

``` r
&gt;     df %&gt;%
+     dplyr::select(
+       exp, rec, r11, r12
+     ) %&gt;%
+     group_by(vars(exp, rec))
Error: Column `vars(exp, rec)` must be length 564 (the number of rows) or one, not 2
```

No dice there. `vars()` is designed to work with the `_at` variants so we can try that.  _et voilà!_

``` r
&gt; df %&gt;%
+     dplyr::select(
+       exp, rec, r11, r12
+     ) %&gt;%
+     group_by_at(vars(exp, rec))

# A tibble: 564 x 4
# Groups:   Experiment, FileName [69]
   Experiment FileName           r11   r12
   &lt;chr&gt;      &lt;chr&gt;            &lt;dbl&gt; &lt;dbl&gt;
 1 190808a    190808a_0020.abf  5.06 0.150
 2 190808a    190808a_0020.abf  5.13 0.160
 3 190808a    190808a_0020.abf  5.11 0.122
 4 190808a    190808a_0020.abf  2.49 0.152
 5 190808a    190808a_0020.abf  2.49 0.195
# ... with 559 more rows
```

---
title: "Placeholder"
author: "Daniel Kick"
date: 2020-5-12
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  Save `R` objects as rds. These could be works in progress or items that take a while to generate (e.g. re-sampled results). This gives the utility of saving a work space without the dangers.

``` r
library(here)
save(df, file = here("data", "df.rds"))
# df is now saved at ./data/df.rds

load(here("data", "df.rds"))
# df is now loaded from ./data/df.rds as df
```

---
title: "Placeholder"
author: "Daniel Kick"
date: 2020-5-15
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  If you're using git for version control, don't forget about the `.gitignore` file. Anything large and static (like .abfs) or procedurally generated (e.g. plots) you can toss in the `.gitignore` and you'll not see it when you commit.

I have tabular data that lives in `./inst/extdata/` is processed by a script and then saved as a .rds in `./data/`.  Here's my `.gitignore`.


``` r
.Rproj.user
.Rhistory
.RData
.Ruserdata
# Don't track ABFs -- large and static
*.abf
# Don't track files that are generated from the scripts
/data/*
```

For more check out &lt;https://git-scm.com/docs/gitignore&gt;

---
title: "Placeholder"
author: "Daniel Kick"
date: 2020-5-19
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  If you need to include greek letters or special characters in a plot use `annotate()` instead of `geom_text()`. On my machine it preforms a *lot* faster.


``` r
# 57.62977 secs
geom_text(aes(x = 15, y = 25, label = "phi~22.5"), parse=TRUE)
# 1.988642 secs
annotate("text", x = 15, y = 25, parse = TRUE, label = as.character(expression(paste(phi, " 22.5"))))
```

---
title: "Placeholder"
author: "Daniel Kick"
date: 2020-5-19
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  Use `geom_rect()` with the min/max set to `-Inf` and `Inf` to add a pleasant shading to your facets.

For example by passing it a data frame with the faceting variables and a column to use for the color (green if positive, red if negative) we can make facets behave like cells in  a heatmap!

``` r
# &gt; tp2
# # A tibble: 23 x 5
#    Condition   Trace             Time  Change StimId
#    &lt;fct&gt;       &lt;chr&gt;             &lt;chr&gt;  &lt;dbl&gt; &lt;chr&gt; 
#  1 PS.0.orig   170623b_0007.abf  60         1 a     
#  2 PS.45.orig  170825a_0008.abf  60        -1 c     

ggplot(df)+
  geom_rect(data = tp2, aes(fill = Change),xmin = -Inf,xmax = Inf, ymin = -Inf,ymax = Inf,alpha = 0.3) +
  geom_hline(yintercept = 0, color = "cornflowerblue")+
  geom_pointline(aes(x = Time, y = rc, group = Experiment), shape = 1, color = "black")+
  ylim(-2, 3.5)+
  scale_fill_gradientn(colors = c("Red", "Grey", "Green"))+
  theme_base()+
  theme(legend.position = "")
```

![image (5).png](image%20(5).png)

---
title: "Placeholder"
author: "Daniel Kick"
date: 2020-5-26
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  Handy trick:
Write your test results into a summary csv. Then while writing you can insert a link to the cell value for a result into your .doc. After that, changing post hoc corrections, re-sampling iterations, or data QC just requires you to re-run your code and let word refresh all the links.

You can also do this from excel, but you'll have to update the formula for the cell used instead.

&lt;#C010AS4PSUX|data_analysis&gt;
![AutoUpdateStats.gif](AutoUpdateStats.gif)

---
title: "Placeholder"
author: "Daniel Kick"
date: 2020-6-4
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  If you're looking at linear relationships, give this code a spin. It'll automatically flag outliers based on 1.5*IQR and show you the fit with and without the outliers.  Since the output is a ggplot, you and add to it or tweak the aesthetics of the output (see example below). If anyone has a favorite outlier detection method and ends up modifying this, please post it here so we can see it! &lt;@U0103JSQYLB&gt; &lt;@U010CH4JLKE&gt; &lt;@U010M1M3N9E&gt;

Here's the function.

``` r
scatter_w_wo_outliers &lt;- function(temp = filter(M, Time == "Baseline"),
                                  X = "bkkca", 
                                  Y = "Ihtk.0"){
  
  # flag outliers based on 1.5xIQR from median
  X_low  &lt;- median(temp[[X]], na.rm = T) - (1.5 * IQR(temp[[X]], na.rm = T))
  X_high &lt;- median(temp[[X]], na.rm = T) + (1.5 * IQR(temp[[X]], na.rm = T))
  Y_low  &lt;- median(temp[[Y]], na.rm = T) - (1.5 * IQR(temp[[Y]], na.rm = T))
  Y_high &lt;- median(temp[[Y]], na.rm = T) + (1.5 * IQR(temp[[Y]], na.rm = T))
  
  X_pass &lt;- (temp[[X]] &gt; X_low) * (temp[[X]] &lt; X_high)
  Y_pass &lt;- (temp[[Y]] &gt; Y_low) * (temp[[Y]] &lt; Y_high)

  temp$flag &lt;- ifelse((X_pass * Y_pass) == 1, T, F)
  
  
  # Duplicate so we have dataset 1, 2 (introduces duplicates)
  temp &lt;- rbind(temp[temp$flag == T, ], mutate(temp, flag = F))
  
  formula1 &lt;- y ~ x
  
  plt &lt;- ggplot(temp, aes_string(X, Y, color = "flag"))+
    geom_smooth(data = temp, method = lm, se = F, fullrange = T)+
    geom_point(data = temp)+
    geom_point(data = temp, color = "black", shape = 1)+
    geom_point(data = temp[temp$flag, ])+
    ggpmisc::stat_poly_eq(aes(label =  paste(stat(eq.label), "*" with "*", 
                                             stat(rr.label), "*", "*", 
                                             stat(f.value.label), "*", and "*",
                                             stat(p.value.label), "*"."",
                                             sep = "")),
                          formula = formula1, parse = TRUE, size = 4)+
    
    scale_color_manual(values = c("darkgray", "black"))+
    theme_bw()+
    theme(legend.position = "")
  
  return(plt)
}
```

Here's a reproducible example. We're creating a Simpson's paradox by giving the "outliers" a negative slope and the real data a positive slope. I've added a red line showing the true relationship.

``` r
set.seed(45645684)
df &lt;- data.frame(x = rnorm(30, mean = 10, sd = 4),
                 noise = runif(30, min = -2, max = 2),
                 y = NA,
                 is_outlier = rbinom(30, 1, prob = 0.2))


df$y &lt;- ifelse(df$is_outlier, 
               -5*df$x+df$noise,
               2*df$x+df$noise)

scatter_w_wo_outliers(temp = df,
                      X = "x",
                      Y = "y")+
  geom_abline(slope = 2, intercept = 0, color = "firebrick")
```

![image (7).png](image%20(7).png)


2020-6-4
Daniel
Here's a related utility function. For a given column it'll return a logical vector where outliers are `FALSE`.

``` r
w_in_x_iqr &lt;- function(col_in, multiplier = 1.5){
  col_in &lt;- as.vector(col_in)
  
  X_low  &lt;- median(col_in, na.rm = T) - (multiplier * IQR(col_in, na.rm = T))
  X_high &lt;- median(col_in, na.rm = T) + (multiplier * IQR(col_in, na.rm = T))
  X_pass &lt;- (col_in &gt; X_low) * (col_in &lt; X_high)
  
  return(as.logical(X_pass))
}
```

e.g.

``` r
&gt; mutate(M, ex = w_in_x_iqr(bkkca)) %&gt;% select(bkkca, ex) %&gt;% arrange(bkkca) %&gt;% tail()

#  A tibble: 6 x 2
#   bkkca ex   
#   &lt;dbl&gt; &lt;lgl&gt;
# 1 3056. TRUE 
# 2 3222. TRUE 
# 3 3255. TRUE   # Within bounds
# 4 3552. FALSE  # Outside bounds 
# 5 3817. FALSE
# 6 6740. FALSE
```

---
title: "Placeholder"
author: "Daniel Kick"
date: 2020-6-5
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  ggplot's, `element_text()` comes with a color argument. Why does that matter? It doesn't just accept atomics, you can hand it a _vector_! This effectively gives one access to conditional formatting. This works on `ggplot2_3.3.0` but "vectorized input to `element_text()` is not officially supported" so YMMV with newer versions.

Here's an example I think it makes an otherwise unbearable figure a little more so without requiring duplicated labels.

One thing to be aware of is the ordering of a character/ factor may differ between a data.frame and the plot. In the visualized example, I had `mRNA` as type character instead of factor so it got alphabetized when it was plotted rather than by the order appearing in the data.frame.
![image (8).png](image%20(8).png)

---
title: "Placeholder"
author: "Daniel Kick"
date: 2020-6-12
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  Trying something new here-- "Function Friday". If you've written a function you're rather pleased with, used one in a creative way, or know one that you think deserves it's time in the sun please share it here.

I'll start. Y'all should be using `unlist()`. `unlist` is a crazy handy function when you pair it with the `purrr` library. It'll take a list and try to give you a vector. Strictly speaking it's probably best to be using `map_dbl()` or `map_chr()` instead but let's not worry about that at the moment.

Why do I love `unlist` so much? Because `unlist(map( ))` gives you a flexible, effective way to iterate (that is parallel friendly with minor changes).

Here's an example: I have a bunch of traces in a folder and I need to know if there are any experiments that didn't copy. Additionally, I'd like to know what kind of data is in the file. I _could_ look through them one by one, but that's no fun. Thankfully, R has functions that perform similarly to shell functions. (More on these some other time.)

So we start by defining a data.frame with all file names and their sizes. Note that calling `unlist(map())` _inside_ `data.frame()` lets us do a lot work very quickly. We find all the files, get information about each one, and then selectively return the size.

``` r
traces_dir &lt;- "C:/Users/Daniel/Documents/Trace_Holding"

files_df &lt;- data.frame(files = list.files(traces_dir),
                       bytes = unlist(map(list.files(traces_dir), 
                                          function(abf){ &lt;http://file.info|file.info&gt;(paste0(traces_dir, "/",abf))$size })))

#              files    bytes
# 1 190808a_0000.abf 15006208 &lt;- long gap free recording
# 2 190808a_0001.abf 15006208
# 3 190808a_0002.abf 15006208
```

The experiment is embedded in the file name for each. Once again , with a little help from `unlist` we can split the file names into a list of list (i.e. `"190808a_0000.abf"` becomes `[[1]] [[1]] "190808a" [[2]] "0000.abf"` select only the first part and populate a new column.

``` r
files_df$Experiment &lt;- unlist(transpose(str_split(files_df$files, pattern = "_"))[[1]])
```


Okay, now we can make use of this. I've defined a data.frame for metadata about the experiments.

``` r
file_groups
#    Experiment       Group
# 1      190924    Baseline
# ... 
# 19     190918     Delayed
```

We can join these data frames and apply a little `tidyverse` magic to see what experiments are missing (we could also compare the sets of experiments directly).

``` r
full_join(file_groups, files_df) %&gt;% 
  filter(&lt;http://is.na|is.na&gt;(files)) %&gt;% 
  group_by(Group, Experiment) %&gt;% 
  tally()

#  Experiments that didn't transfer:

#   Group       Experiment     n
# 1 Baseline    190924         1
# 2 Baseline    190924a        1
# ...
```

We can repeat the same strategy to programattically look at the protocol types (e.g. based on file size or channel number via `&lt;http://file.info|file.info&gt;` | `readABF()`). Moral of the story, you should so stop `apply`ing yourself and give `unlist` and `purrr` functions a try.

---
title: "Placeholder"
author: "Daniel Kick"
date: 2020-6-15
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  If you run into a situation where you're using a ton of memory (e.g. manipulating transcriptomic data, resampling, or working with electrophysiology traces)  use `rm()` to selectively get rid off objects in the environment. A useful pattern is to write out large objects you'll need in the future, remove them, and then read them back in when you need them. This is usually not important, but when it is, it is.

A side note -- unlike listing items where the function matching the unix command acts on the environment and a new command acts on the files system (`ls()` and `list.files()`) the functions for removing items don't follow this logic. `rm()` acts on objects in your environment whereas `unlink()` acts on system files.

---
title: "Placeholder"
author: "Daniel Kick"
date: 2020-6-25
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  The `tictoc` library has convenience functions for timing code.  Here’s the basic usage relative to timing with base R.

``` r
library(tictoc)
tic()
# code here
toc()

tic &lt;- Sys.time()
# code here
toc &lt;- Sys.time()
print(toc - tic)
```

Where this library excels is when you want to time multiple parts of your code. Each `tic` pushes the time onto a stack and each `toc` pops the most recent time from said stack. That means you don’t have to worry about assigning several timing variables even if you want to time nested code.

``` r
tic()
# stack is 1 deep
for (i in 1:10) {
     tic()
     # stack is now 2 deep
     for (j in 1:10){
          tic()
          # stack is now 3 deep
          toc()
     }
     toc()
}
toc()
```

---
title: "Placeholder"
author: "Daniel Kick"
date: 2020-7-28
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  In tidyverse you can use `.` to reference the current dataframe. This is really useful for plotting. For example, I'd like to plot channels In7 and In12 in that order but by default In12 will come first. We could save an intermediate dataframe and re-level the factors, but by referencing the dataframe piped into mutate we can skip this step.


``` r
temp %&gt;% # temp is a down-sampled ephys recording
  ungroup() %&gt;% 
  gather(key, value, c("In7", "In12")) %&gt;% 
  mutate(key = factor(.$key, levels = c("In7", "In12"))) %&gt;%      # Note that if you have groupings, you'll need to get rid of them or supply a column of the same length as the group. 
  ggplot(aes(Time, value, color = key, group = interaction(key, Sweep)))+
  geom_path()
```


![image (11).png](image%20(11).png)

---
title: "Placeholder"
author: "Daniel Kick"
date: 2020-7-30
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  If you want to make future you more positively disposed to present you, then organize your work with `R` packages, save your custom functions in the `/R/` directory. The function `roxygen2::roxygenise()` will documentation comments for your functions into help pages. For example running `roxygenise()` with the below function saved produces the attached help page accessible via `?shrug`.


``` r
#' @title Print Shrug
#' @description This function prints a shrug emoji a specified number of times, provided the input value is a numeric greater than zero.
#' @param n how many shrugs should be printed
#' @author Daniel Kick (\email{daniel.r.kick@@gmail.com})
#' @export
#' @examples
#' shrug(5)

shrug &lt;- function(n = 1, ...){
  if (is.numeric(n) &amp; n&gt;0){
    for (i in seq(1, n)){
        cat("¯\_(ツ)_/¯
")
    }
  }
}
```


![image (12).png](image%20(12).png)

---
title: "Placeholder"
author: "Daniel Kick"
date: 2020-8-22
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  Be mindful of your data types. Sometimes `T == 1 == 1.0` (logical, int, double) but assuming these are equivalent can get you into trouble. For example, in these animations, the only difference is the data is coerced to logicals or integers.
![conway_bool.gif](conway_bool.gif)
![conway_int.gif](conway_int.gif)

---
title: "Placeholder"
author: "Daniel Kick"
date: 2020-9-19
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  If you're working on data that takes a long time to process, consider adding a the following to your analysis.

``` r
save.image(file='myEnvironment.RData')
load('myEnvironment.RData')
```

This will let you reload your environment. One can also by default save your environment when closing Rstudio but that may make it easy to reference objects that no longer are generated in the document itself thereby speeding software rot.

---
title: "Placeholder"
author: "Daniel Kick"
date: 2020-11-4
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  &lt;@U010CH4JLKE&gt; As an example to show how accessible this is, here's bit of code that resamples an anova and computes an empirical p value.

`temp` is a dataframe containing the data
`Condition` is a column with exactly that
`temp_col` is the name of a dependent variable. It's a string to make this easy to reuse.
if you haven't used `map` before it's basically a for loop that returns a list. When the output get's passed into `unlist` it becomes an array.


``` r
temp_shuffle &lt;- temp
resample_array &lt;- map(1:1000, function(i){
     temp_shuffle$Condition &lt;- sample(temp_shuffle$Condition, replace = F)
     fm &lt;- lm(as.formula(paste0(temp_col, " ~ Condition")), data = temp_shuffle)      
     return(car::Anova(fm)[1,3])
}) %&gt;% unlist()
ep &lt;- mean(resample_array &gt;= car::Anova(fm)[1,3])
```

The down side is that it takes orders of magnitude more time to run because you're running the same code hundreds or thousands of times. This is only a problem if you need crazy high precision or have a really complex/hard to fit model. For reference using the code above took about ~2 seconds/dv for 1000 iterations on my machine.

2020-11-4
Daniel
&lt;@U0103JSQYLB&gt;  A handy pattern is to use map to summarize data and then bind it.

``` r
map_res &lt;- map(names(M)[names(M) != "Sample"], function(i){
  res &lt;- shapiro.test(M[[i]])

  return(
    list(
    mrna = i,
    p = res$p.value
    )
  )
})

shapiro_res &lt;- do.call(rbind, map_res)
```

---
title: "Placeholder"
author: "Daniel Kick"
date: 2020-11-20
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  For all you classics fiends out there, R has a convenience function just for you! (As long as your numbers aren't too big)

``` r
&gt; as.numeric(as.roman("MCXXIII"))
[1] 1123
&gt; as.roman(1123)
[1] MCXXIII
&gt; as.roman(11234)
[1] &lt;NA&gt;
```

&lt;https://twitter.com/geokaramanis/status/1327286201443905538|source&gt;

---
title: "Placeholder"
author: "Daniel Kick"
date: 2020-12-7
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  Get ready to celebrate in the streets: native pipes and terse lambdas are coming to R!

&gt; *R* now provides a simple native pipe syntax `|&gt;` as well as a shorthand notation for creating functions, e.g. `\(x) x + 1` is parsed as `function(x) x + 1`. 
&lt;https://cran.r-project.org/doc/manuals/r-devel/NEWS.html&gt;


2020-12-7
David Schulz
I am a little torn though.  I LOVE me a terse lambda.

---
title: "Placeholder"
author: "Daniel Kick"
date: 2020-12-9
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  Check out `DiagrammeR` (or mermaid) if you need clean easy flow diagrams. I find they're not to hard to make and _even easier_ to forget that you've made them.


``` r
DiagrammeR::grViz("digraph {
graph [layout = dot]

# define the global styles of the nodes. We can override these in box if we wish
node [shape = circle, style = filled, fillcolor = LightSteelBlue]

data1 [label = 'Input', shape = folder, fillcolor = Beige]

glmnet [label = 'Lasso 
Regression', shape = box, fillcolor = Linen]
mnnet [label = 'Mulitnomial 
Neural 
Network', shape = box, fillcolor = Linen]
nnet [label = 'Neural 
Network', shape = box, fillcolor = Linen]
knn [label = 'k-Nearest 
Neighbor', shape = box, fillcolor = Linen]
ranger [label = 'Random 
Forest', shape = box, fillcolor = Linen]
svml [label = 'SVM 
Linear', shape = box, fillcolor = Linen]
svmr [label = 'SVM 
Radial', shape = box, fillcolor = Linen]

data2 [label = '5-fold CV 
Accuracy', shape = folder, fillcolor = Beige]

# edge definitions with the node IDs
data1 -&gt; {glmnet mnnet nnet knn ranger svml svmr} -&gt; data2
{alpha lambda} -&gt; glmnet
decay -&gt; mnnet
{size decay} -&gt; nnet
k -&gt; knn
{mtry splitrule minNodeSize} -&gt; ranger
cost -&gt; svml
sigma -&gt; svmr
}")
```

![image (15).png](image%20(15).png)

---
title: "Placeholder"
author: "Daniel Kick"
date: 2020-12-18
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  I just ran into the same issue &lt;@U010N31RYMV&gt; described after updating R on OSX —  none of the libraries are associated with the new version. I tested and abandoned directly copying the libraries over in favor programmatic reinstallation. The downside to this is that it’s slower. The upside is that if you just copy them, R will ask you to update the packages anyway.

This won’t work for libraries that aren’t on CRAN but drastically reduces the number of libraries you’re installing by hand. In my case this took care of all but about 3% of the libraries I had installed for 3.6.


``` r
all_packages &lt;- list.files("/Library/Frameworks/R.framework/Versions/3.6/Resources/library/")
installed_packages &lt;- list.files("/Library/Frameworks/R.framework/Versions/4.0/Resources/library/")
all_packages &lt;- all_packages[!(all_packages %in% installed_packages)]

options(install.packages.compile.from.source = "always")

for (package in all_packages){
  try(install.packages(package))
}

options(install.packages.compile.from.source = "interactive")
```

---
title: "Placeholder"
author: "Daniel Kick"
date: 2020-12-22
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  You can set ggplot's font using the `theme` function. Particularly if combined with functions from `ggthemes` or `ggsci` you can get very pleasing visualizations quickly. Beyond accessing fonts already on your system you can import and fonts with minimal hassle.

e.g. to get the font &lt;https://fonts.google.com/specimen/Metal+Mania?sidebar.open=true&amp;amp;selection.family=JetBrains+Mono:wght@100&amp;amp;query=metal|Metal Mania&gt; :metal: ready to use one might run:

``` r
library(showtext)
font_add_google(name = "Metal Mania", family = "Metal+Mania")
```


``` r
library(palmerpenguins)
library(tidyverse)
library(ggthemes)
library(patchwork) # for adding together plots at the end
library(extrafont)
# font_import()) # &lt;--- run this once
loadfonts(device = "win", quiet = TRUE) # &lt;--- run this once per session
# extrafont::fonts() # see fonts that are available

plt1 &lt;- palmerpenguins::penguins %&gt;%
  filter(!&lt;http://is.na|is.na&gt;(sex)) %&gt;%
  mutate(sex = case_when(sex == "male" ~ "m",
                         sex == "female" ~ "f")) %&gt;% 
  ggplot(aes(sex, body_mass_g, fill = species, group = interaction(species, sex)))+
  geom_boxplot()+
  ggthemes::scale_fill_colorblind()+
  ggthemes::theme_clean()+
  theme(legend.position = "")+
  facet_grid(.~species)+
  labs(title = "Default")

plt2 &lt;- plt1+
  theme(text = element_text(family = "Consolas"))+
  labs(title = "Consolas")

plt3 &lt;- plt1+
  theme(text = element_text(family = "Garamond"))+
  labs(title = "Garamond")

plt1 + plt2 + plt3
```


![image (17).png](image%20(17).png)

---
title: "Placeholder"
author: "Daniel Kick"
date: 2021-2-2
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  Anyone run into issues with `corrplot::corrplot()`  cutting off the title?
![image (18).png](image%20(18).png)



2021-2-2
`ggcorrplot` makes okay plots with ggplot2's logic. Not as clean as the above but it'll work with `patchwork`  and `cowplot` . Unfortunately, `scale_colour_stepsn`  doesn't override the  scaling.

``` r
library(ggcorrplot)

p.mat &lt;- cor_pmat(cor_df)
ggcorrplot(cor(cor_df, use = "pairwise.complete.obs"), 
           p.mat = p.mat,
           insig = "blank",
           type = "upper",
           outline.col = "white",
           colors = RColorBrewer::brewer.pal(n = 9, name = "PuOr")[c(1,5,9)]
           )+
  labs(title = "Brian_AP_Delayed")
```


![image (19).png](image%20(19).png)

2021-2-2
`ggcorrplot` appears to call internal functions which makes modifying it quickly impractical (one would probably be best forking the package and modifying that). I think I have a workaround that gets the same binning behavior:

After the significance matrix (`p.mat`) is generated overwrite the correlation matrix with the middle value of each desired bin.

``` r
          bkkca      cav1      cav2
bkkca 1.0000000 0.3452702 0.5603564
cav1  0.3452702 1.0000000 0.7880727
cav2  0.5603564 0.7880727 1.0000000
&gt; # bin the correlations so there are fewer colors used in the figure
&gt; cor_bins &lt;- seq(-1, 1, length.out = 9)
&gt; for (i in 1:(length(cor_bins)-1)){
+   test[test &gt; cor_bins[i] &amp;amp; test &lt; cor_bins[i+1]] &lt;- ((cor_bins[i] + cor_bins[i+1])/2)
+ }
&gt; test
      bkkca  cav1  cav2
bkkca 1.000 0.375 0.625
cav1  0.375 1.000 0.875
cav2  0.625 0.875 1.000
```

Here this makes very slight changes to the plot. (legend dropped to not imply a continuous fill)
![image (20).png](image%20(20).png)

2021-2-2
Last update, this is harder to read up will use the more extreme value to get closer to `corrplot`


``` r
test &lt;- seq(-1, 1, length.out = 5)+.0000001
test
#-0.9999999 -0.4999999  0.0000001  0.5000001  1.0000001
for (i in 1:(length(cor_bins)-1)){
  # test[test &gt; cor_bins[i] &amp;amp; test &lt; cor_bins[i+1]] &lt;- (cor_bins[i] + cor_bins[i+1])
  test[test &gt; cor_bins[i] &amp;amp; test &lt; cor_bins[i+1]] &lt;-   cor_bins[c(i, (i+1))[which.max(abs(cor_bins[i:(i+1)]))]]
}
test
#-1.00 -0.50  0.25  0.75  1.00
```


![image (21).png](image%20(21).png)

---
title: "Placeholder"
author: "Daniel Kick"
date: 2021-2-3
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  &lt;@U0103JSQYLB&gt; and I just talked through how one might compare two sets of p values. We're thinking about the following approach and wanted to see if anyone can see a flaw in this approach or has other ideas.

Setup:
You have two replicates of an experiment (`r1, r2`). In each experiment you measured three mRNAs (_`a, b, c`_) in control and treatment (`c, t`). You want to know if the same trends in `r1` show up in `r2` but there is a batch effect that will prevent comparing them directly (e.g. you can't run a t test on `r1 _a_ c` and `r2 _a_ c` )

Proposed Solution:
1. Compare each mRNA within a replicate and note the sign of change and if the p value reached a pre-determined cutoff

``` r
rep  |mrna |sign |pval |sig  |
r1   | a   | -   |0.04 |1    |
r1   | b   | +   |0.10 |0    |
...  |     |     |     |     |
r2   | c   | -   |0.02 |1    |
```

2. Multiply the sign by the significance code so that -1 = "significant decrease", 0 = "no significant change", +1 = "significant increase"

``` r
rep  |mrna |sign |pval |sig  |sxp  |
r1   | a   | -   |0.04 |1    |+1   |
r1   | b   | +   |0.10 |0    |0    |
...  |     |     |     |     |     |
r2   | c   | -   |0.02 |0    |-1   |
```

3. Reshape these as two vectors and treat them as categorical data. Then compare the "assignment" between these two lists using a jaccard index as if we were comparing an assignment from clustering against reality.

``` r
r1 &lt;- as.character(c(1, 0, 0))
r2 &lt;- as.character(c(0, 0, -1))
jaccard(r1, r2)
```

4. Use resampling to find an empirical p value for this observed jaccard index.

Does that seem reasonable? Is there another way you would go about it?

---
title: "Placeholder"
author: "Daniel Kick"
date: 2021-2-10
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  Here's an implementation for two sets of correlations. Here we bin the correlations into 5 bins use a jaccard index to assess whether the bin assignments are the same for both datasets (Brian's and mine). To confirm that the measured jaccard index (0.23) isn't anything to write home about we can generate  an empirical p value (ep = 0.13).


``` r
# cor_comp_df
#
#    Source Time     x     y        Corr
#    &lt;fct&gt;  &lt;fct&gt;    &lt;chr&gt; &lt;chr&gt;   &lt;dbl&gt;
#  1 Brian  Baseline bkkca cav1    0.441
#  2 Brian  Baseline bkkca cav2    0.476
#  3 Brian  Baseline bkkca inx1    0.435
#  4 Brian  Baseline bkkca inx2    0.159
#  5 Brian  Baseline bkkca inx3   -0.174
#  6 Brian  Baseline bkkca inx4   -0.167

n_bins &lt;- 5

temp &lt;- cor_comp_df %&gt;% 
  mutate(Bins = cut(Corr, breaks = seq(-1, 1, length.out = n_bins))) %&gt;% 
  select(-Corr) %&gt;% 
  pivot_wider(names_from = "Source", values_from = "Bins")

# temp
#
#   Time     x     y     Brian    Daniel  
#   &lt;fct&gt;    &lt;chr&gt; &lt;chr&gt; &lt;fct&gt;    &lt;fct&gt;   
# 1 Baseline bkkca cav1  (0,0.5]  (0,0.5] 
# 2 Baseline bkkca cav2  (0,0.5]  (0,0.5] 
# 3 Baseline bkkca inx1  (0,0.5]  (0.5,1] 
# 4 Baseline bkkca inx2  (0,0.5]  (0.5,1] 
# 5 Baseline bkkca inx3  (-0.5,0] (0.5,1] 
# 6 Baseline bkkca inx4  (-0.5,0] (-0.5,0]


library('clusteval')
obs_jaccard &lt;- cluster_similarity(temp$Brian, temp$Daniel, similarity="jaccard")  

# 0.2304234

null_jaccard &lt;- map(1:10000, function(i){
  cluster_similarity(sample(temp$Brian, replace = F), 
                     temp$Daniel, similarity="jaccard")
  }) %&gt;% 
  unlist()


temp &lt;- with(density(null_jaccard), data.frame(x, y))
temp &lt;- temp %&gt;% mutate(xmax = max(x),
                obs = obs_jaccard)

ggplot(data = temp, aes(x = x, y = y))+
  geom_line()+
  geom_vline(xintercept = obs_jaccard)+
  geom_ribbon(data = temp[temp$x &gt; temp$obs, ], 
              aes(xmin = obs, xmax = xmax, ymin = 0, ymax = y))+
  labs(subtitle = paste("empirical p=", as.character(mean(null_jaccard &gt;= obs_jaccard))))
```


![image (22).png](image%20(22).png)

2021-2-10
It's worth generating an empirical p value for each comparison you're making. For example here I'm comparing the results of an experiment replicate. Each dependent variable is assigned a group based on if one would conclude it there was a difference (0 or 1) between groups and the sign of that difference (+ or -). Seeing a Jaccard index of 0.61 (out of 1) we might conclude we replicated most of the findings. _However,_ the empirical p value is 1 because most of the comparisons were non-significant in both groups resulting in a high floor for the index.
![image (23).png](image%20(23).png)

---
title: "Placeholder"
author: "Daniel Kick"
date: 2021-2-21
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  I think I have a solution to get decent enough dendrograms without fussing with base graphics.

The overview of my workaround is to cluster with `pvclust`, extract `$hclust` , plot it as a dendrogram, coerce into a ggplot.  This makes it easy enough to replicate the functionality of the `colored_bars()`  function by making additional plots.  The function below makes a few plots in addition to the dendrogram. If you end up working with base graphics anyway, `dendextend` is still worth a look.

Here's an example:

``` r
# needed 
library(pvclust)
library(tidyverse)
library(dendextend) # for color_labels
library(ggnewscale) # to accommodate two fill scales 

# recommended
# install.packages("palmerpenguins")
library(palmerpenguins)
library(patchwork)
library(scales) # for overiding scientific notation on dendrogram y axis 

# Make a demo dataset
tux &lt;- select(palmerpenguins::penguins, 
              species,
              bill_length_mm, bill_depth_mm, 
              flipper_length_mm, body_mass_g) 

tux &lt;- tux[complete.cases(tux), ]

set.seed(54646)
tux &lt;- tux[sample(1:nrow(tux), 30), ] # for faster demo clustering

# Example use
o &lt;- 
  mk_hclust_plts(
    df = mutate(tux, uid = paste(seq(1, nrow(tux)), species, sep = "-")),
    cluster_by = c("bill_length_mm", "bill_depth_mm", 
                   "flipper_length_mm", "body_mass_g"),
    uid_col = "uid",
    n_clusters = 3,
    true_groups = "species",
    true_colors = RColorBrewer::brewer.pal(3, "Set2"),
    cluster_colors = RColorBrewer::brewer.pal(3, "Set1") 
  )


# Patchwork to arrange the output plots
(o$dendrogram_both+
    scale_y_continuous(limits = c(-.0001, 0.00022), labels = scales::comma)
)/ 
  (o$group_compare_tile+
     # lims(y =  c(2, -2))+ # y axis can be flipped like so
     theme(legend.position = "")
  ) / 
  (o$heatmap_raw + theme(legend.position = "right")) / 
  (o$heatmap_z + theme(legend.position = "right")) + 
  patchwork::plot_layout(heights = c(5, .3, 1.25, 1.25))


# example 2

# o &lt;- 
# mk_hclust_plts(
#   df = mutate(iris, uid = paste(seq(1, nrow(iris)), Species, sep = "-")),
#   cluster_by = c("Sepal.Length", "Sepal.Width", "Petal.Length", "Petal.Width"),
#   uid_col = "uid",
#   n_clusters = 3,
#   true_groups = "Species",
#   true_colors = RColorBrewer::brewer.pal(3, "Set2"),
#   cluster_colors = RColorBrewer::brewer.pal(3, "Set1") 
# )
```

ps, it's worth checking your code on sample datasets (`penguins`, `iris`, `mpg`, etc. ). That'll help iron out weird behavior sooner rather than later.
![image (25).png](image%20(25).png)

2021-2-21
Daniel
And here's the source for the function itself.


2021-2-21
Daniel
Edit: The original function here depended on the factor levels of the clusters and true groups to make the color's consistent between plots (e.g. factors ordered abcdABCD not aAbBcCdD). This breaks down with some cases (e.g. if you start group names with a number (e.g. 0hours)). The below edit uses `ggnewscale`  to fix this.

&lt;https://github.com/eliocamp/ggnewscale&gt;


``` r
mk_hclust_plts &lt;- function(
  df = unite(M_winxiqr, uid, Experiment, Cell, sep = "-"),
  cluster_by = c("vrest", "r11", "r1", "Ihtk.0", "Ihtk.Slope", "Ia.0", "Ia.Slope"),
  uid_col = "uid",
  n_clusters = 3,
  true_groups = "Condition",
  true_colors = RColorBrewer::brewer.pal(3, "Set2"),
  cluster_colors = RColorBrewer::brewer.pal(3, "Set1")
){
  df &lt;- as.data.frame(df) 
  
  if (!exists("cluster_colors")){
    cluster_colors = rainbow(n_clusters)
  }
  ## prep
  # move uid to rowname
  row.names(df) &lt;- df[[uid_col]]
  df_groups &lt;- select(df, all_of(true_groups))
  df &lt;- df[, cluster_by]
  
  ## Cluster
  cluster &lt;- pvclust(t(df),
                     method.hclust = "ward.D2",
                     method.dist = "correlation",
                     use.cor = "pairwise.complete.obs")
  
  
  ## make dendrogram  ####
  dend &lt;- cluster$hclust %&gt;% 
    as.dendrogram() 
  
  # iteratively coloring the labels is a workaround to get the "true" groups shown
  dend_labs &lt;- rownames_to_column(df_groups, var = "rownames")[, ]
  dend_labs &lt;- full_join(data.frame(rownames = labels(dend)), dend_labs)
  for(i in seq_along(unique(df_groups[[true_groups]]))){
    true_group &lt;- unique(df_groups[[true_groups]])[i]
    true_color &lt;- true_colors[i]
    
    dend &lt;- dend %&gt;% 
      dendextend::color_labels(
        col = true_color, 
        labels = dend_labs[dend_labs[[true_groups]] == true_group, "rownames"]) 
    
    
  }
  
  dend &lt;- dend %&gt;% 
    set("branches_k_color", 
        k = n_clusters, 
        value = cluster_colors
    ) %&gt;% 
    set("branches_lwd", 0.7) %&gt;%
    set("labels_cex", 0.6) 

  dend_cluster_only &lt;- dend %&gt;% 
    set("labels_colors",
        k = n_clusters,
        value = cluster_colors) %&gt;%
    as.ggdend()
  
  
  dend &lt;- dend %&gt;% 
    as.ggdend()
  
  
  
  plt_dend_cluster_only &lt;- ggplot(dend_cluster_only)+
    theme(axis.ticks.y = element_line(),
          axis.text.y = element_text(),
          axis.line.y = element_line())
  
  
  plt_dend &lt;- ggplot(dend)+
    theme(axis.ticks.y = element_line(),
          axis.text.y = element_text(),
          axis.line.y = element_line())  
  

  ## Add reality ribbon with or without clustering result ####
  groups_to_plt &lt;- full_join(
    as.data.frame(dend$labels),
    rownames_to_column(var = "label", df_groups))
  
  plt_grouping &lt;- groups_to_plt %&gt;% 
    ggplot(aes_string(x="x", y="0", fill = true_groups))+
    geom_tile()+
    scale_fill_manual(values = true_colors)+
    theme_void()+
    labs(x = "", y = "")+
    theme(legend.position = "left")
  
  plt_grouping_contrast &lt;- ggplot()+
    geom_tile(data = groups_to_plt, aes_string(x="x", y="0.5", fill = true_groups))+
    scale_fill_manual(values = true_colors)+
    
    ggnewscale::new_scale("fill") +
    geom_tile(data = data.frame(x = seq_along(dend_cluster_only$labels$col),
                                cluster_groups = as.character(as.numeric(as.factor(dend_cluster_only$labels$col)))
    ),
    aes_string(x="x", y= "-0.5", fill = "cluster_groups"),
    )+
    scale_fill_manual(values = cluster_colors)+
    
    theme_void()+
    labs(x = "", y = "")+
    theme(legend.position = "left")
  
  
  ## Add heatmap  ####
  data_to_plt &lt;- full_join(
    as.data.frame(dend$labels),
    rownames_to_column(var = "label", df)) 
  
  data_to_plt &lt;- 
    data_to_plt %&gt;% 
    gather("key", "value", 
           names(data_to_plt)[
             !(names(data_to_plt) %in% c("x", "y", 
                                         "label", "col", "cex", 
                                         true_groups))
           ])
  
  plt_heatmap_raw &lt;- data_to_plt %&gt;% 
    ggplot(aes(x, 
               y = key, 
               fill = value))+
    geom_tile()+
    scale_fill_viridis_c()+
    labs(x = "", y = "")+
    theme(panel.background = element_blank(),
          axis.ticks.x = element_blank(),
          axis.text.x = element_blank(),
          legend.position = "left")
  
  
  plt_heatmap_z &lt;- data_to_plt %&gt;% 
    group_by(key) %&gt;% 
    mutate(mean = mean(value, na.rm = T),
           sd = sd(value, na.rm = T)) %&gt;% 
    mutate(value = ((value - mean)/sd)) %&gt;% # Now Z scores
    ggplot(aes(x, 
               y = key, 
               fill = value))+
    geom_tile()+
    scale_fill_viridis_c()+
    labs(x = "", y = "")+
    theme(panel.background = element_blank(),
          axis.ticks.x = element_blank(),
          axis.text.x = element_blank(),
          legend.position = "left")
  
  
  ## Return plots, manually tweak layout  ####
  return(
    list(
      pvclust_out = cluster,
      dendrogram_clusters = plt_dend_cluster_only,
      dendrogram_both = plt_dend,
      group_tile = plt_grouping,
      group_compare_tile = plt_grouping_contrast,
      heatmap_raw = plt_heatmap_raw,
      heatmap_z = plt_heatmap_z
    )
  )
}
```


2021-2-21
David Schulz
I am going to have to find a reason to play around with this, because it looks great.  Thanks!  How can we ensure the world can benefit? Is this new and exciting to others as well?





2021-2-22
Daniel
It's potentially worth making available online but I don't know that it's all together too useful. The components of it make for a good starting place to write from but the function itself is fairly inflexible (e.g. in terms of clustering method and tree aesthetics). Where I think it could work nicely is in a sort of "lab reference" with recommended/first pass solutions to needs that come up with the sort of data we collect.

---
title: "Placeholder"
author: "Daniel Kick"
date: 2021-3-3
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  Does anyone have an effective way to deduplicate/shave correlation data frames prior to running a KS test? An inelegant workaround seems to be to order the data frame by correlation and retain every other row. Here's some example data either with or without these duplications.
![image (26).png](image%20(26).png)
![image (27).png](image%20(27).png)

2021-3-3
Daniel
I also thought about using a mutate and ifelse to paste the x/y values into an alphabetical identifier (e.g. shal, bkkca, 0.## would become bkkca-shal, 0.##) but doing that seems like a huge pain.

2021-3-3
David Schulz
I think this is code you generated for me, no?

2021-3-3
David Schulz

``` r
temp.list &lt;- map(unique(M$Cell), function(i){
  M %&gt;% 
    filter(Cell == i) %&gt;% 
    select(-c("Cell")) %&gt;% 
    correlate(use = "pairwise.complete.obs", method = "pearson") %&gt;% 
    shave() %&gt;% # Prevent double counting of correlations
    mutate(Cell = i)
})
```


2021-3-3
David Schulz
I'll put the whole thing in here, just to make sure.  Either you generated this for me (which it has your fingerprints on it), or I pilfered this from somewhere.

2021-3-3
David Schulz

``` r

{r}
library(corrr)
M &lt;- STGGOI

temp.list &lt;- map(unique(M$Cell), function(i){
  M %&gt;% 
    filter(Cell == i) %&gt;% 
    select(-c("Cell")) %&gt;% 
    correlate(use = "pairwise.complete.obs", method = "pearson") %&gt;% 
    shave() %&gt;% # Prevent double counting of correlations
    mutate(Cell = i)
})



M &lt;- do.call(rbind, temp.list)
M &lt;- M %&gt;% 
  gather(colname, cor, names(select(M, -c("rowname", "Cell")))) 

#M &lt;- subset(M, cor &gt;= 0.5)
plot_ecdf_ks &lt;- function(
  df = temp.df,
  data.col = "cor",
  group.col = "data",
  group1 = "0",
  group2 = "24",
  colors = c("#4d4d4d", 
             #"#67a9cf", 
             "#1c9099")) {
  
  # Adapted from:
  # &lt;https://rpubs.com/mharris/KSplot&gt;
  df &lt;- filter(df, df[[group.col]] %in% c(group1, group2))
  
  df[df[[group.col]] == group1, data.col]
  
  
  data1 &lt;- unlist(df[df[[group.col]] == group1, data.col])
  data2 &lt;- unlist(df[df[[group.col]] == group2, data.col])
  
  ecdf1 &lt;- ecdf(data1)
  ecdf2 &lt;- ecdf(data2)
  
  # used to get the most extreme difference between the two samples
  MostExtremeDiff &lt;- seq(min(data1, data2, na.rm = T), max(data1, data2, na.rm = T), length.out = length(data1))
  x0 &lt;- MostExtremeDiff[which(abs(ecdf1(MostExtremeDiff) - ecdf2(MostExtremeDiff)) == 
                                max(abs(ecdf1(MostExtremeDiff) - ecdf2(MostExtremeDiff))))]
  y0 &lt;- ecdf1(x0)
  y1 &lt;- ecdf2(x0)
  
  graph.df &lt;- data.frame(data1, data2) %&gt;% gather(Condition, Value, 1:2)
  graph.df[graph.df$Condition == "data1", "Condition"] &lt;- group1
  graph.df[graph.df$Condition == "data2", "Condition"] &lt;- group2
  
  # Run two sided KS test on data
  test.res &lt;- ks.test(data1, data2)
  
  plt &lt;- 
    ggplot(graph.df)+
    geom_segment(aes(x = x0[1], y = y0[1], xend = x0[1], yend = y1[1]),
                 linetype = "dashed", color = "black", size = 1)+
    geom_point(aes(x = x0[1] , y= y0[1]), color="black", size=2) +
    geom_point(aes(x = x0[1] , y= y1[1]), color="black", size=2) +
    stat_ecdf(aes(x = Value, group = Condition, color = Condition))+
    labs(x = "Sample", 
         y = "ECDF", 
         title = paste("K-S Test", as.character(group1), "vs", as.character(group2), 
                       "
p-value:", as.character(test.res$p.value, digits = 4)))+
    theme_minimal()+
    theme(legend.position = "bottom")+
    scale_color_manual(values = colors)#+
    # theme(text=element_text(family="Calibri Light", size=14)) 
  
  return(plt)
  
}
```


``` r


2021-3-3
Daniel
Yeah I think this is from me. `shave`  is from `library(corrr)`  and achieves this nicely. I was thinking after the data's in long format, but the solution may just be to do it on the front end instead of after it's long.

2021-3-3
David Schulz
So in other words I answered your question by facilitating a conversation between you and yourself. You're welcome.  :slightly_smiling_face:

2021-3-3
Daniel
I appreciate it.

For anyone using this in the future, `shave`  doesn't work with dataframes or the output of `cor`  so you'll need to run `as.data.frame` after shaving it.

2021-3-3
Daniel
Found some code I wrote to do this following `cor`.
This is probably only worth doing this way if you're calculating p values (for corplot) since the same strategy can be applied to a `p.mat`.


```
  cor_df &lt;- cor(cor_df, use = "pairwise.complete.obs")
  
  # Shave to diagonal
  for (j in seq(1, nrow(cor_df))){
    cor_df[j, 1:j] &lt;- NA
  }
``` r

---
title: "Placeholder"
author: "Daniel Kick"
date: 2021-3-5
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  Two tricks today:
1. Scale fill functions can accept breaks and limit arguments so you don't have to use hacky workarounds like binning the data before plotting (which is what I usually do).
2. `library(ggtext)` lets you render markdown within plots (e.g. for those pesky mRNAs)

```
&gt; df
# A tibble: 540 x 4
   term   Condition term2    Cor
   &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt;  &lt;dbl&gt;
 1 bkkca  0h        vrest -0.420
 2 cav1   0h        vrest -0.252
 3 cav2   0h        vrest -0.276
 4 inx1   0h        vrest -0.313
 5 inx2   0h        vrest -0.327
``` r

Before:

```
df %&gt;% 
  ggplot(aes(term, term2, fill = Cor))+
  geom_tile()+
  labs(x= "mRNA", y = "")+
  scale_fill_distiller(palette = "PuOr")+
  coord_fixed()
``` r

After:

```
library(ggtext)
library(glue)

df %&gt;% 
  mutate(term = glue(("&lt;i&gt;{term}&lt;/i&gt;"))) %&gt;%  
  ggplot(aes(term, term2, fill = Cor))+
  geom_tile()+
  labs(x= "mRNA", y = "")+
  
  theme(axis.text.x = element_markdown(angle = 45))+ # &lt;-- Note that we have element_markdown not element_text
  
  scale_fill_stepsn(
    colors=RColorBrewer::brewer.pal(n = 8, name = "PuOr"),
    na.value = "transparent",
    breaks=round(seq(-1, 1, length.out = 8), digits = 2),
    limits=c(-1,1)
  )+
  coord_fixed()
``` r

![image (28).png](image%20(28).png)
![image (29).png](image%20(29).png)

2021-3-5
Daniel
&lt;https://github.com/wilkelab/ggtext&gt;

---
title: "Placeholder"
author: "Daniel Kick"
date: 2021-3-9
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  Another useful trick is to pass expressions into ggplot. Here I've used the following as arguments in `labs()`.


```
c("r11", "r1", "Ihtk.0", "Ihtk.Slope", "Ia.0", "Ia.Slope", "vrest")
c(expression(M~Omega), expression(M~Omega), "nA", expression(frac(nA, mV)), "nA", expression(frac(nA, mV)), "mV" )
``` r


![image (30).png](image%20(30).png)

2021-3-9
Daniel
You can also do something like this `theme(plot.title = element_text(face="italic"))`


---


2021-3-11
Reordering a discrete axis in ggplot after generation a lot simpler than one might expect.  Rather than converting a character column to a factor (what if the data gets pivoted?), or using one column for the position and one for the labels, you can use `xlim` or `ylim`.

```
&gt; mrna_cols # desired order
# [1] "nav"    "cav1"   "cav2"   "bkkca"  "shaker" "shal"   "shab"   "shaw1" 
# [9] "shaw2"  "inx1"   "inx2"   "inx3"  

o_mrna$heatmap_z / # ggplot object within a list
o_mrna$heatmap_z+ylim(mrna_cols)
``` r


![image (31).png](image%20(31).png)

---
title: "Placeholder"
author: "Daniel Kick"
date: 2021-3-23
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  R's distribution simulation functions (e.g. `dbinom`, `runif`) make it quick and easy to double check one's intuitions. For example, I'd been thinking that under H0 the distribution of correlations from normal samples should drop off sharply as you go away from 0 such that a shift in correlation from 0 -&gt; 0.1 is much more likely than 0.8 -&gt; 0.9.


So I used `purrr::map()` to run a quick simulation. Here we simulate the null distribution based on 100,000 observations and compute the chance of a value being above 0.7. If it was uniform we would expect ~15% (.03/2) of the distribution to be here but end up with ~1.2% with the drop off.

```
set.seed(89745)
cor_check &lt;- map(1:100000, function(i){ 
  cor(rnorm(10), rnorm(10), method = "pearson")
})
cor_check &lt;- data.frame(cor = do.call(rbind,  cor_check))
ggplot(cor_check, aes(x = cor))+  geom_histogram(binwidth = 0.05)
mean(cor_check$cor &gt;= 0.7)*100
# 1.227 Percent
``` r


![image (32).png](image%20(32).png)

---
title: "Placeholder"
author: "Daniel Kick"
date: 2021-9-7
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  &lt;https://github.com/seatgeek/fuzzywuzzy|https://github.com/seatgeek/fuzzywuzzy&gt; this is a tool that isn’t necessary most of the time but when it is it can save a ton of time. It lets you do approximate string matching. I’ve used it for handling typos and differences in white space/ punctuation/ naming conventions in entry labels and it’s worked nicely. There’s a port for R and a few other languages too. 

2021-9-7
Joe
Thanks zombie Daniel 

---
title: "Placeholder"
author: "Daniel Kick"
date: 2022-5-31
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  ggplot tip: if you use an eight digit hex code for specifying a color value, the first two control transparency. Thus, you can set `fill` `=` `“#00000000”` in ggplot to get a boxplot with no fill.

2022-5-31
Here's the use case: points on top occlude the cross bar but box on top hides the observations. You could also do this by changing alpha but I think that will alter the lines as well.
![Image from iOS](Image from iOS.jpg)

---
title: "Placeholder"
author: "Daniel Kick"
date: 2023-5-23
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  R treats # as a comment in text files. 

A file I have (hapmap of snps) included one in a field of the header. If your look at the words on the first to lines with wc they will agree but R will fail to load the file. The solution is to pass in an explicit comment character like so `read.table(“snps.txt”, comment.char=‘’)`. 

---
title: "Placeholder"
author: "Daniel Kick"
date: 2024-3-21
image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_%281880%29_p.41.jpg"
categories: 
  - code
  - beginner
  - r
draft: false
editor: 
  markdown: 
    wrap: 72
---
  &lt;https://blog.moertel.com/posts/2006-01-20-wondrous-oddities-rs-function-call-semantics.html|https://blog.moertel.com/posts/2006-01-20-wondrous-oddities-rs-function-call-semantics.html&gt;

This is the sort of thing you don’t realize until it would be _really_ useful to access the name of a variable or run text as if it were code. I think the most accessible example of R’s wizardry is in plotting- you pass variables (time, mv) to plot or ggplot instead of strings (“time”, “mv”) and magically you get axis labels. R gets access to the _value_ of a variable _*and*_ its name and no one notices because it just works. :exploding_head:  ```</code></pre>
</div>
</div>
<div id="2d254a6c" class="cell" data-execution_count="35">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1"></a><span class="bu">print</span>(posts[<span class="dv">0</span>])</span>
<span id="cb43-2"><a href="#cb43-2"></a></span>
<span id="cb43-3"><a href="#cb43-3"></a></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>2020-4-1
Daniel
You can keep your working `Rmd` easier to navigate and less buggy by 1) packaging code into functions and 2) adding them to a companion `R` file. Load your functions with `source()` in the same block you load your libraries with a relative path, full path, or ideally with `here()`.

``` r
library(here)
source(here("R", "02MoniterGapJunction.R")) #here's output is effectively ../R/02MoniterGapJunction.R
```</code></pre>
</div>
</div>
<div id="92c2b94d" class="cell" data-execution_count="136">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb45-1"><a href="#cb45-1"></a><span class="cf">for</span> json_entry <span class="kw">in</span> entries:</span>
<span id="cb45-2"><a href="#cb45-2"></a></span>
<span id="cb45-3"><a href="#cb45-3"></a>    ts, name, text, files <span class="op">=</span> parse_entry(entry<span class="op">=</span>json_entry)</span>
<span id="cb45-4"><a href="#cb45-4"></a></span>
<span id="cb45-5"><a href="#cb45-5"></a>    quarto_text <span class="op">=</span> <span class="st">'</span><span class="ch">\n</span><span class="st">'</span>.join([</span>
<span id="cb45-6"><a href="#cb45-6"></a>    <span class="st">'---'</span>,</span>
<span id="cb45-7"><a href="#cb45-7"></a>    <span class="ss">f'title: "Placeholder"'</span>,</span>
<span id="cb45-8"><a href="#cb45-8"></a>    <span class="ss">f'author: "</span><span class="sc">{</span>name<span class="sc">}</span><span class="ss">"'</span>,</span>
<span id="cb45-9"><a href="#cb45-9"></a>    <span class="ss">f'date: "</span><span class="sc">{</span>ts<span class="sc">}</span><span class="ss">"'</span>,</span>
<span id="cb45-10"><a href="#cb45-10"></a>    <span class="st">'image: "https://upload.wikimedia.org/wikipedia/commons/3/3b/Alphabet_</span><span class="sc">%281880%</span><span class="st">29_p.41.jpg"'</span>,</span>
<span id="cb45-11"><a href="#cb45-11"></a>    <span class="st">'categories: '</span>,</span>
<span id="cb45-12"><a href="#cb45-12"></a>    <span class="st">'  - code'</span>,</span>
<span id="cb45-13"><a href="#cb45-13"></a>    <span class="st">'  - beginner'</span>,</span>
<span id="cb45-14"><a href="#cb45-14"></a>    <span class="st">'  - r'</span>,</span>
<span id="cb45-15"><a href="#cb45-15"></a>    <span class="st">'draft: false'</span>,</span>
<span id="cb45-16"><a href="#cb45-16"></a>    <span class="st">'editor: '</span>,</span>
<span id="cb45-17"><a href="#cb45-17"></a>    <span class="st">'  markdown: '</span>,</span>
<span id="cb45-18"><a href="#cb45-18"></a>    <span class="st">'    wrap: 72'</span>,</span>
<span id="cb45-19"><a href="#cb45-19"></a>    <span class="st">'---'</span>,</span>
<span id="cb45-20"><a href="#cb45-20"></a>    <span class="st">''</span>,</span>
<span id="cb45-21"><a href="#cb45-21"></a>    text,</span>
<span id="cb45-22"><a href="#cb45-22"></a>    <span class="st">''</span>, </span>
<span id="cb45-23"><a href="#cb45-23"></a>    (<span class="kw">lambda</span> x: <span class="st">''</span> <span class="cf">if</span> x <span class="op">==</span> [] <span class="cf">else</span> <span class="st">'</span><span class="ch">\n</span><span class="st">'</span>.join(x))(files)</span>
<span id="cb45-24"><a href="#cb45-24"></a>    ])</span>
<span id="cb45-25"><a href="#cb45-25"></a></span>
<span id="cb45-26"><a href="#cb45-26"></a>    <span class="co"># os.listdir('./')</span></span>
<span id="cb45-27"><a href="#cb45-27"></a>    <span class="co"># print(quarto_text)</span></span>
<span id="cb45-28"><a href="#cb45-28"></a></span>
<span id="cb45-29"><a href="#cb45-29"></a></span>
<span id="cb45-30"><a href="#cb45-30"></a>    dir_name <span class="op">=</span> ts.replace(<span class="st">'-'</span>, <span class="st">''</span>)<span class="op">+</span><span class="st">'_Placeholder'</span></span>
<span id="cb45-31"><a href="#cb45-31"></a></span>
<span id="cb45-32"><a href="#cb45-32"></a>    mkdir(path <span class="op">=</span> <span class="st">'./output/'</span>)</span>
<span id="cb45-33"><a href="#cb45-33"></a>    mkdir(path <span class="op">=</span> <span class="ss">f'./output/</span><span class="sc">{</span>dir_name<span class="sc">}</span><span class="ss">'</span>)</span>
<span id="cb45-34"><a href="#cb45-34"></a></span>
<span id="cb45-35"><a href="#cb45-35"></a>    files_to_mv <span class="op">=</span> [re.findall(<span class="st">'\(.*\)'</span>, e)[<span class="dv">0</span>].replace(<span class="st">'('</span>, <span class="st">''</span>).replace(<span class="st">')'</span>, <span class="st">''</span>) <span class="cf">for</span> e <span class="kw">in</span> files]</span>
<span id="cb45-36"><a href="#cb45-36"></a>    <span class="cf">if</span> files_to_mv <span class="op">!=</span> []:</span>
<span id="cb45-37"><a href="#cb45-37"></a>        <span class="cf">for</span> e <span class="kw">in</span> files_to_mv:</span>
<span id="cb45-38"><a href="#cb45-38"></a>            <span class="cf">if</span> os.path.exists(loc_assets<span class="op">+</span>e):</span>
<span id="cb45-39"><a href="#cb45-39"></a>                shutil.copy(</span>
<span id="cb45-40"><a href="#cb45-40"></a>                    loc_assets<span class="op">+</span>e, </span>
<span id="cb45-41"><a href="#cb45-41"></a>                    <span class="ss">f'./output/</span><span class="sc">{</span>dir_name<span class="sc">}</span><span class="ss">/</span><span class="sc">{</span>e<span class="sc">}</span><span class="ss">'</span>)</span>
<span id="cb45-42"><a href="#cb45-42"></a>        </span>
<span id="cb45-43"><a href="#cb45-43"></a>    <span class="cf">with</span> <span class="bu">open</span>(<span class="ss">f'./output/</span><span class="sc">{</span>dir_name<span class="sc">}</span><span class="ss">/index.qmd'</span>, <span class="st">'w'</span>) <span class="im">as</span> f:</span>
<span id="cb45-44"><a href="#cb45-44"></a>        f.write(quarto_text)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="b29f237c" class="cell" data-execution_count="94">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb46-1"><a href="#cb46-1"></a>loc_assets</span>
<span id="cb46-2"><a href="#cb46-2"></a></span>
<span id="cb46-3"><a href="#cb46-3"></a><span class="im">import</span> os</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="94">
<pre><code>'./better_code_assets/'</code></pre>
</div>
</div>
<div id="ba25c87e" class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb48-1"><a href="#cb48-1"></a><span class="cf">with</span> <span class="bu">open</span>(loc_assets<span class="op">+</span><span class="st">'temp.qmd'</span>, <span class="st">'w'</span>) <span class="im">as</span> f:</span>
<span id="cb48-2"><a href="#cb48-2"></a>    f.write(write_text)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>



</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p><a href="mailto:hello@danielkick.com">Contact Daniel</a></p>
</div>
    <div class="nav-footer-right">
<p><a href="https://github.com/danielkick/danielkick.github.io">View the Source</a></p>
</div>
  </div>
</footer>




</body></html>