---
name: Daniel R.
surname: Kick PhD
position: Research Geneticst
address: Room 213, Curtis Hall, University of Missouri-Columbia Columbia, MO 65211
phone: null
www: orcid.org/0000-0002-9002-1862
email: daniel.r.kick@protonmail.com
github: danielkick
linkedin: https://www.linkedin.com/in/daniel-kick-5a449b9a/
bibliography: cites.bib
date: "`r format(Sys.time(), '%B %Y')`"
output:  vitae::latexcv
editor_options:
  markdown:
    wrap: 72
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
library(vitae)
library(tibble)
```

# In Brief:

-   I'm interested in applying statistical and computational tools to better predict and explain biological effects. I aim to use these techniques to improve food security by assisting in generation of cultivates with high yield that are robust to variable weather.
-   During graduate school I focused on connecting ion channel mRNA abundances to electrophysiological properties and homeostatic tuning within neurons.
-   Now, as a Post-Doc with the USDA-ARS I focus on using deep learning to improve yield prediction by better representing interaction effects between genetic, enviromental, and management effects.

# Selected Skills

-   R (Tidyverse, Shiny (see
    [source](https://github.com/danielkick/MultiMethodPhysApp),
    [deployed](https://danielkick.shinyapps.io/BioSc3700_1/))

-   Python (Pandas, Numpy, Plotly)

-   Bash, git, Vim, slurm, basic singularity, basic docker

-   Statistics (Parametric, basic Bayesian modeling, Resampling)

-   Machine Learning, Deep Learning (Keras, Scikit-learn, Caret)

-   Electrophysiology (voltage, current, and dynamic clamp),
    Microdissection

-   Scientific communication, Scientific writing, Figure generation (Inkscape)


# Professional Experience

```{r}
tribble(
  ~Title,  ~Year, ~Institution, ~Desc,
  "Research Geneticist",  
  "2021-Present", 
  "USDA-ARS", 
  "Responsibilities include: Aggregating, cleaning, and imputing public datasets, tuning and training deep learning models using Keras, assisting an undergraduate run high throughput phenotyping project, and producing scripts to streamline the same. This is a post-doctoral research position overseen by Dr. Jacob D. Washburn."
  ) %>% 
  detailed_entries(Title, Year, Institution, Desc)
```

# Education

```{r}
tribble(
  ~University,
  ~Degree,
  ~Coursework,
  ~GPA,
  
"University of Missouri, Columbia, MO", 
"2021 Ph.D.", 
"Coursework included: Grant Writing, Quantitative Methods in the Life Sciences, and Machine Learning Methods for Biomedical Informatics",
"(3.97/4 GPA)",

"Truman State University, Kirksville, MO", 
"2015 B.S.", 
"Coursework included: Next Generation Sequence Data and Analysis, Bioinformatics, Analysis of Variance and Experimental Design, Non-Parametric Statistics, and Economic & Medicinal Botany.", 
"(3.65/4 GPA)"
) %>% 
  detailed_entries(
    University,
  Degree,
  Coursework,
  GPA)
```

# Short Courses and Workshops

```{r}
tribble(
  ~Title,
  ~Institution,
  ~Desc,
  ~Year,
  "Software Carpentry: Python",
"University of Missouri, Columbia, MO",
"Two-day workshop providing hands on experience working with Python, Git, and Unix Shell.",
"2020",

"Diversity & Inclusion Workshop",
"University of Missouri, Columbia, MO",
"Six-hour workshop covering implicit bias and inclusive practices.",
"2017",

"Big Data in Biology",
"University of Texas, Austin, TX",
"Weeklong short courses in bioinformatics: Introduction to Core NGS Tools and Introduction to RNA-Seq",	
"2016") %>% 
  detailed_entries(Title, Year, Institution, Desc)

```


# Research Experience

```{r}
tribble(
  ~Year,
  ~Role,
  ~Lab, 
  ~Desc,
  
  "2016—2021",
  "PhD Candidate",
  "Dr. David Schulz (University of Missouri)",
  "Analyzed the efficacy of machine learning methods to reproduce neuron cell types given mRNA abundances. (see Northcutt et al. 2019). • Described regulation of gap junction conductance controlled by timing of activity in the Cancer borealis cardiac ganglion. • Contrasted responses in cell activity, membrane currents, and channel mRNA abundances in response to potassium channel blockade after one or twenty-four hours in the Cancer borealis cardiac ganglion.",

  "2015",
  "PhD Student",
  "Dr. Lorin Milescu (University of Missouri)",
  "• Devised and tested a protocol for live imaging Drosophila with a two-photon microscope.",

  "2015",
  "PhD Student",
  "Dr. Bing Zhang (University of Missouri)",
  "• Designed a machine to assay Drosophila climbing (used in Willenbrink et al. 2016) and quantified sleep patterns of Drosophila tyrosine hydroxylase mutants.",

"2014-2015",
"Undergraduate Research Assistant",
"Dr. Diane Janick-Buckner & Dr. Brent Buckner (Truman State University) ",
"• Prototyped a hydroponic growth chamber for root morphology characterization of maize mutants.",


"2014",
"REU Student",
"Dr. Rahul Kanadia (University of Connecticut)",
" • Conduced a pilot study using in situ hybridization which implicated upregulation of the minor spliceosome in postponing retinal cell death. ",

"2013",
"REU Student",
"Dr. Christian Lorsen (University of Missouri)",
" • Measured motor function decline in a mouse model of spinal muscular atrophy receiving an oligonucleotide treatment.",

"2011", 
"Student Volunteer", 
"Dr. Laszlo Kovacs (Missouri State University)", 
"• Refined a set of teaching labs to be used in a genetics course."
) %>%
 detailed_entries(Role, Year, Lab, Desc)
```

# Teaching Experience

-   Teaching Assistant (Animal Physiology Lab, University of Missouri)

-   Oversaw and coordinated transition of the lab to a hybrid then fully
    online class model due to Covid-19.

-   Adapted and expanded materials to be suitable for remote learning.

-   Constructed and deployed a statistics web tool (source, deployed).
    2018 -- 2021 Curriculum Consultant (Animal Physiology Lab,
    University of Missouri) 2018

-   Updated curriculum and redesigned experiments placing a greater
    focus primary literature and data analysis.

Teaching Assistant (Animal Physiology Lab, University of Missouri) •
Assisted students through experiments. • Provided short weekly lectures.
2015 -- 2016"

# Honors and Awards

```{r}
tribble(
  ~Title, ~Location, ~Awards, ~Year,
  "J. Perry Gustafson Award", 
  "University of Missouri ", 
  "$2000", 
  "2019", 
  
  "NIH T32 Training Grant Recipient", 
  "", 
  "$27,000 yearly", 
  "2016 –2018", 
  
  "NIH T32 Travel Award", 
  "", 
  "$750", 
  "2017",
  
  "NIH T32 Travel Award", 
  "", 
  "$750", 
  "2016",
  
  "Graduated Cum Laude", 
  "Truman State University", 
  "",
  "2015"  
  )  %>% 
  detailed_entries(Title, Year, Location, Awards)

```

# Outreach and Communication

```{r}
tribble(
  ~Title, ~Type, ~Location, ~Year,
  "Tools and Techniques for a Jupyter Based Scientific Workflow", 
  "Invited Workshop, Bioinformatics in Plant Science (BIPS)",
  "University of Missouri",
  "2022",
    
  "Can mRNA expression recapitulate neuron cell types?", 
  "Alumni Research Presentation", 
  "Truman State University", 
  "2019",
  
  "Spare the synapse, spoil the circuit", 
  "Public presentation, Science on Tap", 
  "Columbia Missouri", 
  "2019", 
  
  "Voltage Dependent modification of Electrical Synapses", 
  "Biological Sciences Recruitment Poster Session", 
  "University of Missouri", 
  "2019", 
  
  "Gap Junction Conductance Modulation Via Voltage", 
  "Alumni Research Presentation", 
  "Truman State University", 
  "2018", 
  
  "Please mind the gap: Network homeostatic plasticity in the Cancer borealis cardiac ganglion", 
  "Alumni Research Presentation", 
  "Truman State University", 
  "2017", 
  	
  "The Tell-Tale Heart: Applying crustacean neurogenic hearts to basic neurosciences questions",
  "Alumni Research Presentation", 
  "Truman State University", 
  "2016", 
  
  "Scientific Poster Judge", 
  "Spring Undergraduate Research and Creative Achievements Forum", 
  "University of Missouri", 
  "2016")  %>% 
  detailed_entries( Title, Year, Type, Location)

```

# Mentoring

```{r}
tribble(~Student, ~Type, ~Year,
"Abby Beckerdite", "Undergraduate Student",      "2017 – 2019", 
"Ayla Ross",       "NSF REU Student",            "2018 – 2018", 
"Katlyn Sullivan", "NSF REU Student",            "2017 – 2017", 
"Rody Kingston",   "Post-baccalaureate Scholar", "2016 – 2016", 
"Kelly Hiersche",  "NSF REU Student",            "2016 – 2016"
) %>% detailed_entries(Type, Year, Student)
```

# Publications

```{r}
tribble(
  ~Title, ~Author, ~Year, ~Journal, 
  
  ## ====
	"Timing dependent potentiation and depression of electrical synapses contributes to network stability in the crustacean cardiac ganglion",
	"Kick, Daniel R. and Schulz, David J.",
	"2022",
	"Journal of Neuroscience", 
  
  ## ====
  "Studying gap junctions with PARIS",
  "Kick, Daniel R and Schulz, David J",
   "2019",  
  "Elife",
  
    ## ====
  "Molecular profiling of single neurons of known identity in two
               ganglia from the crab Cancer borealis",
  "Northcutt, Adam J and Kick, Daniel R and Otopalik, Adriane G and
               Goetz, Benjamin M and Harris, Rayna M and Santin, Joseph M and
               Hofmann, Hans A and Marder, Eve and Schulz, David J",
   "2019",
  "Proc. Natl. Acad. Sci. U.S.A.",
  
  ## ====
  "Dopamine maintains network synchrony via direct modulation of
               gap junctions in the crustacean cardiac ganglion",
  "Lane, Brian J and Kick, Daniel R and Wilson, David K and Nair,
               Satish S and Schulz, David J",
   "2018",
  "Elife",

  ## ====
  "Variability in neural networks",
  "Daniel R Kick; David J Schulz",
"2018", 
"Elife",
  
  
  ## ====
  "The Hillary Climber trumps manual testing: an automatic system for studying Drosophila climbing",
  "Alex M. Willenbrink and Margo K. Gronauer and Leon F. Toebben and Daniel R. Kick and Madalyn Wells and Bing Zhang",
  "Journal of Neurogenetics",
  "2016"

) %>% detailed_entries(Title, Author, Year, Journal)



```


```{r}
# library(dplyr)
# # knitr::write_bib(, "packages.bib")
# bibliography_entries("cites.bib") %>%
#   arrange(desc(author$family), issued)
```
